{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting xlrd\n",
      "  Downloading xlrd-1.2.0-py2.py3-none-any.whl (103 kB)\n",
      "\u001b[K     |████████████████████████████████| 103 kB 527 kB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: xlrd\n",
      "Successfully installed xlrd-1.2.0\n"
     ]
    }
   ],
   "source": [
    "# pip install pybedtools==0.8.1\n",
    "# pip install pandas\n",
    "# pip install matplotlib\n",
    "!pip install xlrd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import pandas as pd\n",
    "import os,sys\n",
    "import pybedtools as pbt\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import time\n",
    "from mapper import expand, parse_mapping_table, apply_mappers\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "chr_dict = dict(zip(range(1,22),map(str,range(1,22))))\n",
    "chr_dict.update({22: 'X', 23: \"Y\"})\n",
    "\n",
    "# root_dir = \"/home/nisahopcu/\"\n",
    "# gene_coords_file = root_dir + \"ref_GRCh37.p5_top_level.gff3.bed\" # must contain chromosome, start, end and Entrez Gene ID for hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_marker_thr = 5\n",
    "# to detect 1 copy gains or losses presenting at CCF >= 0.3\n",
    "pos_seg_mean_thr = 0.20\n",
    "neg_seg_mean_thr = -0.23\n",
    "preprocessed_dir = \"preprocessed/CNA/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "### functions for TCGA and CCLE #################################\n",
    "def filter_lowconf_segments(df,num_marker_thr,pos_seg_mean_thr, neg_seg_mean_thr ):\n",
    "    # filter low-confidence segments with too few probes\n",
    "    df = df[df[\"Num_Probes\"] >= num_marker_thr ]\n",
    "    # filter low-confidence segments with Segment_Mean too close to zero:\n",
    "    df = df[ (df[\"Segment_Mean\"] >= pos_seg_mean_thr) | (df[\"Segment_Mean\"] <= neg_seg_mean_thr)]\n",
    "    return df\n",
    "\n",
    "def sample_type(barcode):\n",
    "    if barcode[13:16] in [\"10A\",\"10B\",\"11A\",\"11B\",\"10C\",\"11C\"]:\n",
    "        return \"Normal\"\n",
    "    else:\n",
    "        return \"Tumor\"\n",
    "\n",
    "def find_matching_normal(tumor_barcode,barcodes_list):\n",
    "    patient_id = tumor_barcode[:12]\n",
    "    normal_barcodes = []\n",
    "    for barcode in barcodes_list:\n",
    "        if barcode.startswith(patient_id) and sample_type(barcode) == \"Normal\":\n",
    "            normal_barcodes.append(barcode)\n",
    "    return normal_barcodes\n",
    "\n",
    "def cnv2bed(seg):\n",
    "    #cnv_bed = seg[[\"Chromosome\",\"Start\",\"End\",\"Segment_Mean\"]]\n",
    "    #cnv_bed.columns = [\"chrom\",\"start\",\"stop\",\"Segment_Mean\"]\n",
    "    cnv_bed = seg.rename({\"Chromosome\":\"chrom\",\"Start\":\"start\",\n",
    "                          \"End\":\"stop\"},axis=\"columns\")\n",
    "    cnv_bed = cnv_bed.loc[:,[\"chrom\",\"start\",\"stop\",\"Segment_Mean\",\"Sample\",\"Num_Probes\"]]\n",
    "    return  pbt.BedTool.from_dataframe(cnv_bed)\n",
    "\n",
    "def bed2cnv(cnv_bed):\n",
    "    cnv_bed = str(cnv_bed)\n",
    "    if len(cnv_bed) > 0:\n",
    "        seg = pd.read_csv(StringIO(cnv_bed),sep = \"\\t\",header=None)\n",
    "        seg.columns = [\"Chromosome\",\"Start\",\"End\",\"Segment_Mean\",\"Sample\",\"Num_Probes\"]\n",
    "        seg = seg.loc[:,[\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"]]\n",
    "    else:\n",
    "        seg = pd.DataFrame(columns=[\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"])\n",
    "    return seg\n",
    "def remove_ovelapping_segments(tumor, normal,sample_name):\n",
    "    tumor_bed = cnv2bed(tumor)\n",
    "    normal_bed = cnv2bed(normal)\n",
    "    tumor_wo_germline = tumor_bed.subtract(normal_bed,r=True,f=0.8,A=True)\n",
    "    tumor_wo_germline  = bed2cnv(tumor_wo_germline)\n",
    "    n_segs_removed = tumor.shape[0] - tumor_wo_germline.shape[0]\n",
    "    if n_segs_removed*1.0/tumor.shape[0] > 0.5 and n_segs_removed>5 :\n",
    "        print(n_segs_removed,\"of\",tumor.shape[0],\"segments removed in\",sample_name,\"due to overlap with normal\",file = sys.stderr)\n",
    "    return tumor_wo_germline\n",
    "\n",
    "def cnv2genelevel(cnv_bed,gene_intervals_bed,sample_name,verbose = True,sorted_index = \"\"):\n",
    "    \n",
    "    # intersect \n",
    "    cnv2gene = str(gene_intervals_bed.intersect(cnv_bed,wb = True,wa=True))\n",
    "    if len(cnv2gene)==0: # if no intersection, return all zeroes\n",
    "        print(sample_name,\"has no genes with altered CN\",file = sys.stderr)\n",
    "        return pd.DataFrame(columns=[sample])\n",
    "    cnv2gene = pd.read_csv(StringIO(cnv2gene),sep = \"\\t\",header=None)\n",
    "    cnv2gene = cnv2gene[[3,7]].copy()\n",
    "    cnv2gene.columns = [\"gene\",\"Segment_Mean\"] \n",
    "    \n",
    "    # find genes overlapping with more than one segment:\n",
    "    # take the most exterme segement_mean value\n",
    "    \n",
    "    dups = cnv2gene.loc[cnv2gene.duplicated(subset=[\"gene\"],keep=False),]\n",
    "    if dups.shape[0] > 0:\n",
    "        cnv2gene = cnv2gene.drop_duplicates(subset=[\"gene\"],keep=False)\n",
    "        dups[\"abs_seg_mean\"] = abs(dups[\"Segment_Mean\"])\n",
    "        if verbose:\n",
    "            print(sample_name,\"contain \",len(set(dups[\"gene\"].values)),\"genes overalpped with more than one segment\",file=sys.stderr)\n",
    "            #print(dups.head(10),file=sys.stderr)\n",
    "        dups = dups.groupby(['gene'], group_keys=False).apply(lambda row: row.loc[row['abs_seg_mean'].idxmax()])\n",
    "        cnv2gene = pd.concat([cnv2gene,dups],sort=False)\n",
    "\n",
    "    cnv2gene = cnv2gene[[\"gene\",\"Segment_Mean\"]]\n",
    "    cnv_baseline = cnv2gene.copy()\n",
    "    cnv_baseline[\"Segment_Mean\"] = [0]*cnv_baseline.shape[0]\n",
    "    cnv_baseline = cnv_baseline[[\"gene\",\"Segment_Mean\"]]\n",
    "    cnv_baseline.set_index(\"gene\",inplace=True,drop=True)\n",
    "    cnv_baseline.sort_index(inplace=True)\n",
    "    sorted_index = list(cnv_baseline.index.values)\n",
    "    \n",
    "    cnv2gene.set_index(\"gene\",inplace=True,drop=True)\n",
    "    cnv2gene.rename(int,axis=0,inplace=True)\n",
    "    # add copy-neutral genes with 0s\n",
    "    cnv2gene = cnv2gene.loc[sorted_index,:]\n",
    "    cnv2gene.columns = [sample_name]\n",
    "    return cnv2gene\n",
    "\n",
    "\n",
    "### functions for GDSC and PDX #################################\n",
    "\n",
    "def CN2log2R(col, median_ploidy=2 ):\n",
    "    # this is fr GDSC only\n",
    "    lRs = []\n",
    "    genes = col.index.values\n",
    "    for code in col.values:\n",
    "        if not code == \"-1,-1,-,-\":\n",
    "            [max_cn,min_cn,zygosity,disruption] = code.split(\",\")\n",
    "            if int(max_cn) == 0:\n",
    "                lRs.append(-4.32) # CN=0 with 95% purity\n",
    "            else:\n",
    "                max_lR = np.log2(float(max_cn)/median_ploidy)\n",
    "                if not disruption == \"D\":\n",
    "                    lRs.append(max_lR)\n",
    "                else:\n",
    "                    if int(min_cn) == 0:\n",
    "                        min_lR = -4.32\n",
    "                    else:\n",
    "                        min_lR = np.log2(float(min_cn)/median_ploidy)\n",
    "                    if abs(min_lR) > abs(max_lR):\n",
    "                        lRs.append(min_lR)\n",
    "                    else:\n",
    "                        lRs.append(max_lR)\n",
    "                \n",
    "        else:\n",
    "            lRs.append(np.NaN)\n",
    "    return pd.Series(dict(zip(genes, lRs)))\n",
    "\n",
    "def define_avg_ploidy(col):\n",
    "    n,pl = 0,0\n",
    "    CN_non_disrupted = []\n",
    "    for code in col.values:\n",
    "        if not code == \"-1,-1,-,-\":\n",
    "            [max_cn,min_cn,zygosity,disruption] = code.split(\",\")\n",
    "            n+=1\n",
    "            cn = (int(max_cn)+int(min_cn))*0.5\n",
    "            pl += cn\n",
    "            if not disruption == \"D\":\n",
    "                CN_non_disrupted.append((cn))\n",
    "    return pd.Series({\"avg_pl\":pl/n , \"median_pl\":np.median(CN_non_disrupted)})\n",
    "\n",
    "def clean_logR(logR_value, pos_seg_mean_thr, neg_seg_mean_thr):\n",
    "    if logR_value >= pos_seg_mean_thr:\n",
    "        return logR_value \n",
    "    elif logR_value <= neg_seg_mean_thr:\n",
    "            return logR_value \n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "def handle_dups(df,corr_thr = 0.75):\n",
    "    '''Detect dupliated row IDs. Merge 2 or more rows with the same ID, \n",
    "    if averaged correlation in all pairvise comparision is >= corr_thhr;\\n\n",
    "    otherwise drop all duplicates.  Keeps abs. max value (negative preferred).'''\n",
    "    dups = df.index\n",
    "    dups = list(set(dups[dups.duplicated()]))\n",
    "    if len(dups)==0:\n",
    "        print(\"No duplicated row IDs. Do nothing.\")\n",
    "        return df\n",
    "    print(len(dups), \"duplicated IDs in\",df.loc[dups,:].shape[0],\"rows found.\")\n",
    "    dups_merge = [] # if corr > corr_thr\n",
    "    dups_remove = [] # corr < \n",
    "    for dup in dups:\n",
    "        r = df.loc[dup,:].T.corr()\n",
    "        n_dups = df.loc[dup,:].shape[0]\n",
    "        r_avg = []\n",
    "        for i in range(0,n_dups):\n",
    "            for j in range(i+1,n_dups):\n",
    "                r_avg.append(r.iloc[i,j])\n",
    "        if np.average(r_avg) < corr_thr :\n",
    "            #print(dup,r_avg, n_dups)\n",
    "            dups_remove.append(dup)\n",
    "        else:\n",
    "            dups_merge.append(dup)\n",
    "    \n",
    "    # remove not similar duplicates\n",
    "    df_size = df.shape[0]\n",
    "    df = df.loc[~df.index.isin(dups_remove),:]\n",
    "    print(\"duplicate rows removed due to low correlation of duplicated profiles\",df_size -df.shape[0] )\n",
    "    df_size = df.shape[0]\n",
    "    \n",
    "    # merge simialr duplicates\n",
    "    d1 = df.loc[~df.index.isin(dups_merge),:]\n",
    "    d2 = df.loc[dups_merge,:]\n",
    "    d2 = d2.groupby(d2.index).agg(lambda x: -max(-x.max(),-x.min(),key= abs))\n",
    "    df = pd.concat([d1,d2])\n",
    "    df.sort_index(inplace=True)\n",
    "    print(\"Merged \",df_size-df.shape[0]+len(dups_merge),\"duplicated rows into\",len(dups_merge),\"rows\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "segments in tumor 204 segemtns in normal 121\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57803</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>456120</td>\n",
       "      <td>8896255</td>\n",
       "      <td>4489.0</td>\n",
       "      <td>-0.0113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57804</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>8899400</td>\n",
       "      <td>8899668</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.3344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57805</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>8900394</td>\n",
       "      <td>126596817</td>\n",
       "      <td>67487.0</td>\n",
       "      <td>0.0010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57806</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>126596926</td>\n",
       "      <td>127130276</td>\n",
       "      <td>453.0</td>\n",
       "      <td>-1.0306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57807</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>127132920</td>\n",
       "      <td>128342803</td>\n",
       "      <td>864.0</td>\n",
       "      <td>-0.0031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57808</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>128342819</td>\n",
       "      <td>128350888</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.2824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57809</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>128353007</td>\n",
       "      <td>134142530</td>\n",
       "      <td>3708.0</td>\n",
       "      <td>0.0082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome      Start        End  \\\n",
       "57803  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11     456120    8896255   \n",
       "57804  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11    8899400    8899668   \n",
       "57805  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11    8900394  126596817   \n",
       "57806  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  126596926  127130276   \n",
       "57807  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  127132920  128342803   \n",
       "57808  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  128342819  128350888   \n",
       "57809  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  128353007  134142530   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57803      4489.0       -0.0113  \n",
       "57804         3.0       -1.3344  \n",
       "57805     67487.0        0.0010  \n",
       "57806       453.0       -1.0306  \n",
       "57807       864.0       -0.0031  \n",
       "57808        44.0        0.2824  \n",
       "57809      3708.0        0.0082  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#file_path = \"../../TCGA/CNA/data/gdac.broadinstitute.org_CESC.Merge_snp__genome_wide_snp_6__broad_mit_edu__Level_3__segmented_scna_minus_germline_cnv_hg19__seg.Level_3.2015082100.0.0/CESC.snp__genome_wide_snp_6__broad_mit_edu__Level_3__segmented_scna_minus_germline_cnv_hg19__seg.seg.txt\"\n",
    "\n",
    "file_path = \"TCGA/CNA/data__2016_01_28/gdac.broadinstitute.org_CESC.Merge_snp__genome_wide_snp_6__broad_mit_edu__Level_3__segmented_scna_minus_germline_cnv_hg19__seg.Level_3.2016012800.0.0/CESC.txt\"\n",
    "df = pd.read_csv(file_path, sep = \"\\t\")\n",
    "tumor_barcode = \"TCGA-ZJ-AAXJ-01A-11D-A42N-01\"\n",
    "t = df.loc[df[\"Sample\"]==tumor_barcode,:]\n",
    "t_shape = t.shape[0]\n",
    "n = find_matching_normal(tumor_barcode,list(set(df[\"Sample\"].values)))\n",
    "n = df.loc[df[\"Sample\"]==n[0],:]\n",
    "print(\"segments in tumor\",t.shape[0],\"segemtns in normal\",n.shape[0])\n",
    "\n",
    "n.loc[n['Chromosome']==11,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57960</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>456120</td>\n",
       "      <td>64200041</td>\n",
       "      <td>34710.0</td>\n",
       "      <td>0.0054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57961</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64208988</td>\n",
       "      <td>64319750</td>\n",
       "      <td>61.0</td>\n",
       "      <td>-0.6748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57962</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64325209</td>\n",
       "      <td>126596817</td>\n",
       "      <td>37207.0</td>\n",
       "      <td>0.0571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57963</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>126596926</td>\n",
       "      <td>127130276</td>\n",
       "      <td>454.0</td>\n",
       "      <td>-1.0760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57964</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>127132920</td>\n",
       "      <td>132080656</td>\n",
       "      <td>3591.0</td>\n",
       "      <td>0.0449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57965</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132080885</td>\n",
       "      <td>132099465</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-0.6123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57966</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132099856</td>\n",
       "      <td>134142530</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>0.0483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome      Start        End  \\\n",
       "57960  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11     456120   64200041   \n",
       "57961  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64208988   64319750   \n",
       "57962  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64325209  126596817   \n",
       "57963  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  126596926  127130276   \n",
       "57964  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  127132920  132080656   \n",
       "57965  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132080885  132099465   \n",
       "57966  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132099856  134142530   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57960     34710.0        0.0054  \n",
       "57961        61.0       -0.6748  \n",
       "57962     37207.0        0.0571  \n",
       "57963       454.0       -1.0760  \n",
       "57964      3591.0        0.0449  \n",
       "57965        15.0       -0.6123  \n",
       "57966      1010.0        0.0483  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.loc[t[\"Chromosome\"] ==11,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "segemtns in normal after dropping low.conf.: 38\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57804</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>8899400</td>\n",
       "      <td>8899668</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.3344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57806</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>126596926</td>\n",
       "      <td>127130276</td>\n",
       "      <td>453.0</td>\n",
       "      <td>-1.0306</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome      Start        End  \\\n",
       "57804  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11    8899400    8899668   \n",
       "57806  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  126596926  127130276   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57804         3.0       -1.3344  \n",
       "57806       453.0       -1.0306  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = filter_lowconf_segments(n,0,0.46, -0.68 )\n",
    "print(\"segemtns in normal after dropping low.conf.:\",n.shape[0])\n",
    "n.loc[n['Chromosome']==11,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "segemtns in tumor after removing germlines: 194\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>456120</td>\n",
       "      <td>64200041</td>\n",
       "      <td>34710.0</td>\n",
       "      <td>0.0054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64208988</td>\n",
       "      <td>64319750</td>\n",
       "      <td>61.0</td>\n",
       "      <td>-0.6748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64325209</td>\n",
       "      <td>126596817</td>\n",
       "      <td>37207.0</td>\n",
       "      <td>0.0571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>127132920</td>\n",
       "      <td>132080656</td>\n",
       "      <td>3591.0</td>\n",
       "      <td>0.0449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132080885</td>\n",
       "      <td>132099465</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-0.6123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132099856</td>\n",
       "      <td>134142530</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>0.0483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Sample  Chromosome      Start        End  \\\n",
       "96   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11     456120   64200041   \n",
       "97   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64208988   64319750   \n",
       "98   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64325209  126596817   \n",
       "99   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  127132920  132080656   \n",
       "100  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132080885  132099465   \n",
       "101  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132099856  134142530   \n",
       "\n",
       "     Num_Probes  Segment_Mean  \n",
       "96      34710.0        0.0054  \n",
       "97         61.0       -0.6748  \n",
       "98      37207.0        0.0571  \n",
       "99       3591.0        0.0449  \n",
       "100        15.0       -0.6123  \n",
       "101      1010.0        0.0483  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = remove_ovelapping_segments(t, n,tumor_barcode)\n",
    "print(\"segemtns in tumor after removing germlines:\",t.shape[0])\n",
    "t.loc[t[\"Chromosome\"] ==11,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "segemtns in tumor after dropping low.conf.: 101\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64208988</td>\n",
       "      <td>64319750</td>\n",
       "      <td>61.0</td>\n",
       "      <td>-0.6748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132080885</td>\n",
       "      <td>132099465</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-0.6123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Sample  Chromosome      Start        End  \\\n",
       "97   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64208988   64319750   \n",
       "100  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132080885  132099465   \n",
       "\n",
       "     Num_Probes  Segment_Mean  \n",
       "97         61.0       -0.6748  \n",
       "100        15.0       -0.6123  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = filter_lowconf_segments(t,num_marker_thr,pos_seg_mean_thr, neg_seg_mean_thr )\n",
    "print(\"segemtns in tumor after dropping low.conf.:\",t.shape[0])\n",
    "t.loc[t[\"Chromosome\"] ==11,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TCGA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACC samples: 180 CNA events per sample on avg.: 116.95555555555555\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 180 tumors: 90 normals: 90\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 88\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "ACC samples: 90 Segments per sample on avg.: 106.42222222222222\n",
      "BLCA-FFPE samples: 2 CNA events per sample on avg.: 1064.0\n",
      "after filtering\n",
      "BLCA-FFPE samples: 2 Segments per sample on avg.: 368.5\n",
      "BRCA-FFPE samples: 7 CNA events per sample on avg.: 677.5714285714286\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 2 tumors: 2 normals: 0\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 7 tumors: 7 normals: 0\n",
      "\ttumors without matched normal 7\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "BRCA-FFPE samples: 7 Segments per sample on avg.: 184.42857142857142\n",
      "CESC samples: 586 CNA events per sample on avg.: 101.45051194539249\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 586 tumors: 297 normals: 289\n",
      "\ttumors without matched normal 16\n",
      "\ttumors with at least one sCNA 280\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "CESC samples: 296 Segments per sample on avg.: 58.38513513513514\n",
      "CHOL samples: 85 CNA events per sample on avg.: 89.05882352941177\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 85 tumors: 36 normals: 49\n",
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 36\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "CHOL samples: 36 Segments per sample on avg.: 56.80555555555556\n",
      "COAD-FFPE samples: 12 CNA events per sample on avg.: 379.5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 12 tumors: 12 normals: 0\n",
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 96 tumors: 52 normals: 44\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "COAD-FFPE samples: 12 Segments per sample on avg.: 100.08333333333333\n",
      "DLBC samples: 96 CNA events per sample on avg.: 97.32291666666667\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 10\n",
      "\ttumors with at least one sCNA 40\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "DLBC samples: 50 Segments per sample on avg.: 44.72\n",
      "ESCA samples: 373 CNA events per sample on avg.: 163.01072386058982\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 373 tumors: 185 normals: 188\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 181\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "ESCA samples: 184 Segments per sample on avg.: 142.08152173913044\n",
      "GBM samples: 1104 CNA events per sample on avg.: 133.018115942029\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1104 tumors: 590 normals: 514\n",
      "\ttumors without matched normal 78\n",
      "\ttumors with at least one sCNA 511\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "GBM samples: 589 Segments per sample on avg.: 70.35144312393888\n",
      "HNSC samples: 1089 CNA events per sample on avg.: 101.27548209366391\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1089 tumors: 530 normals: 559\n",
      "\ttumors without matched normal 28\n",
      "\ttumors with at least one sCNA 498\n",
      "\ttumors without any somatic CNA 4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "HNSC samples: 526 Segments per sample on avg.: 60.77186311787072\n",
      "KICH samples: 132 CNA events per sample on avg.: 77.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 132 tumors: 66 normals: 66\n",
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 66\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 3 tumors: 3 normals: 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "KICH samples: 66 Segments per sample on avg.: 51.06060606060606\n",
      "KIRC-FFPE samples: 3 CNA events per sample on avg.: 284.0\n",
      "after filtering\n",
      "KIRC-FFPE samples: 3 Segments per sample on avg.: 59.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIRP samples: 590 CNA events per sample on avg.: 79.51525423728813\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 590 tumors: 288 normals: 302\n",
      "\ttumors without matched normal 15\n",
      "\ttumors with at least one sCNA 271\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "KIRP samples: 286 Segments per sample on avg.: 22.16083916083916\n",
      "LAML samples: 380 CNA events per sample on avg.: 74.53684210526316\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 380 tumors: 191 normals: 189\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 172\n",
      "\ttumors without any somatic CNA 16\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LAML samples: 175 Segments per sample on avg.: 7.171428571428572\n",
      "LGG samples: 1015 CNA events per sample on avg.: 78.61182266009853\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1015 tumors: 530 normals: 485\n",
      "\ttumors without matched normal 33\n",
      "\ttumors with at least one sCNA 494\n",
      "\ttumors without any somatic CNA 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LGG samples: 527 Segments per sample on avg.: 29.286527514231498\n",
      "LIHC samples: 760 CNA events per sample on avg.: 122.8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 760 tumors: 373 normals: 387\n",
      "\ttumors without matched normal 21\n",
      "\ttumors with at least one sCNA 349\n",
      "\ttumors without any somatic CNA 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LIHC samples: 370 Segments per sample on avg.: 81.1891891891892\n",
      "LUAD-FFPE samples: 12 CNA events per sample on avg.: 739.8333333333334\n",
      "after filtering\n",
      "LUAD-FFPE samples: 12 Segments per sample on avg.: 185.91666666666666\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 12 tumors: 12 normals: 0\n",
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LUSC samples: 1032 CNA events per sample on avg.: 130.68217054263565\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1032 tumors: 501 normals: 531\n",
      "\ttumors without matched normal 23\n",
      "\ttumors with at least one sCNA 476\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LUSC samples: 499 Segments per sample on avg.: 94.85571142284569\n",
      "MESO samples: 172 CNA events per sample on avg.: 106.59883720930233\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 172 tumors: 87 normals: 85\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 83\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "MESO samples: 85 Segments per sample on avg.: 60.35294117647059\n",
      "OV samples: 1168 CNA events per sample on avg.: 224.04109589041096\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1168 tumors: 597 normals: 571\n",
      "\ttumors without matched normal 26\n",
      "\ttumors with at least one sCNA 571\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "OV samples: 597 Segments per sample on avg.: 208.14405360134003\n",
      "PAAD samples: 365 CNA events per sample on avg.: 95.36438356164383\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 365 tumors: 185 normals: 180\n",
      "\ttumors without matched normal 10\n",
      "\ttumors with at least one sCNA 161\n",
      "\ttumors without any somatic CNA 14\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "PAAD samples: 171 Segments per sample on avg.: 32.60233918128655\n",
      "PCPG samples: 346 CNA events per sample on avg.: 90.33526011560694\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 346 tumors: 168 normals: 178\n",
      "\ttumors without matched normal 6\n",
      "\ttumors with at least one sCNA 161\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "PCPG samples: 167 Segments per sample on avg.: 43.37125748502994\n",
      "PRAD-FFPE samples: 6 CNA events per sample on avg.: 453.1666666666667\n",
      "after filtering\n",
      "PRAD-FFPE samples: 6 Segments per sample on avg.: 110.83333333333333\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 6 tumors: 6 normals: 0\n",
      "\ttumors without matched normal 6\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "READ samples: 316 CNA events per sample on avg.: 113.18037974683544\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 316 tumors: 166 normals: 150\n",
      "\ttumors without matched normal 23\n",
      "\ttumors with at least one sCNA 141\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "READ samples: 164 Segments per sample on avg.: 70.1280487804878\n",
      "SARC samples: 513 CNA events per sample on avg.: 208.0682261208577\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 513 tumors: 263 normals: 250\n",
      "\ttumors without matched normal 17\n",
      "\ttumors with at least one sCNA 245\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "SARC samples: 262 Segments per sample on avg.: 187.37786259541986\n",
      "SKCM samples: 937 CNA events per sample on avg.: 115.35112059765208\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 937 tumors: 472 normals: 465\n",
      "\ttumors without matched normal 7\n",
      "\ttumors with at least one sCNA 463\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "SKCM samples: 470 Segments per sample on avg.: 83.15957446808511\n",
      "STAD samples: 904 CNA events per sample on avg.: 130.9612831858407\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 904 tumors: 442 normals: 462\n",
      "\ttumors without matched normal 26\n",
      "\ttumors with at least one sCNA 410\n",
      "\ttumors without any somatic CNA 6\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "STAD samples: 436 Segments per sample on avg.: 96.62614678899082\n",
      "STES samples: 1277 CNA events per sample on avg.: 140.32263116679718\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1277 tumors: 627 normals: 650\n",
      "\ttumors without matched normal 29\n",
      "\ttumors with at least one sCNA 591\n",
      "\ttumors without any somatic CNA 7\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "STES samples: 620 Segments per sample on avg.: 110.11612903225806\n",
      "TGCT samples: 304 CNA events per sample on avg.: 83.8125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 304 tumors: 156 normals: 148\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 154\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "TGCT samples: 156 Segments per sample on avg.: 38.1025641025641\n",
      "THCA samples: 1013 CNA events per sample on avg.: 54.409674234945705\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1013 tumors: 506 normals: 507\n",
      "\ttumors without matched normal 15\n",
      "\ttumors with at least one sCNA 389\n",
      "\ttumors without any somatic CNA 102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "THCA samples: 404 Segments per sample on avg.: 3.873762376237624\n",
      "THYM samples: 248 CNA events per sample on avg.: 62.78629032258065\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 248 tumors: 125 normals: 123\n",
      "\ttumors without matched normal 5\n",
      "\ttumors with at least one sCNA 105\n",
      "\ttumors without any somatic CNA 15\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "THYM samples: 110 Segments per sample on avg.: 8.89090909090909\n",
      "UCEC-FFPE samples: 4 CNA events per sample on avg.: 349.5\n",
      "after filtering\n",
      "UCEC-FFPE samples: 4 Segments per sample on avg.: 90.0\n",
      "UCS samples: 111 CNA events per sample on avg.: 173.85585585585585\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 4 tumors: 4 normals: 0\n",
      "\ttumors without matched normal 4\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 111 tumors: 56 normals: 55\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 54\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "UCS samples: 56 Segments per sample on avg.: 179.28571428571428\n",
      "UVM samples: 160 CNA events per sample on avg.: 81.08125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 160 tumors: 80 normals: 80\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "UVM samples: 80 Segments per sample on avg.: 38.5375\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 80\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    }
   ],
   "source": [
    "data_dir = \"TCGA/CNA/data__2016_01_28//\"\n",
    "\n",
    "dfs = {}\n",
    "dfs_normals = {}\n",
    "tumors_without_CNA = {}\n",
    "for f in os.listdir(data_dir):\n",
    "    cohort = f.split(\".\")[2].replace(\"org_\",\"\")\n",
    "    file_path = f+\"/\"+cohort+\".txt\"\n",
    "    df = pd.read_csv(data_dir+file_path, sep = \"\\t\")\n",
    "\n",
    "    df[\"Chromosome\"] = df[\"Chromosome\"].map(chr_dict)\n",
    "    print(cohort,\"samples:\",len(set(df[\"Sample\"].values)),\n",
    "          \"CNA events per sample on avg.:\",float(df.shape[0])/len(set(df[\"Sample\"].values)))\n",
    "\n",
    "    #### remove segments overlapping with segemnts in normals by 80% or more reciprocally ####\n",
    "    df[\"type\"] = df[\"Sample\"].apply(sample_type)\n",
    "    df_normals = df.loc[df[\"type\"]== \"Normal\",[\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"]]\n",
    "    df_tumors = df.loc[df[\"type\"]== \"Tumor\",[\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"]]\n",
    "    normal_samples = list(set(df_normals[\"Sample\"].values))\n",
    "    tumor_samples = list(set(df_tumors[\"Sample\"].values))\n",
    "    print(\"total samples:\", len(set(df[\"Sample\"].values)),\n",
    "          \"tumors:\",len(tumor_samples),\"normals:\",len(normal_samples),file= sys.stderr)\n",
    "\n",
    "    tumors_without_somatic_CNA = []\n",
    "    tumors_germline_removed = []\n",
    "    tumors_without_matching_normal = []\n",
    "    filtered_normals = []\n",
    "    for tumor_sample in tumor_samples:\n",
    "        #print(sample, find_matching_normal(sample,list(set(d[\"Sample\"]))))\n",
    "        tumor = df_tumors.loc[df_tumors [\"Sample\"]== tumor_sample,:]\n",
    "        matching_normals = find_matching_normal(tumor_sample,normal_samples)\n",
    "        if len(matching_normals) >0:\n",
    "            n_segs = tumor.shape[0]\n",
    "            for normal_sample in matching_normals:\n",
    "                normal  =  df_normals.loc[df_normals[\"Sample\"]== normal_sample,:]\n",
    "                # thresholds for +1 and -1 copy in 75% of normal cell;\n",
    "                # this is to retain segments appeared due to slight tumor contamination\n",
    "                normal =  filter_lowconf_segments(normal,0,0.46, -0.68 )\n",
    "                filtered_normals.append(normal)\n",
    "                tumor = remove_ovelapping_segments(tumor, normal,tumor_sample)\n",
    "            #if n_segs > tumor.shape[0]:\n",
    "            #    print(n_segs - tumor.shape[0],\"segments removed in sample\",tumor_sample,\n",
    "            #    tumor.shape[0],\"remained\",file= sys.stderr)\n",
    "            tumor = filter_lowconf_segments(tumor,num_marker_thr,pos_seg_mean_thr, neg_seg_mean_thr )\n",
    "            if tumor.shape[0] == 0:\n",
    "                   tumors_without_somatic_CNA.append(tumor_sample)\n",
    "            else:\n",
    "                tumors_germline_removed.append(tumor)\n",
    "        else:\n",
    "            tumor = filter_lowconf_segments(tumor,num_marker_thr,pos_seg_mean_thr, neg_seg_mean_thr )\n",
    "            if tumor.shape[0] == 0:\n",
    "                   tumors_without_somatic_CNA.append(tumor_sample)\n",
    "            else:\n",
    "                tumors_without_matching_normal.append(tumor)\n",
    "        \n",
    "\n",
    "    print(\"\\ttumors without matched normal\",len(tumors_without_matching_normal),file= sys.stderr)\n",
    "    print(\"\\ttumors with at least one sCNA\",len(tumors_germline_removed),file= sys.stderr)\n",
    "    print(\"\\ttumors without any somatic CNA\",len(tumors_without_somatic_CNA),file= sys.stderr)\n",
    "    #dfs[cohort] = df\n",
    "\n",
    "#     if(len(tumors_germline_removed) > 0) | (len(tumors_without_matching_normal) > 0):\n",
    "    filtered_tumors = pd.concat(tumors_germline_removed+tumors_without_matching_normal)\n",
    "\n",
    "    dfs[cohort] = filtered_tumors\n",
    "\n",
    "    if(len(filtered_normals) > 0):\n",
    "        filtered_normals = pd.concat(filtered_normals)\n",
    "\n",
    "    dfs_normals[cohort] = filtered_normals\n",
    "    tumors_without_CNA[cohort] = tumors_without_somatic_CNA\n",
    "    print(\"after filtering\")\n",
    "    print(cohort,\"samples:\",len(set(filtered_tumors[\"Sample\"].values)),\n",
    "          \"Segments per sample on avg.:\",float(filtered_tumors.shape[0])/len(set(filtered_tumors[\"Sample\"].values)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aggregating to gene-level\n",
    "Gene annotation must be:\n",
    "\n",
    "with Entrez gene IDs\n",
    "in hg19 coordinates\n",
    "with columns \"chrom\",\"start\",\"stop\",\"gene\" (this is foru-column bed format)\n",
    "wget ftp://ftp.ncbi.nih.gov/genomes/H_sapiens/ARCHIVE/BUILD.37.3/GFF/ref_GRCh37.p5_top_level.gff3.gz\n",
    "\n",
    "echo -e \"chrom\\tstart\\tstop\\tgene\\tname\" > ef_GRCh37.p5_top_level.gff3.bed; zcat ref_GRCh37.p5_top_level.gff3.gz | awk '$3==\"gene\"' | cut -f 1,4,5,9| sed -e 's/;/\\t/g'| cut -f 1-3,5,6 | grep GeneID | sed -re 's/(Dbxref=GeneID:[0-9]*),.*/\\1/' | sed -e 's/Name=//' -e 's/Dbxref=GeneID://' | awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$5\"\\t\"$4}' >> ref_GRCh37.p5_top_level.gff3.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "rename_chroms = {'NC_000001.10':1,\"NC_000002.11\":2,\"NC_000003.11\":3,\"NC_000004.11\":4,\n",
    "                 \"NC_000005.9\":5,\"NC_000006.11\":6,\"NC_000007.13\":7,\"NC_000008.10\":8,\n",
    "                 \"NC_000009.11\":9,\"NC_000010.10\":10,\"NC_000011.9\":11,\"NC_000012.11\":12,\"NC_000013.10\":13,\n",
    "                 \"NC_000014.8\":14,\"NC_000015.9\":15,\"NC_000016.9\":16,\"NC_000017.10\":17,\n",
    "                 \"NC_000018.9\":18,\"NC_000019.9\":19,\"NC_000020.10\":20,\"NC_000021.8\":21,\n",
    "                 \"NC_000022.10\":22,\"NC_000023.10\":23,\"NC_000024.9\":24}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36019, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10954</td>\n",
       "      <td>11507</td>\n",
       "      <td>100506145</td>\n",
       "      <td>LOC100506145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12190</td>\n",
       "      <td>13639</td>\n",
       "      <td>100652771</td>\n",
       "      <td>LOC100652771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>653635</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>100302278</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom  start   stop       gene          name\n",
       "0      1  10954  11507  100506145  LOC100506145\n",
       "1      1  12190  13639  100652771  LOC100652771\n",
       "2      1  14362  29370     653635        WASH7P\n",
       "3      1  30366  30503  100302278     MIR1302-2"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_intervals = pd.read_csv('ref.bed', sep=\"\\t\")\n",
    "gene_intervals = gene_intervals.loc[gene_intervals[\"chrom\"].isin(rename_chroms.keys()),:]\n",
    "gene_intervals[\"chrom\"] = gene_intervals[\"chrom\"].apply(lambda x : rename_chroms[x])\n",
    "gene_intervals['gene'] = gene_intervals['gene'].map(lambda x: x.lstrip('Dbxref=GeneId:'))\n",
    "#print(\"chromosomes:\",list(set(gene_intervals[\"chrom\"].values)))\n",
    "gene_intervals = gene_intervals.sort_values(by=[\"chrom\",\"start\",\"stop\"],ascending=True)\n",
    "gene_intervals.to_csv(\"ref_GRCh37.p5_top_level.gff3.chroms_renamed.bed\",sep = \"\\t\",index=False)\n",
    "print(gene_intervals.shape)\n",
    "gene_intervals.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_intervals_bed = pbt.BedTool.from_dataframe(gene_intervals[[\"chrom\",\"start\",\"stop\",\"gene\"]])\n",
    "# prepare copy-neutral table\n",
    "cnv_baseline = gene_intervals.copy()\n",
    "cnv_baseline[\"Segment_Mean\"] = [0]*cnv_baseline.shape[0]\n",
    "cnv_baseline = cnv_baseline[[\"gene\",\"Segment_Mean\"]]\n",
    "cnv_baseline.set_index(\"gene\",inplace=True,drop=True)\n",
    "cnv_baseline.sort_index(inplace=True)\n",
    "sorted_index = list(cnv_baseline.index.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ACC\n",
      "TCGA-OR-A5KQ-01A-11D-A309-01 has no genes with altered CN\n",
      "TCGA-OR-A5KV-01A-11D-A29H-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACC (31922, 90)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BLCA-FFPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLCA-FFPE (7775, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BRCA-FFPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA-FFPE (15440, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CESC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n",
      "CESC (31922, 297)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CHOL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHOL (31922, 36)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "COAD-FFPE\n",
      "DLBC\n",
      "TCGA-G8-6914-14A-01D-2209-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DLBC (31874, 52)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ESCA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "ESCA (31922, 185)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GBM\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-06-0165-01A-01D-0236-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n",
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-06-5410-01A-01D-1694-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-06-0119-01A-08D-0214-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 500 processed.\n",
      "GBM (31922, 590)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "HNSC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n",
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-D6-6824-01A-11D-1910-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n",
      "... 500 processed.\n",
      "HNSC (31922, 530)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KICH\n",
      "TCGA-KN-8437-01A-11D-2308-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KICH (31922, 66)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KIRC-FFPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIRC-FFPE (2460, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KIRP\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-DW-7838-01A-11D-2135-01 has no genes with altered CN\n",
      "TCGA-A4-7828-01A-11D-2135-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-GL-A4EM-01A-11D-A253-01 has no genes with altered CN\n",
      "TCGA-4A-A93Y-01A-11D-A36W-01 has no genes with altered CN\n",
      "TCGA-Y8-A8S1-01A-11D-A36W-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIRP (31922, 288)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LAML\n",
      "TCGA-AB-2978-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-3006-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2854-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2947-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2808-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2840-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2845-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2932-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2884-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2969-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2837-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2931-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2991-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2880-03A-01D-0756-21 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-AB-2826-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2922-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2842-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2836-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2824-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2851-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2871-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2844-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2998-03A-01D-0756-21 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LAML (30616, 191)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LGG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-S9-A6WI-01A-21D-A33S-01 has no genes with altered CN\n",
      "TCGA-HT-7680-01A-11D-2252-01 has no genes with altered CN\n",
      "TCGA-P5-A5EY-01A-11D-A27J-01 has no genes with altered CN\n",
      "TCGA-HT-8106-01A-11D-2391-01 has no genes with altered CN\n",
      "TCGA-DU-5872-02A-21D-A36N-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-FG-8181-01A-11D-2252-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-HT-7602-01A-21D-2085-01 has no genes with altered CN\n",
      "TCGA-DU-7011-01A-11D-2023-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-TM-A84B-12A-01D-A366-01 has no genes with altered CN\n",
      "TCGA-CS-6669-01A-11D-1892-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 500 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-FG-8189-01B-11D-A288-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LGG (31922, 530)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LIHC\n",
      "TCGA-G3-A25V-01A-11D-A16U-01 has no genes with altered CN\n",
      "TCGA-BC-A10X-01A-11D-A12Y-01 has no genes with altered CN\n",
      "TCGA-ED-A5KG-01A-11D-A27H-01 has no genes with altered CN\n",
      "TCGA-2V-A95S-10D-01D-A36Z-01 has no genes with altered CN\n",
      "TCGA-DD-A3A6-01A-11D-A22E-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-CC-A9FV-01A-11D-A36W-01 has no genes with altered CN\n",
      "TCGA-MR-A520-01A-11D-A25U-01 has no genes with altered CN\n",
      "TCGA-UB-AA0V-01A-11D-A381-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-DD-A4NL-01A-11D-A28W-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LIHC (31922, 373)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LUAD-FFPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LUAD-FFPE (12203, 12)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LUSC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-56-8623-01A-11D-2391-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n",
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-98-A53H-01A-12D-A25M-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n",
      "LUSC (31922, 501)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MESO\n",
      "TCGA-TS-A8AS-01A-21D-A39Q-01 has no genes with altered CN\n",
      "TCGA-3H-AB3O-01A-11D-A39Q-01 has no genes with altered CN\n",
      "TCGA-TS-A7P8-01A-11D-A34B-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MESO (31922, 87)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OV\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n",
      "... 300 processed.\n",
      "... 400 processed.\n",
      "... 500 processed.\n",
      "OV (31922, 597)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PAAD\n",
      "TCGA-IB-AAUR-01A-21D-A38F-01 has no genes with altered CN\n",
      "TCGA-HZ-8002-01A-11D-2200-01 has no genes with altered CN\n",
      "TCGA-XD-AAUG-01A-61D-A40V-01 has no genes with altered CN\n",
      "TCGA-Z5-AAPL-01A-12D-A40V-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-IB-A5SQ-01A-11D-A32M-01 has no genes with altered CN\n",
      "TCGA-IB-AAUS-01A-12D-A38F-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAAD (31922, 185)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PCPG\n",
      "TCGA-SA-A6C2-01A-11D-A35H-01 has no genes with altered CN\n",
      "TCGA-RW-A7CZ-01A-11D-A35C-01 has no genes with altered CN\n",
      "TCGA-WB-A817-01A-11D-A35H-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "PCPG (31922, 168)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PRAD-FFPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRAD-FFPE (2249, 6)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "READ\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "READ (31922, 166)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SARC\n",
      "TCGA-QQ-A5V2-01A-11D-A32H-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-WK-A8XS-10E-01D-A37E-01 has no genes with altered CN\n",
      "TCGA-WK-A8Y0-10D-01D-A419-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n",
      "SARC (31922, 263)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SKCM\n",
      "TCGA-EB-A4OZ-01A-12D-A25P-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n",
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-ER-A19A-06A-21D-A191-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n",
      "SKCM (31922, 472)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "STAD\n",
      "TCGA-RD-A8NB-01A-12D-A396-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-D7-A6ET-01A-32D-A32M-01 has no genes with altered CN\n",
      "TCGA-BR-7196-01A-11D-2052-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-D7-6522-01A-11D-1799-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-HU-A4GJ-01A-11D-A253-01 has no genes with altered CN\n",
      "TCGA-BR-7957-01A-11D-2200-01 has no genes with altered CN\n",
      "TCGA-BR-6563-01A-13D-2052-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-MX-A5UG-01A-21D-A31K-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STAD (31922, 442)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "STES\n",
      "TCGA-RD-A8NB-01A-12D-A396-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-D7-A6ET-01A-32D-A32M-01 has no genes with altered CN\n",
      "TCGA-BR-7196-01A-11D-2052-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n",
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-D7-6522-01A-11D-1799-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-HU-A4GJ-01A-11D-A253-01 has no genes with altered CN\n",
      "TCGA-BR-7957-01A-11D-2200-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 500 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-BR-6563-01A-13D-2052-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 600 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-MX-A5UG-01A-21D-A31K-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STES (31922, 627)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TGCT\n",
      "TCGA-YU-A90S-01A-11D-A434-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "TGCT (31922, 156)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "THCA\n",
      "TCGA-EM-A1CS-01A-11D-A13V-01 has no genes with altered CN\n",
      "TCGA-EL-A3ZT-01A-12D-A23L-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UY-01A-21D-A22C-01 has no genes with altered CN\n",
      "TCGA-EM-A1CU-01A-11D-A13V-01 has no genes with altered CN\n",
      "TCGA-EL-A3CL-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UX-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-EL-A4K4-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DE-A69J-01A-21D-A396-01 has no genes with altered CN\n",
      "TCGA-EM-A4FO-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EL-A4KD-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-E8-A414-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-BJ-A2P4-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-DJ-A13X-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-J8-A3YH-01A-11D-A22Y-01 has no genes with altered CN\n",
      "TCGA-ET-A3DP-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-EL-A3TA-01A-12D-A22C-01 has no genes with altered CN\n",
      "TCGA-ET-A3DV-01A-12D-A201-01 has no genes with altered CN\n",
      "TCGA-ET-A3DW-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EM-A3FK-01A-11D-A219-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-DE-A4MA-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-FY-A3RA-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-BJ-A0Z9-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V0-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EM-A3FO-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-EL-A4K2-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EL-A4K1-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-FY-A3I4-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-BJ-A0Z5-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-ET-A39T-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-DJ-A2Q3-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-GE-A2C6-01A-11D-A16M-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UT-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-CE-A3ME-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-E3-A3E5-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-E3-A3E3-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-CE-A485-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-BJ-A18Y-01A-11D-A13V-01 has no genes with altered CN\n",
      "TCGA-FY-A4B3-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-EM-A3O7-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-FY-A3R8-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-EL-A4JZ-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EM-A2CN-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-ET-A39O-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-E8-A418-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-L6-A4EQ-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V5-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A2PO-01A-21D-A19I-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-EL-A3N2-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-BJ-A0ZG-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-E8-A437-01A-12D-A23T-01 has no genes with altered CN\n",
      "TCGA-DE-A0Y2-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-J8-A3O2-06A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-DJ-A2PP-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-E3-A3E1-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-ET-A39J-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-DE-A4M8-01A-21D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A2PS-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-EL-A3D4-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EL-A3TB-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-EL-A3GW-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-BJ-A0ZJ-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-EL-A3H8-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-KS-A4I5-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A3VE-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-CE-A483-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-EL-A3CX-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UO-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-E8-A433-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-ET-A39L-01A-12D-A19I-01 has no genes with altered CN\n",
      "TCGA-FY-A4B4-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-ET-A3BX-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-BJ-A45D-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-EL-A3GX-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-EM-A4FM-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EM-A1YD-01A-11D-A14V-01 has no genes with altered CN\n",
      "TCGA-MK-A4N7-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V2-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-FY-A76V-01A-11D-A396-01 has no genes with altered CN\n",
      "TCGA-EM-A2CU-01A-12D-A17S-01 has no genes with altered CN\n",
      "TCGA-KS-A4I7-01A-11D-A256-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-EM-A3O8-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V4-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EL-A4K7-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-ET-A39M-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-E8-A415-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-ET-A3BU-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-J8-A3O2-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-ET-A40Q-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-DJ-A1QM-01A-11D-A14V-01 has no genes with altered CN\n",
      "TCGA-EL-A3T6-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-DE-A0XZ-01A-11D-A17S-01 has no genes with altered CN\n",
      "TCGA-IM-A41Z-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-EL-A3D1-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EM-A3FM-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-ET-A2MZ-01A-12D-A19I-01 has no genes with altered CN\n",
      "TCGA-DJ-A3VK-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-DJ-A4UT-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A2Q2-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-E3-A3DZ-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-EM-A22K-01A-11D-A17S-01 has no genes with altered CN\n",
      "TCGA-EM-A3AP-01A-12D-A20A-01 has no genes with altered CN\n",
      "TCGA-EL-A3D0-01A-12D-A201-01 has no genes with altered CN\n",
      "TCGA-KS-A4I9-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-BJ-A28S-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EM-A1CW-01A-21D-A13V-01 has no genes with altered CN\n",
      "TCGA-ET-A2N1-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-EM-A2P3-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UN-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-EM-A3AL-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-FE-A3PC-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-EM-A4FQ-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A3VJ-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-EL-A3MY-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-ET-A40T-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-ET-A25N-01A-11D-A16M-01 has no genes with altered CN\n",
      "TCGA-DE-A4M9-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-FY-A3TY-01A-11D-A22Y-01 has no genes with altered CN\n",
      "TCGA-BJ-A18Z-01A-21D-A13V-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n",
      "THCA (31922, 506)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "THYM\n",
      "TCGA-XU-AAXZ-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZB-A96G-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-XM-A8RB-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-XU-A92X-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96E-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-3Q-A9WF-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-4X-A9F9-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZC-AAAA-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZB-A96M-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-X7-A8M8-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-X7-A8M4-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96A-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-X7-A8DC-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96Q-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-3S-AAYX-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-YT-A95E-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-X7-A8M1-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-X7-A8M7-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-X7-A8D8-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-4V-A9QQ-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-XH-A853-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZT-A8OM-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZB-A96R-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-XM-AAZ3-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96B-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZC-AAAH-01A-11D-A427-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-X7-A8DB-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96O-01A-11D-A427-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THYM (31922, 125)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UCEC-FFPE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UCEC-FFPE (13704, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UCS\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UCS (31922, 56)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UVM\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UVM (31922, 80)\n"
     ]
    }
   ],
   "source": [
    "for cohort in dfs.keys():\n",
    "    print(cohort, file=sys.stderr)\n",
    "    df = dfs[cohort]\n",
    "    cna_table = []\n",
    "    n_samples = 0\n",
    "    if(cohort ==\"COAD-FFPE\"):\n",
    "        continue\n",
    "    for sample in list(set(df.Sample.values)):\n",
    "        n_samples +=1\n",
    "        cnv2gene = cnv2genelevel(cnv2bed(df[df.Sample == sample]),gene_intervals_bed,sample,\n",
    "                                 verbose = False,sorted_index = sorted_index)\n",
    "        cna_table.append(cnv2gene)\n",
    "        if n_samples % 100 == 0:\n",
    "            print(\"...\",n_samples, \"processed.\")\n",
    "    cna_table = pd.concat(cna_table,axis =1)\n",
    "    \n",
    "\n",
    "    for sample in tumors_without_CNA[cohort]:\n",
    "        cna_table[sample] = 0\n",
    "    \n",
    "    cna_table.fillna(0, inplace = True)\n",
    "    cna_table.to_csv(preprocessed_dir+\"/TCGA-\"+cohort+\".Segment_Mean.CNA.tsv\",\n",
    "                     sep = \"\\t\",header=True,index=True)\n",
    "    print(cohort,cna_table.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'t = time.time()\\ncnv2gene = cnv2genelevel(cnv2bed(df[df.Sample == sample]),gene_intervals_bed,sample,\\n                         verbose = False,sorted_index = sorted_index)\\nprint( time.time() - t)\\ncnv2gene'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"t = time.time()\n",
    "cnv2gene = cnv2genelevel(cnv2bed(df[df.Sample == sample]),gene_intervals_bed,sample,\n",
    "                         verbose = False,sorted_index = sorted_index)\n",
    "print( time.time() - t)\n",
    "cnv2gene\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CCLE\n",
    "the same pipeline as for TCGA except filtering out germline CNA (because no )\n",
    "\n",
    "wget https://data.broadinstitute.org/ccle_legacy_data/dna_copy_number/CCLE_copynumber_2013-12-03.seg.txt\n",
    "\n",
    "? should we use a stronger segment_mean threshold because this data are for cell lines and purity must be 100%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No duplicated row IDs. Do nothing.\n",
      "CCLE: genes: 33850 samples 1043\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"CCLE/CCLE_copynumber_2013-12-03.seg.txt\",sep = \"\\t\")\n",
    "df.rename({\"CCLE_name\":\"Sample\"},inplace=True, axis=\"columns\")\n",
    "df[\"End\"] = df[\"End\"].apply(int)\n",
    "ccle = []\n",
    "for sample_name in list(set(df[\"Sample\"].values)):\n",
    "    cl = df.loc[df[\"Sample\"]==sample_name, :]\n",
    "    # keep high-conf segments \n",
    "    cl_filtered = filter_lowconf_segments(cl,num_marker_thr,pos_seg_mean_thr, neg_seg_mean_thr )\n",
    "    #print(sample_name, cl.shape[0], \"after filtration\",cl_filtered.shape[0])\n",
    "    # map to genes \n",
    "    cnv2gene = cnv2genelevel(cnv2bed(cl_filtered),gene_intervals_bed,sample_name,\n",
    "                                 verbose = False,sorted_index = sorted_index)\n",
    "    ccle.append(cnv2gene)\n",
    "    \n",
    "ccle = pd.concat(ccle,axis =1)\n",
    "ccle.fillna(0, inplace = True)\n",
    "ccle = handle_dups(ccle)\n",
    "ccle.to_csv(preprocessed_dir+\"/\"+\"CCLE\"+\".Segment_Mean.CNA.tsv\",\n",
    "                 sep = \"\\t\",header=True,index=True)\n",
    "print(\"CCLE:\",\"genes:\",ccle.shape[0],\"samples\",ccle.shape[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GDSC\n",
    "Assume that supplementary file with gene-level CN is downloaded :\n",
    "\n",
    "wget\n",
    "\n",
    "GDSC provides gene-level integer estimated CN, max. and min. CN over all segments covering a gene. In order to make it comparable with TCGA and CCLE, we divide estimated CN by CN of copy-neutral state and log2-transform it.\n",
    "\n",
    "1) Copy-neutral state was defined from average ploiy, as median of integer CN values in non-disrupted genes.\n",
    "\n",
    "2) Compute log2(CN/neutral-CN) for min and max CN; keep the value with most extreme estimate\n",
    "\n",
    "3) Replace estimates below thresholds with zeroes.\n",
    "\n",
    "DGSC uses 4 comma-separated values for gene-level CN (max_cn,min_cn,zygosity,disruption): e.g. (from \"legend\" tab)\n",
    "\n",
    "2,2,H,- Gene resides on a single genomic segment in a diploid region of the genome. 2,0,L,D Gene spans multiple segments, higest copy number is 2 but part of the coding sequence is homozygously deleted, the gene is disrupted. 13,13,H,- Gene resides on a single genomic segment of copy number 13 in a heterozygous part of the genome (amplification). 14,12,L,D Gene spans multiple genomic segments all of which are amplified to 12 or more copies, some or all segments have LOH, the gene is disrupted. 0,0,0,- Complete gene sequence falls within a homozygous deletion. -1,-1,-,- gene level CN not assigned\n",
    "\n",
    "min and max CN are integers\n",
    "zygosity - can be L (LOH in any overlapping segment) or H (heterozygous) or 0 (homozygous deleteion of the whole gene) or - (undefined)\n",
    "disruption - D (if disrupted) or \"-\" (not disrupted)\n",
    "Average ploidies of cell lines were downloaded from COSMIC:\n",
    "\n",
    "wget https://cog.sanger.ac.uk/cosmic/GRCh37/cell_lines/v86/PICNIC_average_ploidies.tsv?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1540792525&Signature=mcSB6oFv%2BXCF4%2Fezm4a3Ds1JXo4%3D\n",
    "\n",
    "wget ftp:// ftp.sanger.ac.uk/pub/project/cancerrxgene/releases/release-7.0/Gene_level_CN.xlsx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>201T</th>\n",
       "      <th>22RV1</th>\n",
       "      <th>23132-87</th>\n",
       "      <th>42-MG-BA</th>\n",
       "      <th>451Lu</th>\n",
       "      <th>5637</th>\n",
       "      <th>...</th>\n",
       "      <th>WSU-NHL</th>\n",
       "      <th>YAPC</th>\n",
       "      <th>YH-13</th>\n",
       "      <th>YKG-1</th>\n",
       "      <th>YMB-1-E</th>\n",
       "      <th>YT</th>\n",
       "      <th>ZR-75-30</th>\n",
       "      <th>huH-1</th>\n",
       "      <th>no-10</th>\n",
       "      <th>no-11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1287381</td>\n",
       "      <td>924100</td>\n",
       "      <td>910924</td>\n",
       "      <td>687561</td>\n",
       "      <td>1287706</td>\n",
       "      <td>687452</td>\n",
       "      <td>...</td>\n",
       "      <td>909785</td>\n",
       "      <td>909904</td>\n",
       "      <td>909905</td>\n",
       "      <td>687592</td>\n",
       "      <td>1303911</td>\n",
       "      <td>946358</td>\n",
       "      <td>909907</td>\n",
       "      <td>1298146</td>\n",
       "      <td>908452</td>\n",
       "      <td>908450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>...</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14363.0</td>\n",
       "      <td>29806.0</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>...</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  chr    start     stop       201T      22RV1   23132-87   42-MG-BA  \\\n",
       "0      NaN  NaN      NaN      NaN    1287381     924100     910924     687561   \n",
       "1  DDX11L1    1  11869.0  14412.0  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "2   WASH7P    1  14363.0  29806.0  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "\n",
       "       451Lu       5637  ...    WSU-NHL       YAPC      YH-13      YKG-1  \\\n",
       "0    1287706     687452  ...     909785     909904     909905     687592   \n",
       "1  -1,-1,-,-  -1,-1,-,-  ...  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "2  -1,-1,-,-  -1,-1,-,-  ...  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "\n",
       "     YMB-1-E         YT   ZR-75-30      huH-1      no-10      no-11  \n",
       "0    1303911     946358     909907    1298146     908452     908450  \n",
       "1  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  \n",
       "2  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  \n",
       "\n",
       "[3 rows x 1000 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GDSC_CNA = \"GDSC/Gene_level_CN.xlsx\"\n",
    "\n",
    "gdsc = pd.read_excel(GDSC_CNA,\"Gene_level_CN\")\n",
    "gdsc.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25 gene IDs excluded due to string to datetime conversion in Excel.\n",
      "Strings containing duplicated gene IDs: 0\n"
     ]
    }
   ],
   "source": [
    "gdsc.set_index(\"gene\",inplace = True)\n",
    "gdsc.drop([\"chr\",\"start\",\"stop\"],inplace=True,axis=1)\n",
    "gdsc.columns = gdsc.iloc[0,:]\n",
    "gdsc = gdsc.iloc[1:,:]\n",
    "gdsc.columns.name = None\n",
    "# replace 2001-12-01 with DEC1 and get remove gene names converted to datetimes\n",
    "gdsc.index.values[37778] = \"DEC1\"\n",
    "df_size = gdsc.shape[0]\n",
    "ndxs=pd.Series(gdsc.index).apply(lambda x : type(x) == str)\n",
    "gdsc = gdsc.loc[gdsc.index.values[ndxs[ndxs].index],:]\n",
    "print(df_size - gdsc.shape[0],\"gene IDs excluded due to string to datetime conversion in Excel.\")\n",
    "\n",
    "gdsc.index.name = \"gene_id\"\n",
    "ids = gdsc.index\n",
    "ids = list(set(ids[ids.duplicated()]))\n",
    "print(\"Strings containing duplicated gene IDs:\",gdsc.loc[ids,:].shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## distribution of averaged ploidies in GDSC\n",
    "we compared average ploidies reported in PICNIC_average_ploidies.tsv provided by COSMIC with"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1020\n",
      "1016\n"
     ]
    }
   ],
   "source": [
    "GDSC_Ploidies = \"GDSC/PICNIC_average_ploidies.tsv\"\n",
    "GDSC_Ploidies = pd.read_csv(GDSC_Ploidies,sep = \"\\t\")\n",
    "GDSC_Ploidies.drop(\"#sample_name\",axis = 1, inplace= True)\n",
    "GDSC_Ploidies.set_index(\"sample_id\",inplace=True)\n",
    "print(GDSC_Ploidies.shape[0])\n",
    "GDSC_Ploidies.dropna(inplace=True)\n",
    "print(GDSC_Ploidies.shape[0])\n",
    "\n",
    "est_ploidies = gdsc.apply(define_avg_ploidy).T\n",
    "df_ploidies = pd.DataFrame.from_dict({\"est. avg. ploidy from CN profile\":est_ploidies[\"avg_pl\"],\"PICNIC avg. pl.\":GDSC_Ploidies[\"average_ploidy\"],\n",
    "                                     \"est. median. ploidy\":est_ploidies[\"median_pl\"]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIQAAAE/CAYAAAA35xgnAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3de5hlVX3n//dHUGO8odAhyMX2giZqklY7qPESIhoRUNQoQgyiIWlNMDFj5mfQTMRLnMFJDJqY6KAgkAiCIIqKGYnXmBFNgwS56AjYDI0ttCAXxZAA398fe5Wcrq7qOt1V59K136/nqaf2WXudc761u89a+3z3WmunqpAkSZIkSVJ/3GPSAUiSJEmSJGm8TAhJkiRJkiT1jAkhSZIkSZKknjEhJEmSJEmS1DMmhCRJkiRJknrGhJAkSZIkSVLPmBDSVEuyb5L1Q9Z9X5I/28L+SvLIbYjhPkk+keTmJB/Z2uf3QZKnJvl2kh8meUGSTyc5ou17RZIvTzpGSVruknwhye9MOg5J6ovZ57ntXPjhE4hjqO85SV6W5DNb2G8/0jMmhJapJCtbw7DjpGMZl6p6dVW9bQQv/WJgV2DnqnrJCF5/QUmek+RLSW5NsjHJF5M8v+17Rfu3fv2s56xPsu+YQnwr8J6qul9VfayqnltVJ4/pvSUtA0nWJflxO5m+LslJSe7X9m1ygprkAUneleT/tfpXtse7DLzW9UnuO/Cc30nyhYHHm5w8J3lUko8k+X67AHBxktcl2WEsB0CStKBp/47TzoWvmnQc86mqD1XVr086Dk0PE0LSwh4K/N+qumOunaPukJK8GPgIcAqwB11y6k3A8waq3Qi8Psn9R/D+w/x9DwUuXer3ltQ7z6uq+wFPAFYD/212hST3Aj4LPBbYH3gA8BTgBmCfgao7AK8d5k2TPAL4KnAN8AtV9UDgJS2GJW9XJUmSpoEJoe1AkockOauNDPlOkj8c2LdPkrVJbmlXVP+q7fpS+31Tu3r6lCHe55VJLm+jUK5K8qqBfZcnOWjg8Y4tnie0xy9PcnWSG5L8Wbs6+6wh/751Sd6Q5LIkP0jywSQ/NU/dn29Xim9KcunMKJm276Qkfz7w+P9LsiHJd5P89kD5L7djtcNA2YuS/Nsc7/cWuuTLS9txPLKNyPmXJMcluQF4c5IHJjmlHZOrk/y3JPdorzFY/6Z2bH+llV/TrmIfMc/fG+CvgLdV1Qeq6uaququqvlhVvztQ9XLgK8DrhjzmJ6WbYnde+/f+YpKHDuyvJEcl+Tbw7Vb2u0muSHJjknOSPKSVXwk8HPhEO0b3nn01f9Z7/1x73xuTfCvJIcPELKk/qupa4NPA4+bY/XJgL+CFVXVZaxOvr6q3VdW5A/X+AvivSXYa4i3fAvyfqnpdVW1oMXyrqn6zqm6aXTnJg5J8srX5P2jbe7R9L02ydlb9/5LknLa9c7ppyLck+dckf54hp9UO9CfvSTeK6ZtJ9hvmuZI0bcb4HecLra39P+05n2ht8YcG2uKVA/XnPVdtzzunPe9rwCNmvddPRp8mOTDJ11vda5K8eaDezEinI9KNdv1+kj/dwt+wxXP3WXUX+l4yOMXt2a0vuTnJe4C08nu1v/8XBur+TJLbkqxY6Jhr+2FCaMq1D+8ngH8Ddgf2A/4oyXNalXcD766qB9A1SGe08me03zu1oYtfGeLtrgcOorva+krguLSED3AacNhA3ecA36+qC5M8Bvg74GXAbsADW6xb42XtNR8BPIq5rwrfk+5YfAb4GeAPgA8lefQcdfcH/ivwbGBv4CfJqar6V7oryYPDJQ+nG4Gziao6BvjvwOntOJ7Qdj0JuIputM7bgb+h+7sfDvwq3ReWVw681JOAi4GdgVOBDwO/DDwS+C3gPWlTI2Z5NLAncOYc+2b7M7r/Gw8eoi50x/xtwC7ARcCHZu1/QYv7MUmeCfwP4BC6f+Or299AVT0C+H+0K/tVdft8b5hu+sZ5dMfgZ4BDgb9r/4ckCYAkewIHAF+fY/ezgH+sqh8u8DJrgS/Q9QULeRbDtbMz7gF8kG505F7Aj4H3tH2fAB6dZO+B+r9J1+4B/C3wI+BngSPaz9Z4EnAlXdt9DPDRrWj3JWkqjPk7DnTnnIe393oE3YXUDwIPpruwekyLa6Fz1b8F/p3ufPi32898fkT3nWAn4EDg95K8YFadp9Gd7+8HvCnJz2/h9RY6d5+x0PcS2t+6C/BRuu9du9D1LU8FqKr/oDvX/62BpxwGfLaqNm4hRm1nTAhNv18GVlTVW6vqP9qc1PfTNU4A/wk8MskuVfXDqjp/W9+oqj5VVVdW54t0iZent92nAs9P8tPt8W/SJYmgW2PnE1X15dZ4vAmorXz791TVNVV1I12C5bA56jwZuB9wbDsWnwM+OU/dQ4APVtUlVfUj4M2z9p9Ma+DaifRzuPtkfRjfraq/adPI/oPu3+MNVXVrVa0D3knX6cz4TlV9sKruBE6nS/K8tapur6rPtNeYayG4ndvvDQsFVFUX0XVgfzLk3/CpqvpSS+D8KfCU9iVsxv+oqhur6sd0HdCJVXVhq/+GVn/lkO814yBgXTsWd1TV14Gz6KZmSNLHktwEfBn4Il1CfradGaJNbN4E/MEQVzO35jWpqhuq6qyquq2qbqXrt3617bsN+Ditb2qJoZ8Dzkk3MvU3gGPacy+j64+2xvXAu6rqP6vqdOBbdF80JGl7MrbvOM0H2/ecm+lGoF5ZVf/UzuU/Ajy+1Zv3XHWgDX9TVf2oqi5hC214VX2hqr7RRrJeTPfd6VdnVXtLVf24qv6NLjn2S1v4GxY6d6fFuND3khkHAJdW1ZlV9Z/Au4DvDew/GTgsSdrjw4G/30J82g6ZEJp+DwUekm6q0U3tRPmNdCNTAI6kG1HzzTbc8aD5XmghSZ6b5Pw2PPAmukZiF4CquoIue/68lhR6PncnUB5Ct+4Cre5tdCNwtsY1A9tXt9ec7SHANVV116y6c41G2iSmVm/QP9D9LfelSx7988w0gW2IdxfgnrPeY3Zc1w1s/xigqmaXzTVCaOY47jZkXG+iu/qw64I1N/03+yHdOkQPmWt/K796Vv0b2PqRYA8FnjTr//PL6K6US9ILqmqnqnpoVf1+S0jPdgNDtontZP2TwNELVB36NQGS/HSS/9WG4t9CN4Vhp9w9FflU7r5Y8ZvAx1rfuALYkU3b18HtYVxbVYMXXebrMyVpmo3tO04z+7x7vvPwLZ2rztWGz/6O8RNJnpTk823q1s3Aq2nfrQYMJmBuY+7vAzMWOneH4b6XzJj9Ha5mPf5qi2nfJD9Hd/H6nC3Ep+2QCaHpdw3d6JKdBn7uX1UHAFTVt6vqMLohje8AzmxJjq0aoZPk3nTZ778Edq2qnYBzafNIm5lpYwcDl7UkEXRXVfcYeK37cPfIlmENZrf3Ar47R53vAnvOzIEdqHvtHHU3zPGaP1Hd+hRfAV7EtmW7B4/v9+muYgzO450vrq31Lbr/A78xVFBV36Qb+jnvHOQBPzk+bbrag9n0uA/+jd9l4O9r/8d2Zuv/xmuAL876/3y/qvq9rXwdSf31T8BzMnAHsQUcA/wuW05g/xNDtrPNH9MN8X9Sm84wM4Vhps88D1iRZBVdvzlzAWUjcAcDfSab9lXD2H3gai3M32dK0jQby3ecbYxrvnPVmTZ83u8Ys5xKl0DZs7qbFbyPTb9bba2Fzt1h676XbPJ9qfUts/ukmVkVhwNnVtW/b2vwmk4mhKbf14Bbk/xJkvsk2SHJ45L8MkCS30qyoo2amVn48i66Busuurmjw7gXcO/2vDuSPJdN19iBbh7prwO/x6bTq86kG23zK+nu/vJmtr6xOyrJHm361p/STauabSZL/fok90x3S/XntbhmOwN4RZLHtBFNx8xR5xTg9cAv0CVRtkmbBnYG8PYk928LvL2ObhTSorRM/euAP0u36PcDktwjydOSHD/P095CN094oYVUD2ivcy+6+cjnV9V8V6pPA16ZZFVLHv534KttGOrW+CTwqCSHt3/De6Zb5HtL86UladDf052wn5Vu4c97pFvk841JDphduV28OB34w9n7BhwD/EqSv0jyswBJHpnkHzL3otT3p7uifFPrtzbpY9rQ+4/QLWz9YLoE0Ux/8VG6mxH8dLvi+vKt+uu7L0d/2NrPlwA/T3cBR5K2J+P6jrO15j1XnaMNfwxbXgfu/sCNVfXvSfahGzG6GAueu2/l95JPAY9Nd3OdHen6ydmj9v8BeCFdUmiz9Va1/TMhNOXah/ogYBXwHbqs7wfoFgqD7pa7lyb5Id3ia4e2eai30a1p8C9tuOOTkzy91ZvrfW6lawTOAH5A12CdM6vOBrpRNb/CQMKmqi6lW+D5w3SZ5h/SrXFwO0CSlyVZ6Jbkp9KtWXQV3YJmfz67QnXrEz0PeG47Dn8HvLyNipld99N082A/B1zRfs92Nl32/Ox2vBbjD+gWjruKbu2LU4ETF/maAFTVmcBL6Rat+y7dENc/p1ujYq7636H7wrTQ1fNT6b7E3Ag8kU0XjZv9mv9Et2j1WXT/xo/g7jneQ2v/z369Pfe7dMNk30GXjJSkBbW1E54FfJMu0XIL3ReLXeguHMzlrWyhTayqK+luXb+Srk+9ma69WwvcOsdT3gXch64vOh/4xznqnNri/Ehbo2LGa+j68O/RtdWn0fpLgHR30HzZfLHS/Y17t/d+O/DiqtpsmvaW+nxJmrRxfcfZhrgWOld9Dd20ru8BJ9EtTD2f3wfemuRWumUdzthC3WEMe+4+1PeSqvo+3Tqex9JNnd4b+JdZda4BLqQbmfXPi4xfUyibTkOXFq8NYbwJ2LslJxaqvw74nZZ0GKt0t0x/1STee5KSnASsr6rN7uYmSRqfJO8AfraqFrzbWJJX0PWXTxt5YJKkqTHJc/ckJ9LdUMfvDcuQI4S0JJI8rw2dvC/dOkTfANZNNqotS/IbdNnuuUYPSZK05No0t19MZx+6hVPPnnRckiTNlu6Owi8CTphsJBoVE0JaKgfTDav8Lt1ww0NrioefJfkC8F7gqFl3LZMkaZTuT7cGxY/opl+/k3mmAEuSNClJ3gZcAvzFMLM+tH1yypgkSZIkSVLPOEJIkiRJkiSpZ0wISZIkSZIk9cyOkw4AYJdddqmVK1dOOgxJmkoXXHDB96tqxaTjmCT7CUmam31Ex35Ckua2pX5iKhJCK1euZO3atZMOQ5KmUpKrJx3DpNlPSNLc7CM69hOSNLct9RNOGZMkSZIkSeoZE0KSJEmSJEk9Y0JIkiRJkiSpZ0wISZIkSZIk9YwJIUmSJEmSpJ4xISRJkiRJktQzJoQkSZIkSZJ6xoSQJEmSJElSz5gQkiRJkiRJ6hkTQpIkSZIkST1jQkiSJEmSJKlndpx0ANNm5dGfGqreumMPHHEkkqRpZD8hScuL7bqkvnKEkCRpUZKcmOT6JJcMlJ2e5KL2sy7JRa18ZZIfD+x73+QilyRJkvrLEUKSpMU6CXgPcMpMQVW9dGY7yTuBmwfqX1lVq8YWnSRJkqTNmBCSJC1KVX0pycq59iUJcAjwzHHGJEmSJGnLnDImSRqlpwPXVdW3B8oeluTrSb6Y5OnzPTHJmiRrk6zduHHj6COVJEmSesSEkCRplA4DTht4vAHYq6oeD7wOODXJA+Z6YlUdX1Wrq2r1ihUrxhCqJEmS1B8mhCRJI5FkR+BFwOkzZVV1e1Xd0LYvAK4EHjWZCCVJkqT+MiEkSRqVZwHfrKr1MwVJViTZoW0/HNgbuGpC8UmSJEm9ZUJIkrQoSU4DvgI8Osn6JEe2XYey6XQxgGcAF7fb0J8JvLqqbhxftJIkSZLAu4xJkhapqg6bp/wVc5SdBZw16pgkSZIkbdmCI4SSnJjk+iSXDJSdnuSi9rOuXeklycokPx7Y975RBi9JkiRJkqStN8wIoZOA9wCnzBRU1UtntpO8E7h5oP6VVbVqqQKUJEmSJEnS0lowIVRVX0qycq59SQIcAjxzacOSJEmSJEnSqCx2UemnA9dV1bcHyh6W5OtJvpjk6Yt8fUmSJEmSJC2xxS4qfRib3kFmA7BXVd2Q5InAx5I8tqpumf3EJGuANQB77bXXIsOQJEmSJEnSsLZ5hFCSHYEXAafPlFXV7VV1Q9u+ALgSeNRcz6+q46tqdVWtXrFixbaGIUmSJEmSpK20mCljzwK+WVXrZwqSrEiyQ9t+OLA3cNXiQpQkSZIkSdJSGua286cBXwEenWR9kiPbrkPZdLoYwDOAi9tt6M8EXl1VNy5lwJIkSZIkSVqcYe4ydtg85a+Yo+ws4KzFhyVJkiRpOWkzCdYC11bVQUkeBnwY2Bm4ADi8qv4jyb2BU4AnAjcAL62qdRMKW5KWrcXeZUySJEmShvFa4PKBx+8AjquqRwI/AGZmIhwJ/KCVH9fqSZKWmAkhSZIkSSOVZA/gQOAD7XGAZ9ItMwFwMvCCtn1we0zbv1+rL0laQiaEJEmSJI3au4DXA3e1xzsDN1XVHe3xemD3tr07cA1A239zqy9JWkImhCRJkiSNTJKDgOur6oIlft01SdYmWbtx48alfGlJ6gUTQpIkSZJG6anA85Oso1tE+pnAu4Gdkszc5GYP4Nq2fS2wJ0Db/0C6xaU3UVXHV9Xqqlq9YsWK0f4FkrQMmRCSJEmSNDJV9Yaq2qOqVgKHAp+rqpcBnwde3KodAXy8bZ/THtP2f66qaowhS1IvmBCSJEmSNAl/ArwuyRV0awSd0MpPAHZu5a8Djp5QfJK0rO24cBVJkiRJWryq+gLwhbZ9FbDPHHX+HXjJWAOTpB5yhJAkSZIkSVLPmBCSJEmSJEnqGRNCkiRJkiRJPWNCSJIkSZIkqWdMCEmSJEmSJPWMCSFJkiRJkqSeMSEkSZIkSZLUMztOOgBJkiRJ0nRZefSnhqq37tgDRxyJpFFxhJAkSZIkSVLPmBCSJEmSJEnqGRNCkqRFSXJikuuTXDJQ9uYk1ya5qP0cMLDvDUmuSPKtJM+ZTNSSJElSv5kQkiQt1knA/nOUH1dVq9rPuQBJHgMcCjy2PefvkuwwtkglSZIkASaEJEmLVFVfAm4csvrBwIer6vaq+g5wBbDPyIKTJEmSNCcTQpKkUXlNkovblLIHtbLdgWsG6qxvZZIkSZLGyISQJGkU3gs8AlgFbADeubUvkGRNkrVJ1m7cuHGp45MkSZJ6zYSQJGnJVdV1VXVnVd0FvJ+7p4VdC+w5UHWPVjbXaxxfVauravWKFStGG7AkSZLUMyaEJElLLsluAw9fCMzcgewc4NAk907yMGBv4Gvjjk+SJEnqux0nHYAkafuW5DRgX2CXJOuBY4B9k6wCClgHvAqgqi5NcgZwGXAHcFRV3TmJuCVJkqQ+WzAhlORE4CDg+qp6XCt7M/C7wMyiDm8cuKXwG4AjgTuBP6yq/z2CuCVJU6KqDpuj+IQt1H878PbRRSRJkiRpIcNMGTsJ2H+O8uOqalX7mUkGPQY4FHhse87fJdlhqYKVJEmSJEnS4i2YEKqqLwE3Dvl6BwMfrqrbq+o7wBXcvZCoJEmSJEmSpsBiFpV+TZKLk5yY5EGtbHfgmoE661uZJEmSJEmSpsS2JoTeCzwCWAVsAN65tS+QZE2StUnWbty4ceEnSJIkSZIkaUlsU0Koqq6rqjur6i7g/dw9LexaYM+Bqnu0srle4/iqWl1Vq1esWLEtYUiSJEmSJGkbbFNCKMluAw9fCFzSts8BDk1y7yQPA/YGvra4ECVJkiRJkrSUhrnt/GnAvsAuSdYDxwD7JlkFFLAOeBVAVV2a5AzgMuAO4KiqunM0oUuSJEmSJGlbLJgQqqrD5ig+YQv13w68fTFBSZIkSZIkaXQWc5cxSZIkSZIkbYdMCEmSJEmSJPWMCSFJkiRJkqSeMSEkSZIkSZLUMyaEJEmSJEmSesaEkCRJkiRJUs+YEJIkSZIkSeoZE0KSJEmSJEk9Y0JIkiRJkiSpZ0wISZIkSZIk9YwJIUmSJEmSpJ4xISRJkiRJktQzJoQkSZIkSZJ6xoSQJEmSJElSz5gQkiRJkiRJ6hkTQpIkSZIkST1jQkiSJEmSJKlnTAhJkiRJkiT1jAkhSZIkSZKknjEhJElalCQnJrk+ySUDZX+R5JtJLk5ydpKdWvnKJD9OclH7ed/kIpckSZL6y4SQJGmxTgL2n1V2HvC4qvpF4P8CbxjYd2VVrWo/rx5TjJIkSZIGmBCSJC1KVX0JuHFW2Weq6o728Hxgj7EHJkmSJGleJoQkSaP228CnBx4/LMnXk3wxydMnFZQkSZLUZyaEJEkjk+RPgTuAD7WiDcBeVfV44HXAqUkeMM9z1yRZm2Ttxo0bxxOwJGnJJfmpJF9L8m9JLk3yllb+sCRfTXJFktOT3KuV37s9vqLtXznJ+CVpuTIhJEkaiSSvAA4CXlZVBVBVt1fVDW37AuBK4FFzPb+qjq+q1VW1esWKFWOKWpI0ArcDz6yqXwJWAfsneTLwDuC4qnok8APgyFb/SOAHrfy4Vk+StMRMCEmSllyS/YHXA8+vqtsGylck2aFtPxzYG7hqMlFKksahOj9sD+/Zfgp4JnBmKz8ZeEHbPrg9pu3fL0nGFK4k9caCCSFvJyxJ2pIkpwFfAR6dZH2SI4H3APcHzpvVHzwDuDjJRXQn+a+uqhvnfGFJ0rKRZIfW9l9PdyfKK4GbBm5AsB7YvW3vDlwD0PbfDOw83oglafnbcYg6J9Gd2J8yUHYe8IaquiPJO+huJ/wnbd+VVbVqSaOUJE2tqjpsjuIT5ql7FnDWaCOSJE2bqroTWNUuJJ8N/NxiXzPJGmANwF577bXYl5Ok3llwhJC3E5YkSZK0FKrqJuDzwFOAnZLMXKDeA7i2bV8L7AnQ9j8QuGGO13KtOUlahKVYQ8jbCUuSJEmaU1s/bmaJifsAzwYup0sMvbhVOwL4eNs+pz2m7f/czM0JJElLZ5gpY/Pawu2Eb0jyROBjSR5bVbfM8VyHeEqSJEnL327Aye2mAvcAzqiqTya5DPhwkj8Hvs7d041PAP4+yRV0MxUOnUTQkrTcbXNCaOB2wvsN3k6Y7raSVNUFSWZuJ7x29vOr6njgeIDVq1eb8ZckSZKWoaq6GHj8HOVXAfvMUf7vwEvGEJok9do2TRnzdsKSJEmSJEnbrwVHCLXbCe8L7JJkPXAM3V3F7k13O2GA86vq1XS3E35rkv8E7sLbCUuSJEmSJE2dBRNC3k5YkiRJkiRpeVmKu4xJkiRJkiRpO2JCSJIkSZIkqWdMCEmSJEmSJPWMCSFJkiRJkqSeMSEkSZIkSZLUMyaEJEmSJEmSesaEkCRJkiRJUs+YEJIkSZIkSeoZE0KSJEmSJEk9Y0JIkiRJkiSpZ0wISZIkSZIk9YwJIUmSJEmSpJ4xISRJkiRJktQzJoQkSZIkSZJ6xoSQJEmSJElSz5gQkiRJkiRJ6hkTQpIkSZIkST1jQkiSJEmSJKlnTAhJkiRJkiT1jAkhSZIkSZKknjEhJEmSJEmS1DMmhCRJi5LkxCTXJ7lkoOzBSc5L8u32+0GtPEn+OskVSS5O8oTJRS5JkiT1lwkhSdJinQTsP6vsaOCzVbU38Nn2GOC5wN7tZw3w3jHFKEmSJGmACSFJ0qJU1ZeAG2cVHwyc3LZPBl4wUH5Kdc4Hdkqy23gilSRJkjTDhJAkaRR2raoNbft7wK5te3fgmoF661vZZpKsSbI2ydqNGzeOLlJJkiSph0wISZJGqqoKqG143vFVtbqqVq9YsWIEkUmSJEn9NVRCyAVDJUlb6bqZqWDt9/Wt/Fpgz4F6e7QySZIkSWM07Aihk3DBUEnS8M4BjmjbRwAfHyh/ebt48GTg5oGpZZIkSZLGZKiEkAuGSpLmk+Q04CvAo5OsT3IkcCzw7CTfBp7VHgOcC1wFXAG8H/j9CYQsSZIk9d6Oi3ju1i4Y6hVgSVqGquqweXbtN0fdAo4abUSSJEmSFrIki0pvy4Kh3j1GkiRJkiRpMhaTEFrUgqHePUaSJEmSJGkyFpMQcsFQSZIkSZKk7dBQawi1BUP3BXZJsh44hm6B0DPa4qFXA4e06ucCB9AtGHob8MoljlmSJEmSJEmLMFRCyAVDJUmSJEmSlo8lWVRakiRJkiRJ2w8TQpIkSZIkST1jQkiSJEmSJKlnTAhJkiRJkiT1zFCLSkuSpK2z8uhPDVVv3bEHjjgSSZIkaXOOEJIkSZIkSeoZE0KSJEmSJEk9Y0JIkiRJkiSpZ0wISZIkSZIk9YwJIUmSJEmSpJ4xISRJkiRJktQzJoQkSZIkSZJ6xoSQJEmSJElSz5gQkiRJkjQySfZM8vkklyW5NMlrW/mDk5yX5Nvt94NaeZL8dZIrklyc5AmT/QskaXkyISRJkiRplO4A/riqHgM8GTgqyWOAo4HPVtXewGfbY4DnAnu3nzXAe8cfsiQtfyaEJEmSJI1MVW2oqgvb9q3A5cDuwMHAya3aycAL2vbBwCnVOR/YKcluYw5bkpY9E0KSJEmSxiLJSuDxwFeBXatqQ9v1PWDXtr07cM3A09a3stmvtSbJ2iRrN27cOLKYJWm52nHSAUiSpO3fyqM/NVS9dcceOOJIJE2rJPcDzgL+qKpuSfKTfVVVSWprXq+qjgeOB1i9evVWPVeS5AghSZIkSSOW5J50yaAPVdVHW/F1M1PB2u/rW/m1wJ4DT9+jlUmSlpAJIUmSJEkjk24o0AnA5VX1VwO7zgGOaNtHAB8fKH95u9vYk4GbB6aWSZKWiFPGJEmSJI3SU4HDgW8kuaiVvRE4FjgjyZHA1cAhbd+5wAHAFcBtwCvHG64k9YMJIUmSJEkjU1VfBjLP7v3mqF/AUSMNSpJkQkiSNBpJHg2cPlD0cOBNwE7A7wIzt4R5Y1WdO+bwJEmSpF4zISRJGomq+hawCiDJDnQLgp5NN/T/uKr6ywmGJ0mSJPWai0pLksZhP+DKqrp60oFIkiRJMiEkSRqPQ4HTBh6/JsnFSU5M8qBJBSVJkiT11TYnhJI8OslFAz+3JPmjJG9Ocu1A+QFLGbAkafuS5F7A84GPtKL3Ao+gm062AXjnPM9bk2RtkrUbN26cq4okSWI0qxkAABC6SURBVJKkbbTNCaGq+lZVraqqVcAT6W4JeXbbfdzMPhcKlaTeey5wYVVdB1BV11XVnVV1F/B+YJ+5nlRVx1fV6qpavWLFijGGK0mSJC1/SzVlzLUhJEnzOYyB6WJJdhvY90LgkrFHJEmSJPXcUiWEXBtCkrSZJPcFng18dKD4fyb5RpKLgV8D/stEgpMkSZJ6bNEJIdeGkCTNp6p+VFU7V9XNA2WHV9UvVNUvVtXzq2rDJGOUJEmS+mgpRgi5NoQkSZIkSdJ2ZMcleI3N1oYYuNrr2hCSJE2hlUd/aqh66449cMSRSJIkaRIWlRAaWBviVQPF/zPJKqCAdbP2aTvlFwdJkiRJkpaPRSWEqupHwM6zyg5fVESSJGlqDHtBQJIkSduXpbrLmCRJkiRJkrYTS7GGkCRJkiRJwuU2tP0wIbRM2QhJkiRJkqT5OGVMkiRJkiSpZ0wISZIkSZIk9YwJIUmSJEmSpJ4xISRJkiRJktQzJoQkSZIkSZJ6xoSQJEmSJElSz3jb+Z4b9vb0kiRJkiRp+XCEkCRJkiRJUs+YEJIkSZIkSeoZp4xto2GnWq079sARRyJJkiRJkrR1TAhpIkyoSZIkSZI0OSaEJEnS2HhBQJIkaTq4hpAkSZIkSVLPmBCSJEmSJEnqGaeMSZK0jAw7JWvaDfN3OK1MkiRp2zlCSJIkSZIkqWd6M0JouVwxlSRJkiRJWqzeJISWCxNbkiRJkiRpsZwyJkmSJEmS1DMmhCRJkiRJknrGKWNaUk5pkzQoyTrgVuBO4I6qWp3kwcDpwEpgHXBIVf1gUjFKkiRJfeQIIUnSqP1aVa2qqtXt8dHAZ6tqb+Cz7bEkSZKkMTIhJEkat4OBk9v2ycALJhiLJEmS1EuLnjLmdABJ0hYU8JkkBfyvqjoe2LWqNrT93wN2nVh02xGn5EqSJGkpLdUIIacDSJLm8rSqegLwXOCoJM8Y3FlVRZc02kySNUnWJlm7cePGMYQqSZIk9ceopow5HUCSRFVd235fD5wN7ANcl2Q3gPb7+nmee3xVra6q1StWrBhXyJIkSVIvLEVCaGY6wAVJ1rQypwNIUs8luW+S+89sA78OXAKcAxzRqh0BfHwyEUqSJEn9tRS3nX9aVV2b5GeA85J8c3BnVVVbO2ITLXm0BmCvvfZagjAkSVNmV+DsJND1N6dW1T8m+VfgjCRHAlcDh0wwRkmSJKmXFp0QGpwOkGST6QBVtWG+6QBtYdHjAVavXj3n+hF94mKhkpabqroK+KU5ym8A9ht/RJIkSZJmLGrKmNMBJEmSJEmStj+LHSHkdABJkiRJkqTtzKISQk4HkCRJy8Ww07fXHXvgiCORJEkavVHddl6SJEmSJElTyoSQJEmSJElSzyzFbeclSdruebdHSRqNJCcCBwHXV9XjWtmDgdOBlcA64JCq+kG6xUnfDRwA3Aa8oqounETckrTcOUJIkiRJ0iidBOw/q+xo4LNVtTfw2fYY4LnA3u1nDfDeMcUoSb1jQkiSJEnSyFTVl4AbZxUfDJzctk8GXjBQfkp1zgd2SrLbeCKVpH4xISRJkiRp3Hatqg1t+3vArm17d+CagXrrW5kkaYmZEJIkSZI0MVVVQG3t85KsSbI2ydqNGzeOIDJJWt5MCEmSJEkat+tmpoK139e38muBPQfq7dHKNlNVx1fV6qpavWLFipEGK0nLkQkhSZIkSeN2DnBE2z4C+PhA+cvTeTJw88DUMknSEvK285pqw94Get2xB471tSRJkjScJKcB+wK7JFkPHAMcC5yR5EjgauCQVv1culvOX0F32/lXjj1gSeoJE0KSJGm7ZKJf2j5U1WHz7NpvjroFHDXaiCRJYEJIkiQtc8MmjiRJkvrENYQkSZIkSZJ6xoSQJEmSJElSzzhlTMuC0wEkSZIkSRqeI4QkSZIkSZJ6xoSQJEmSJElSz5gQkiRJkiRJ6hnXEJIkaYJcA02SJEmTYEJoxDzRlyRJkiRJ08aEkCRJ0lYY9mLPumMPHHEkkiRJ286EkDSLJ/qSJEmSpOXORaUlSZIkSZJ6xoSQJEmSJElSz5gQkiRJkiRJ6hkTQpKkkUiyZ5LPJ7ksyaVJXtvK35zk2iQXtZ8DJh2rJEmS1DfbnBDyRF+StIA7gD+uqscATwaOSvKYtu+4qlrVfs6dXIiSJElSPy3mLmMzJ/oXJrk/cEGS89q+46rqLxcfniRpe1VVG4ANbfvWJJcDu082KkmSJEmwiBFCVbWhqi5s27cCnuhLkuaUZCXweOCrreg1SS5OcmKSB83znDVJ1iZZu3HjxjFFKkmSJPXDkqwhtC0n+pKkfkhyP+As4I+q6hbgvcAjgFV0I4jeOdfzqur4qlpdVatXrFgxtnglSZKkPlh0QmhbT/S98itJy1+Se9L1ER+qqo8CVNV1VXVnVd0FvB/YZ5IxSpIkSX20qITQYk70vfIrSctbkgAnAJdX1V8NlO82UO2FwCXjjk2SJEnqu21eVHpLJ/ptIVHwRF+S+uypwOHAN5Jc1MreCByWZBVQwDrgVZMJT5IkSeqvxdxlzBN9SdK8qurLQObY5W3mJUmSpAnb5oSQJ/qSJEmSJEnbpyW5y5gkSZIkSZK2HyaEJEmSJEmSemYxawhJvbby6E8NVW/dsQeOOBJJkiRJkraOI4QkSZIkSZJ6xhFCkiRJI+BIUkmSNM0cISRJkiRJktQz2/0IoWGvvkmSJEmSJKnjCCFJkiRJkqSe2e5HCEmStCWOJJUkSZI2Z0JIkiRpO+Ai1ZIkaSk5ZUySJEmSJKlnTAhJkiRJkiT1jAkhSZIkSZKknjEhJEmSJEmS1DMmhCRJkiRJknrGhJAkSZIkSVLPeNt5aUp4O2FJkiRJ0riYEJK2MyaOJEmSJEmL5ZQxSZIkSZKknjEhJEmSJEmS1DMmhCRJkiRJknrGNYSkERt2zR9JUj8tdT+x1GvNuXadJEnLkyOEJEmSJEmSesYRQpIkSZIkaaQccTp9TAhJy5RTBiRJkiRJ8zEhJEmS1EOucSdJUr+NLCGUZH/g3cAOwAeq6thRvZekbecXAk2K/YQkaT72EZI0eiNJCCXZAfhb4NnAeuBfk5xTVZeN4v0kbX+WyxS05fJ3jJv9hLT8LPUFhuXSbtpPbD37CEkaj1GNENoHuKKqrgJI8mHgYMBGXFrmpn3EkSfmU8N+QtIWTao/sf2fCvYRknph0t9NRnXb+d2BawYer29lkiSB/YQkaX72EZI0BhNbVDrJGmBNe/jDJN+aVCwL2AX4/qSDmEIel815TOa2qOOSdyxhJBN831mvt7XH5KFLGsx2YjvqJwbZDmzOYzI3j8vmpuKYTFm/M8wx6WUfAUvaTwz1f2/M/zem4vMwh83imtRnZsA0Hqtp/D8F28mx8v/U5vKORcU0bz8xqoTQtcCeA4/3aGU/UVXHA8eP6P2XTJK1VbV60nFMG4/L5jwmc/O4bM5jAiyjfmKQ/7ab85jMzeOyOY/J5np8TBbsI2Dp+olpPM7TGBNMZ1zGNLxpjMuYhjOqmEY1Zexfgb2TPCzJvYBDgXNG9F6SpO2P/YQkaT72EZI0BiMZIVRVdyR5DfC/6W4VeWJVXTqK95IkbX/sJyRJ87GPkKTxGNkaQlV1LnDuqF5/jLar6Qpj5HHZnMdkbh6XzXlMWFb9xCD/bTfnMZmbx2VzHpPN9faYjLmPmMbjPI0xwXTGZUzDm8a4jGk4I4kpVTWK15UkSZIkSdKUGtUaQpIkSZIkSZpSJoTmkWTPJJ9PclmSS5O8dtIxTVqSn0rytST/1o7JWyYd0zRJskOSryf55KRjmQZJ1iX5RpKLkqyddDzTIslOSc5M8s0klyd5yqRj0uLZZ2zOPmN+9hebs8/YnP3F0hqmnU7nr5NckeTiJE+Ygpj2TXJz+2xclORNI45pwbY7yb2TnN6O01eTrBxlTFsR1yuSbBw4Vr8z6rja+87bpk/iWA0R06SO0xbb+XF//oaMaayfv/aeW2z7l/o4jWwNoWXgDuCPq+rCJPcHLkhyXlVdNunAJuh24JlV9cMk9wS+nOTTVXX+pAObEq8FLgceMOlApsivVdX3Jx3ElHk38I9V9eJ0d0756UkHpCVhn7E5+4z52V/MzT5jU/YXS2uYdvq5wN7t50nAe9vvScYE8M9VddAI4xg0TNt9JPCDqnpkkkOBdwAvnYK4AE6vqteMOJbZttSmT+JYLRQTTOY4wZbb+XF//oaJCcb7+YOF2/4lPU6OEJpHVW2oqgvb9q10H6jdJxvVZFXnh+3hPduPi1ABSfYADgQ+MOlYNL2SPBB4BnACQFX9R1XdNNmotBTsMzZnnzE3+wsNw/5i6Q3ZTh8MnNLar/OBnZLsNuGYxmrItvtg4OS2fSawX5JMQVxjN0SbPvZjtR33M2P9/E2jIdv+JT1OJoSG0Ib2PR746mQjmbw2/PAi4HrgvKrq/TFp3gW8Hrhr0oFMkQI+k+SCJGsmHcyUeBiwEfhgG8b7gST3nXRQWlr2GXezz5iT/cXc7DM2ZX8xQltop3cHrhl4vJ4xJWgW6Due0qZKfTrJY8cQy0Jt90+OU1XdAdwM7DwFcQH8RptGc2aSPUcdEwu36ZM4VsP0M+M+TrBwOz+Jz98wfc84P3/DtP1LepxMCC0gyf2As4A/qqpbJh3PpFXVnVW1CtgD2CfJ4yYd06QlOQi4vqoumHQsU+ZpVfUEumGNRyV5xqQDmgI7Ak8A3ltVjwd+BBw92ZC0lOwzNmWfsSn7iy2yz9iU/cWITGM7vUBMFwIPrapfAv4G+Nio45nWtnuIuD4BrKyqXwTO4+6ROSMxjW36kDGN9TgNmMZ2fqGYxv35G3vbb0JoC9r81LOAD1XVRycdzzRpQ9c+D+w/6VimwFOB5ydZB3wYeGaSf5hsSJNXVde239cDZwP7TDaiqbAeWD9wRetMukZfy4B9xvzsM37C/mIe9hmbsb8YgSHa6WuBwdESe7SyicVUVbfMTJWqqnOBeybZZZQxDbz3fG33T45Tkh2BBwI3jCOmLcVVVTdU1e3t4QeAJ444lGHa9HEfqwVjmsBxmnnfhdr5sX/+FoppAp+/Ydr+JT1OJoTm0eZ2ngBcXlV/Nel4pkGSFUl2atv3AZ4NfHOyUU1eVb2hqvaoqpXAocDnquq3JhzWRCW5b1sckTbM8deBSyYb1eRV1feAa5I8uhXtB/R50eFlwz5jc/YZm7O/mJt9xubsL5bekO30OcDL03kycHNVbZhkTEl+dmbNmST70H1/G1lCYci2+xzgiLb9Yrq2bKTr+QwT16x1VJ5PtybTyAzZpo/1WA0T07iPU3vPYdr5cX/+Foxp3J+/Idv+JT1O3mVsfk8FDge+0eaqAryxZQb7ajfg5CQ70H0Yzqgqb5mruewKnN3azx2BU6vqHycb0tT4A+BD6e4acBXwygnHo6Vhn7E5+wwNyz5jbvYXS2vOdhrYC6Cq3gecCxwAXAHcxuiP+TAxvRj4vSR3AD8GDh1x8mXOtjvJW4G1VXUOXRLr75NcAdxIl3gYtWHi+sMkz6e7e9uNwCvGENdmpuBYLRTTJI7TnO18klfDxD5/w8Q07s8fzNH2j/I4ZfR/jyRJkiRJkqaJU8YkSZIkSZJ6xoSQJEmSJElSz5gQkiRJkiRJ6hkTQpIkSZIkST1jQkiSJEmSJKlnTAhJkiRJkiT1jAkhSZIkSZKknjEhJEmSJEmS1DP/P2SHTNQn8Gh1AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1440x360 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXgAAAEGCAYAAABvtY4XAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3de7xUdb3/8ddnZl+Qi0jgDRBIiQwIUHeiP5C8HD3lhfSIepI6pzoeTp1+pqnRyS5WHuukdjErk7S0X3SSS6WWaShesMQO6AZFsdAiwVTcB5WtMHvvmc/vj1kDszczs+e29tzez8ejB3uvmbXWl5He+7u/6/v9fM3dERGR+hOpdANERCQcCngRkTqlgBcRqVMKeBGROqWAFxGpU02VbkC6UaNG+YQJEyrdDBGRmrF27dpX3H3/TK9VVcBPmDCBNWvWVLoZIiI1w8w2Z3tNQzQiInUq1IA3s/3MbJmZbTSzp83s2DDvJyIie4Q9RHMdcLe7zzOzFmBwyPcTEZFAaAFvZsOBOcCHANy9C+gK634iItJbmEM0bwW2AT8ys8fN7CYzGxLi/UREJE2YAd8EHAnc4O5HAG8A/9H3TWa2wMzWmNmabdu2hdgcEZHq09EZY93zr9LRGSv7tcMcg98CbHH3R4Pvl5Eh4N19EbAIoK2tTaUtRaSudHTG2LJ9J2NH7MPIoa29Xru9fSufXr6e5kiE7kSCq8+extwZY8p279AC3t1fNLPnzezt7v4McBLwVFj3ExGpNrkCvKMzxqeXr2dXd4JdJABYuHw9syaO2usHQbHCngd/IbDYzNYDM4CvhHw/EZGqkB7gO2I97OpOsHD5+t1DMVu276Q50juCmyMRtmzfWbY2hDpN0t3bgbYw7yEiUo1SAZ7qncOeAB85tJWxI/ahO5HodU53IsHYEfuUrQ1aySoiEoL+Anzk0FauPnsag5ojDGttYlBzhKvPnla24Rmoslo0IiL1IhXgC/uMwacH+NwZY5g1cVTWh7ClUsCLiIQknwAfObS17MGeooAXEQlRmAHeH43Bi4jUKQW8iEidUsCLiNQpBbyISJ1SwIuI1CkFvIhInVLAi4jUKQW8iEidUsCLiNQpBbyISJ1SwIuI1CkFvIhInVLAi4gEwtwAuxJUTVJEhPA3wK4E9eBFpOH1t39qrVLAi0jDG4gNsCtBAS8iDW8gNsCuBAW8iDS8gdgAuxL0kFVEhPA3wK4EBbyISKCS+6eGQUM0IiJ1SgEvIjWl3hYjhUlDNCJSM+pxMVKY1IMXkZpQr4uRwqSAF5GaUK+LkcKkgBeRmlCvi5HCpIAXkZpQr4uRwqSHrCJSM+pxMVKYFPAiUlPqbTFSmDREIyJSpxTwIlL1tLipOBqiEZGqpsVNxVMPXkSqlhY3lSbUHryZ/QXYAcSBHndvC/N+IlJfUoubdrFn/ntqcZMetPZvIIZoTnD3VwbgPiJSIzo6Y3lNddTiptJoDF5EQpce6A9veiXvMfXU4qaFfd6v3nt+zN3Du7jZn4HtgAM3uvuiDO9ZACwAGDdu3FGbN28OrT0iMvDSH5J2xRPEEwl60jrlg5oj/O7TJ+YM7Xx7/I3IzNZmG/4Ouwc/2923mtkBwAoz2+juD6W/IQj9RQBtbW3h/bQRkQGX/pA0fRw9XT5j6lrcVJxQZ9G4+9bgz5eBXwBHh3k/EakumSpA9qUx9fCEFvBmNsTMhqW+Bk4BngzrfiJSfTI9JG2OGq1NpoJhAyDMIZoDgV+YWeo+P3X3u0O8n4hUmWwPSVUwbGCEFvDu/hwwPazri0htyFYBUsEePk2TFJHQ6SFpZahUgYgURQXAqp968CJSsNvbt7Jw2XqiESOecK6ZpwJg1Ug9eBEpSEdnjMuWriPWk+DNrjixngSXLFnHppd2VLpp0ocCXkQKsuGF1+iO916T2JNw3vvth7ijfWuFWiWZKOBFpECW8Wh3HJXyrTIKeBHJKtOD1Cmj96UpS3Kkyg5IdVDAi0hGt7dvZdbXVvKBmx5l1tdW7h5+GTm0lW+cO4OW6N49+a54XGUHqogCXqTB5DO9sb+dlObOGMMjnzmJS0+e1CtEEg6/26TtH6qFpkmKNJB89zfNZyelkUNbecuQll41IrvjzsLl65k1cZQWNlUB9eBFGkSmXvllSzNPb8xnJ6WOzhhf+tVTe50bNdM4fJVQwIs0iEyle7vizqnXP7zX9MZUkbBBzZGsVR+3bN+ZcRy+O67yv9VCQzQiDSJTrxygqyfBZcvWM/ngfZl44LDdx7MVCUu/Xk9i7z16rjhjioZnqoR68CINItUrzzj7pSfBqd9exR3tW3s9hB05tJXph+yXMbDTe/lDWqO0NEW46qypzD9m/ED8dSQPoe7JWqi2tjZfs2ZNpZshUtc2vbSDU69/mK6evXvzTRGIRiK0RPvfEDtF+6VWVq49WdWDF2kwEw8cxrXzptGSYbVSTwJiPZmnRmaTq5cvlaWAF2lAc2eM4a4LZ2ccrkmnlam1raiAN7ODyt0QESmffBYzTTxwGNeeM333TJnWJqO5T+BrQ+zaVuwsmpuB08rZEBEpXUdnjMWP/pXv3v8nWqLRXuPomcbK+86U+d2mV/baP1VDL7WrqIB3d4W7SJVJbcIRCx6exnp6gGSFxx27erjy109lXMGavp1ef1MjpbZkDXgze0uuE939f8vfHBEpRmqVaizDzJhoxPjSnRvoivvu0gO5yglo/9T6kasHvxZwMhd/duDQUFokIgXp6Ixx/8aXicf3DndI1odpjkboisd3H+tbV0bqU9aAd/e3DmRDRKRwqeJh7tCdId9bonDFGZO5sk/NGD08bQx5jcGb2T8As0n23Fe5+y9DbZWI9Cu9eFgm0Yhx0z+3MWfSAQxrbdLD0wbUb8Cb2feAicB/B4c+amYnu/vHQ22ZiOSUqaRvunjCGT082UvXw9PGlE8P/kTgHR7UNDCzW4ENobZKRHbLVgogW/GwlKYIvNG1Z9xdD08bTz4BvwkYB2wOvj8kOCYiIUrNaf/Oyj/RFI0QTzjXzJvWqyf++dMn89lfPJnx/GjENM7e4PIJ+GHA02b2B5Jj8EcDa8zsDgB3nxti+0Qa0u3tW/nU0nV0xZPFAFMzYC6+rZ3m6J5iYB8/fiJDWiK80bV3T/4Lp6tsb6PLJ+C/EHorRGS3TS/t4JLb2olnKPSa8GQxsNR89+/cv4lEhoqwg1siTB0zPOymSpXrN+Dd/cGBaIiIpHru6zOGeyYt0QgfOHYcNzzwXK/jCUfDM6JqkiLVIjXtsSvLgqVMuhMJLph9KFedOZWWqDGkJZpxez1pTNqyTyREhWyGseGF14lkXDieWXPUdgf5/GPG856pB2kapPSigBcJSWqVaaYCX5nem14oLB8Rg1kTR+3+XtMgpa9i68F/scztEKkr6atM+9sdKVuhsP768i3RqDbjkJyK7cGvLWsrROpMplWm2Qp8bdm+k0wbKzVHjbs+cRwvvLaL13d2c+nSdmI9e56+qp6M9KfYevB3lrshIvUk0yrTbIH85NbXMs5jb26K8EZXnDmT9gcg4a56MlKQfGrRfDvD4deANe5+e/mbJFL7Rg5t5eqzp/UbyB2dMa789VMZrxFPeK8fCKonI4XKpwc/CDgcWBp8fzbwZ2C6mZ3g7hfnOtnMosAaYKu7n15KY0WqXfqsmbkzxjD54H15eNM2Rg0dxLGHjdzr/dkKhrU0ZZ7qqAepUoh8An4aMMvd4wBmdgOwimT54CfyOP8i4Glg32IbKVILkjNh1hG1CD2JOHMmHZDciCMYNjfgkpMncf7McbtDOtNQTkvUuOvC2Uw8cNgA/w2k3uQzi2YEMDTt+yHAW4LAz75lO2BmY0luzn1T0S0UqQEdnTEuXZJ8CPpmd5yuONz79Mu9VqQ68PUVf+T//NdK7mjfCuwZyhnUHGFYaxODmiNce850hbuURT49+KuBdjN7gGQnZA7wFTMbAtzbz7nfAhaSLFiWkZktABYAjBs3Lo/miFSfR57tIN8p7LGeRK89UTW2LmHJpxbNzWZ2F8kqkgCXu/sLwdefynaemZ0OvOzua83s+BzXXwQsAmhra8uzAodI9Vi8ejNX3FHYFgl9p0xqbF3CkM8smjuBnwJ3uPsbBVx7FjDXzE4l+aB2XzP7ibt/oLimilSfxas389lfZq7HnsuunrjmsEvo8hmDvxY4DnjKzJaZ2TwzG9TfSe7+GXcf6+4TgH8EVircpdZ1dMZY9/yrdHTG6OiM8fkiwh3AM5T4FSm3fMsFPxhMdzwR+Ffgh2hWjDSYVG2ZqBnd8QTnHz0uy26o/WttimZc1SpSTnmtZDWzfYAzgPOAI4FbC7mJuz8APFBg20QqbtNLO2h//lUmjBzMwmXrepUKuOWRzTnOzK07rjIDEr58xuCXkHzAejfwHeBBdy+24yJSM77wyyf48eq/hnLtK87QdnoSvnx68DcD708tdBJpBJte2lF0uBsw860jWP3n7b2ODW6N0h13rjhjMvNnji9PQ0VyyGcM/h4zm2pmk0nOhkkd/3GoLROpoPbnXy363Fs/8i7mTDpg9/DOjEP2Y8SQFs1zlwGXzxDNFcDxwGTgLuC9wMOAAl7q1oqnXyz63NHDk2PrEw8c1mtFqoJdBlo+QzTzgOnA4+7+YTM7EPhJuM0SKY9CtsxLvffX67Zyz4aXi7rfoOZkiV+RapBPwO9094SZ9ZjZvsDLwCEht0ukZIVsmbfowWe5+p6N4NBT4hR1zY6RapFPwK8xs/2AH5DcyakTeCTUVomUKH3LvFQp3oXL1zP54H15oyveq0e/cNk6lqzZUtR9ZowdzsaXdmgTDqlK+Txk/ffgy++b2d3Avu6+PtxmiZQmW531U7+9itamKF3xBKe98yAmjBxcdLhHDW7+0Lt2308PUKXaFLRln7v/JaR2iJRVpjrru7qT33fFewD4+eMv7HVeoX636RXmzhijYJeqlE8tGpGa07fOektTJOPG1qWIe3LYp6Mz57YIIhVT1KbbIrUgvc76yqdf4rqVm8p+j75lf0WqSdaAN7N3AaPc/Td9jp8KvOTua8NunEipUsF7w0PPleV6RnJnppTuhGrKSPXK1YP/GvDhDMc3AD8iWVlSpOolH7gaXXm+/6hxw3li6w7MwB3On3kIE0YOYfbEUTz1t9dZ2GfqpXrvUq1yBfwwd9+rXJ67bzazUSG2SaRgHZ0xHnn2FTZ3vMn4kUM49rCRQDLch7REieW5n94BQ5tZ/u+zsy6QmnjgMG2vJzUjV8CPyPHa4HI3RKRYt7dv5ZO3tZPos0ApGjEMSLjv9Vo235t/FJB7Cz1trye1ItcsmnvN7Coz2z33wJK+DKwMv2ki/evojPGppesyBng84fQk8g/3pgg0N0XL20CRCsrVg78UuAnYZGbtwbHpwBrggrAbJpKPxY/+la54eba/a4pG9MBU6krWgA822H6/mR0KTAkOb3D38kxHEClRR2eM6+79Y8nXGdwcJYHrganUnVzTJI9M+3Zr8Od+qePu/liYDRPJpaMzxp3rXqDUzntrU4Tvf/AopozeV+EudSfXEM3Xc7zmaJqkDLDUzJbVz3Xw9RV/pCde/M6RrVHDIsbVZ09jzqT9y9hKkeqRa4jmhIFsiOxRSA3zRpEq/evuvTa+zlcE+Kdjx3PiOw5k9PBBe1WUFKlHuYZo5uQ4z919VQjtaXiF1DBvFOmlf4sx/+hxXHLKJIW5NJxcQzSfynDMgWkkN/zQfLIyy1bDfNbEUQ0dTlu276QpUlylsNamiMJdGlauIZoz0r83s1nA54AXgQtDbldDylTDvFGKWfUdlurojLHhhdfYun0nKze+TGesuG3wvnD65Lr/7ESyyWfT7ZOAz5PsvX/F3VeE3qoGlamGeSMUs0ofluqKxzlu4ihWbtxG8Y9Qkwa3RJk6ZnhZ2ihSi7KuZDWz08zs98BlwOfc/QSFe7j61jAf1Byp+7nZ6cNSO2I9xHqce8sQ7pAsUVDvPxxFcsnVg78T2AJ0AAvNbGH6i+4+N8yGNar0GuaNMMtjy/adeL61BPoRjRgRnJbmKPGEFi6J5Ap4TZOskEYqZjWkJUqsxNVKQ1r3BHoj/XAU6U+ugH8K2N/dn0o/aGaTgW2htkoaxgU//p+iztunyXAzPn/aZKaOGd4r0BXsIkm5Av564HsZjo8kOZvm/FBaJHUvOUPmdf5j+TpeeK3w/UybDK48852ccPgBCnORHHIF/ER3f6jvQXdfZWY3hNgmqROZVuR+87fP8O2VmyhlUCYSMYW7SB5y7uiU47XmcjdE6ktq6mPUjK54gvlHj2PVpm08u+3Nkq994YlvU7iL5CFXwG8ys1Pd/a70g2b2XkAlgyWrTKUFbnlkr90fi9LaFOH8mePKci2Repcr4C8Gfm1m5wJrg2NtwLHA6WE3TGpL+nDMlu07c24VVozBLRESjqY+ihQgV6mCP5nZO0k+TJ0aHH4Q+Dd33zUQjZPa0LdA2tzpo3mzyMJgfbU2RfjCGZOZOnq4pj6KFChnqQJ3jwE/GqC2SA3a9NIOPrV0HV1x311DZ8maLSVfNwJ88uRJnD9znEJdpEi5ygXvgIyTHYxkueB9Q2uV1ITb27fuDvdyaYrAt86bwbGHNXYFTZFyyDVEk2sWTb/MbBDwENAa3GeZu19RyjWlenR0xrhs6Tq6yxDuUYPBLU2769+fPr2x69+LlEuuHvwg4KPARGA98EN37yng2jHgRHfvNLNm4GEz+427ry6pxVIR6Q9RAe5ct7XkcG+KGHdfdBwjhrSovIBICHKNwd8KdAOrgFOBKcBF+V7Y3R3oDL5tDv5Xvt/lZcCkP0Td2d2DmZHw0v9Tful9U5h4YPIXRQW7SPnlCvjJ7v5OADO7GfhDoRc3syjJKZYTge+6+6MZ3rMAWAAwbpzmN1ebTLtMlfpzOmrw5fdNZf7M8aU3UESyyhXw3akv3L3HrPAt09w9Dswws/2AX5jZVHd/ss97FgGLANra2tTDrzIbXnidCMVtl9eXAde/Xw9QRQZKroCfbmavB18bsE/wfcGzaNz9VTO7H3gP8GR/75fKSR9rf3jTK1y2ZB3dJdZrb4oka7VfM2+6HqCKDKBcs2hK2lTbzPYHuoNw3wc4GfhaKdeUcC1evZkv/eopWqJGV4/TFS99sVJrU4Qf/FMbU0bvq167yADrd0/WEhwM3BqMw0eAJe7+qxDvJyVYvHozn/1l8perrkLmSuXQ2hThmnnTmDNp//JcUEQKElrAu/t64Iiwri/ls+mlHVxx54ayXvNjxx/KBbMPVa9dpILC7MFLDUitRu0p02rUCHDlmVOZf4xmyIhUmgK+gfTdgCM1BbJcpQaiBvdcPGf33HYRqSwFfINIX6zUFU9w2jsP4uDhg+juKU/Vx6YIfOPcGQp3kSqigG8AmRYr/fzxF0q+bnME5s8cz4nvOIApo4drvF2kyijgG8CW7TvLer0mgx9++F0KdZEqp4BvAN098V7b55Viz9THA8pyPREJjwK+zi1evZkr7iht8fCgJnCM/3vC27QBh0gNUcDXsRsffJav/mZjSdc4t20s82eOVylfkRqkgK9T3/ztM1y3clPR57/7baP4/OmTNStGpIYp4OtMR2eMi3/2OKs2dRR9jXPbxnL1vOllbJWIVIICvsalFi8NaYmy7LEt3Pjgc0VXaz9+0ig+d5p67SL1QgFfw1KLl4CSZ8m0ROHr587QOLtIHVHA16j0xUulihpce47CXaTeKOBrUEdnjPs3vkwiUXq4/93hB/C1edMU7iJ1SAFfY1LDMvF4gmI77xGDS0+exN9POUjj7SJ1TAFfQzo6Y3xq6bqiqj9GgOamCADXzJvG3BnaOk+k3inga8jCZcWFe3PU+M0njuONrrgWLIk0EAV8lUpNf+zuiXP/My/z4DPbePJvOwq+TkvUuPac6RqKEWlACvgqtHj1Zr545wYSCaeUvThOm3oQXz5zqnrsIg1KAV9l0je/LtbHjz+Us44Yq167SINTwFeRjs4Ynysx3AE+os2uRYTk5AqpAh2dMT6++LGiywykDG6OlH2DDxGpTerBV4FyDMukJICxI/Ypy7VEpLYp4CukozPGhhde59frX+C2NVsKOvfo8SP4hyPHcPKUg1i+dgtX37ORlmiEBHD12VqVKiJJCvgKuL19K5fc1l7wDBkD/vOsqcyfOX73sQXvPoyzjxrLlu07NcddRHpRwA+wTS/t4JM/a6fQKgNfPWsqp0w5KGOAjxzaqmAXkb0o4AdIR2eMH6x6jkUPPldwuJ/bNpb3p/XaRUTyoYAfAIuCvVGLmSFz0YkT+eQpby97m0Sk/ingQ1TK9nnNUeOLc6f0Gm8XESmEAj4EHZ0xrr/vT9zyyOaCz22JGhee+DbOnzlO4+oiUhIFfBl1dMZY/Ohf+caKPxZ1/seOP5QLtApVRMpEAV+C9A2vlz+2hR+seo6eIjfhuOrMqcw/RsMxIlI+CvgipXZWcodYkakeNfjgMeO58KS3qdcuImWngC9COTa8bokad33iOFV8FJHQqNhYEX674cWSwr1Zm3CIyABQD74AHZ0xzr3x9zy77c2ir/Gd9x/BsYeN1JCMiIROAZ+HUmfHpFx15lROnz66TK0SEckttIA3s0OAHwMHAg4scvfrwrpfuaVmyKx+roNr7tlY9OyY1ibDHa7QoiURGWBh9uB7gEvd/TEzGwasNbMV7v5UiPcsi9QMmZ7uBD0lXOfctrHMnzleVR5FpCJCC3h3/xvwt+DrHWb2NDAGqOqA7+iMsXDZOmI9xe+tdPaRo/nYuyfqIaqIVNSAjMGb2QTgCODRDK8tABYAjBs3biCak9P19/2ppHAf1Bzh8lMnq8cuIhUXesCb2VBgOXCxu7/e93V3XwQsAmhrayt1S9Ki3ffUi3z2l0/y4uuxos6PGjQ3RbSjkohUjVAD3syaSYb7Ynf/eZj3KsVJ197Ps68UP/Vx7vSD+ZfZh2qsXUSqSpizaAy4GXja3b8R1n1KddmS9qLCfdyIVj4y+zBmTxylsXYRqUph9uBnAR8EnjCz9uDY5e5+V4j3zEtqCuT37t/EPU+9VPD5y/7tGNreOjKElomIlE+Ys2geJrlPdFW58cFnueaeZ3D3gje9HtZiPPHlU8NpmIhImTXUStaFS9tZsnZrUedGgAcWnlTeBomIhKghAr6jM8ZvN7xYdLhHDb553gw9QBWRmlL3AX97+1Yuua294OGYqEEkYlww+61ccJx2WRKR2lPXAb/mzx1c/LN2Csl2A5b+2zE0N0U17VFEalpdBnxHZ4xPL1vHvRu3FXzuf541VTNkRKQu1F3A396+lUuXrKMnUdiYTMTgyvdNVcVHEakbdRPwHZ0xHnm2o+Bwb47AZ0+bzBnTR2s4RkTqSl0EfLLX3l5UzfZoNKJwF5G6VPMBnyrvm0+4D26JEOtOgMGgpihxdxUHE5G6VfMBv2X7zryWy17+3sOZeehIxo7YZ/d5miUjIvWs5gP+ya2vsauf+u2Xn3o4C+Yc1uuYgl1E6l1NB3xHZ4wrf519g6iIwZVnamaMiDSmmg74Ldt30hyJsIveA/BRgwVzDtUKVBFpaDUd8GNH7EN3one4t0SNuz5xnGq0i0jDi1S6AaUYObSVq8+exqDmCMNamxjUHOHac6Yr3EVEqPEePMDcGWOYNXGUZsWIiPRR8wEPyZ68gl1EpLeaHqIREZHsFPAiInVKAS8iUqcU8CIidUoBLyJSp8y9wM1KQ2Rm24DNlW5HFqOAVyrdiCqlzyY7fTbZ6bPJrpDPZry775/phaoK+GpmZmvcva3S7ahG+myy02eTnT6b7Mr12WiIRkSkTingRUTqlAI+f4sq3YAqps8mO3022emzya4sn43G4EVE6pR68CIidUoBLyJSpxTw/TCzQ8zsfjN7ysw2mNlFlW5TtTCzQWb2BzNbF3w2X6p0m6qJmUXN7HEz+1Wl21JtzOwvZvaEmbWb2ZpKt6damNl+ZrbMzDaa2dNmdmwp16uLcsEh6wEudffHzGwYsNbMVrh79s1gG0cMONHdO82sGXjYzH7j7qsr3bAqcRHwNLBvpRtSpU5wdy106u064G53n2dmLcDgUi6mHnw/3P1v7v5Y8PUOkv+HHVPZVlUHT+oMvm0O/qen9oCZjQVOA26qdFukNpjZcGAOcDOAu3e5+6ulXFMBXwAzmwAcATxa2ZZUj2AYoh14GVjh7vpskr4FLIQ+O8JLigO/NbO1Zrag0o2pEm8FtgE/Cob2bjKzIaVcUAGfJzMbCiwHLnb31yvdnmrh7nF3nwGMBY42s6mVblOlmdnpwMvuvrbSbalis939SOC9wMfNbE6lG1QFmoAjgRvc/QjgDeA/SrmgAj4PwfjycmCxu/+80u2pRsGvkvcD76l0W6rALGCumf0F+Blwopn9pLJNqi7uvjX482XgF8DRlW1RVdgCbEn7LXgZycAvmgK+H2ZmJMfEnnb3b1S6PdXEzPY3s/2Cr/cBTgY2VrZVlefun3H3se4+AfhHYKW7f6DCzaoaZjYkmLBAMARxCvBkZVtVee7+IvC8mb09OHQSUNJkDs2i6d8s4IPAE8FYM8Dl7n5XBdtULQ4GbjWzKMnOwhJ315RA6c+BwC+SfSeagJ+6+92VbVLVuBBYHMygeQ74cCkXU6kCEZE6pSEaEZE6pYAXEalTCngRkTqlgBcRqVMKeBGROqWAl5zM7ENmNrrS7cglaON3+nnPR83snzIcn2BmBc3BNrP/NrP1ZvbJQtta4H0OMrOfmdmzwZL+u8xsUtBmN7ML0977HTP7UAhtaDWze4Oqj+cFy+cnB6/9xcxGlfueUj6aBy/9+RDJRSgvVLgdJXH375fjOmZ2EPAud5+Y4bUmd+8p032M5ArPW939H4Nj00nOIX+eZO2fi8zsRnfvKvFeudp9BEBQjgLgtlLuJQNLPfgGZGYfCOq4t5vZjUHBsKiZ3WJmTwZ1uj9pZvOANpILL9qD1arZrnm0mT0SFEn6fWo1npmtNrMpae97wMzaglWwK4I68jeZ2eb+eoPBudcFbXnSzPZa3h70blcGPez7zGxccPyLZnZZ8PVRlqxhvw74eNq5D5nZjLTvH90R6QgAAARmSURBVA5CNd1vgTFBG44L2vQtS9Y0v8jMTgo+gyfM7Idm1hpc6y9m9tXgvDVmdqSZ3RP0zj+a4a97AtCd/oPJ3de5+6rg223AfcA/9/OZ3WJm3w/u+cegTk7qt547zGwlcJ+ZvcXMfhl8bqvNbJqZHQD8BHhX0O7DUv/9Mtxnr39TudolA0MB32DM7B3AecCsoFcWB+YDM4Ax7j7V3d8J/MjdlwFrgPnuPsPdd+a49EbguKBI0heArwTHbwPODe59MHCwu68BriC5hH8KyZob4/L8KwwO2v3vwA8zvH49yV7vNGAx8O0M7/kRcKG79w3vm0n+xoKZTQIGufu6Pu+ZCzwbfB6psG1x9zbgu8AtwHnBZ9gEfCzt3L8GbV8VvG8ecAyQaaOUqUB/xcq+BlyWR5hOIFnr5TTg+2Y2KDh+JDDP3d8dtOHx4HO7HPhxUCfmAmBV8Pd9NtPFc/ybkgpTwDeek4CjgP+xZOmFk4BDSS6LPtTMrjez9wCFVswcDiwNxrO/CaR67UtIBhkkg35Z8PVskoW4CJapb8/zPv8dnPMQsK8FtXDSHAv8NPj6/wX32S14/37B+an3pCwFTrdkcbmPkAzhfKSGLd4O/Nnd/xh8fyvJ+t4pdwR/PgE86u473H0bEMvw9+iXuz9HsnT1+f28dYm7J9z9TyT/Ox8eHF/h7v8bfD2b4LNw95XASDPLd6OSbP+mpMI0Bt94jGQP9zN7vZAcjvh74KMkw/gjBVz3SuB+dz/LknXzH4Bk1UAz6zCzaSR7eZmGIwrRt7ZG2WptuPubZrYCeB/Jv/9ReZ76Rp7viwV/JtK+Tn3f9/+LG9jzgzGXr5D8oflgjvdk+8zybXd/sv6bkspSD77x3AfMC8ZXCcZexwfj3xF3Xw58jj1lSncAw/K47nBga/D1h/q8dhvJzS+Gu/v64Njv2DN0cwowIs/2nxecMxt4zd1f6/P670lWcITkMMGq9BeDssavBuen3pPuJpLDOv/j7vn+VpHyDDDBzFIPYD9I7uDNZSXQammbYQTj4selv8ndN5KsOHhGjmudY2YRMzuMZM/6mQzvWUXwWZjZ8cArBex7kPHfVJ7nSogU8A0m2Ev2cyR301kPrCBZFXIM8EDwK/ZPgFRv7BaS47btZraPmX3ZzOZmuPTVwFfN7HH27o0uIxm6S9KOfQk4JRjSOQd4keQPEyw5HTDb1MxdwT2+D/xLhtcvBD4c/N0+SHJf1L4+DHw3+Lta+gvBJh2vkxynL4i77wquvdTMniDZMy9q9o4nqwCeBfxd8CB2A/BVkp9TX1eR3HAlm78CfwB+A3w0aGdfXwSOCj63/6Kfh7d92prt35RUmKpJSkUEs0vi7t5jyZ3jb0ibipftnAeAy4KHtGG1azTJ4aXD3b3mt9szs1uAXwUPzKXBaAxeKmUcsMTMIkAX8K8Vbg+WXAh1FXBJPYS7iHrwIiJ1SmPwIiJ1SgEvIlKnFPAiInVKAS8iUqcU8CIider/A18YV9KRwXPEAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(20,5))\n",
    "plt.subplot(131)\n",
    "tmp = plt.hist(est_ploidies[\"avg_pl\"],bins=30)\n",
    "plt.title(\"est. avg. ploidy from CN profile\")\n",
    "plt.subplot(132)\n",
    "tmp = plt.hist(sorted(list(GDSC_Ploidies[\"average_ploidy\"].values)),bins=30)\n",
    "plt.title(\"PICNIC avg. pl.\")\n",
    "plt.subplot(133)\n",
    "tmp = plt.hist(est_ploidies[\"median_pl\"],bins=30)\n",
    "plt.title(\"est. median ploidy\")\n",
    "\n",
    "tmp = df_ploidies.plot.scatter(x = \"est. avg. ploidy from CN profile\",y=\"PICNIC avg. pl.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXUAAAEcCAYAAAA4BiRaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3de5zcVX3/8debJCZAYiIXo0BIvP10MRRqUqIYIWsUKFjFFrWpt9T4S/lp1wtajcbKpcaSSq0IrSgEEgQSlYtCKCloNmpoxSYlXJd6QUAQRRBCFiIm4fP745xNJrOzu7OZ2Z3Z77yfj8c8dma+Z873M2e/85nvnO/5nq8iAjMzK4a9Gh2AmZnVj5O6mVmBOKmbmRWIk7qZWYE4qZuZFYiTuplZgTipjxCSQtJLGx1HI0maI+nBfpa3fBuVknSfpDfk+5+WdFETxLRO0vurKHeopG5Jo/pYfoaky+of4cjnpD5I+YOyNW9wj0u6XtKURsfVQ9J8SesbHUcrkrRc0ucaHUclEfH5iBgwmTaLiHggIsZHxI5GxzLSOKnvmT+LiPHAC4HfAOc1OJ4hI2l0o2Mws+o5qdcgIn4PXAkc1vOcpImSLpX0W0n3S/qMpL0k7SfpQUl/lsuNl/QzSe/Jj5dLukDSTZK2SPq+pKmV1tvPOtqAC4DX5F8ST/Tx+hdJ+kFez3cl/WvPT1lJ03I3xgJJDwBrc92fyet6JK97Yi7fq0uk7Gf/GZKulPSNvL7/kXRESdmDJF2V38svJH2oZNneuV0el3Q38CdV/FtOlHSvpEclfSHH/hxJv5N0eEndz5f0tKQD+2ij90nqyuv+j57/hZJ/ye3wpKQ7JE2XtBB4J/CJ3PbXDRRoT9tJ+kSu72FJJ0s6UdJPcsyfLim/l6RFkn4u6TFJ35S0X8nyd+f/0WOSFpeta7fuCknfkvRrSZvztvDKkmXL8zZxff6f3SLpJVW0fc8vxZslnZ/rvkfS3D7K9rdd9WyHo/PjFyl9JrZIugk4oKSe6yV1lNV9u6S3VhNz4USEb4O4AfcBb8j39wFWAJeWLL8U+A4wAZgG/ARYkJcdB/waeD5wIXBlyeuWA1uAY4CxwLnA+pLlAby0inXML31dH+/hv4BzgOcAs4Engcvysml5XZcC+wJ7A+8Dfga8GBgPXA18PZefAzzYTxudAWwDTgHGAB8HfpHv7wVsBD6bY3kxcC9wfH7t2cAPgf2AKcCd5esqW28Anbn8obld3p+X/RuwtKTsh4Hr+qjnLfn9tgGjgc8A/5mXHZ9jngQol3lhyf/wc4PYluYA2/P7HwP8X+C3wBX5f/tKYCvwopKYfwQckreRrwIr87LDgO6S7eeLue7S/8NlJet+X17HWOBLwKaybfEx4Kj8/i8HVlX5nubn9X40v6d3AJuB/fLydSX/k/62q2n5/zm6ZJv9Yo73GNJnpWebfTtwS0kMR+T4n9PofNGIW8MDGGk3UsLqBp4gJatfAYfnZaOAPwCHlZT/G2BdyePzgDuAh4D9S55fXvrByRv5DmBKfhzASwdaBwMkdVKy2w7sU/LcZfRO6i8uWf494AMlj1+e3/toqkvqPypZthfwMPA6YBbwQNlrPwVcku/fC5xQsmxh+brKXhtl5T8AfC/fnwU8ACg/3gC8vY96biB/SZbE/DQwFXg96cvi1cBeZa9bzuCT+lZgVH48Ib+HWSVlNgIn5/tdwNySZS8s+T98tmz72TdvJxWTelkck/J6J5a8j4tKlp8I3FPle5pP+kyo5LkfA+/O99exK6n3t131bIej2bXN7ltS9gp2bbPjgMeBl+XH5wD/NpR5oJlv7n7ZMydHxCTSxvS3wPclvYD0k3AMcH9J2fuBg0sefw2YDiyPiMfK6v1lz52I6AZ+BxxUVqaadfTnIOB3EfF0pfX28dxBFdY3Gphc5TpL39ezwIO5zqnAQZKe6LkBny6p96CyOEpjGHBdufxBeb23kBLzHEmvIH1BXttHHVOBc0ti+h1pr/zgiFgLnA/8K/CIpK9Jem4VcfXlsdh1MHBr/vubkuVbSV/wPXFdUxJXF+mLfzJlbRURT5H2VnuRNErS2bkb50nSlzCUdGmQflH2eLokhmo8FDm7Zjv/D2Wq3a4OAh7P76m0LLCzG/QbwLsk7QXMA74+iHgLxUm9BhGxIyKuJn2wZgOPkvY0SvvCDyXtlaM0POtrpK6ND6j38Ludo2gkjSd1I/yqrEy/6yDt3fTnYWA/SftUWm/p2yu5/6sK69tOSj5PkbqheuIeBZT3U5e+r71I3Qe/IiWhX0TEpJLbhIg4sSTW0tgOHeC9lb+XQ9m9/VYA7wLeTer6+n0fdfwS+JuyuPaOiP8EiIgvR8QMUpfH/wH+Lr+urlOeSlrH7gnul8CflsU1LiIeoqyt8v93/z6q/itSF9MbgImkvWJIX1z1cLCk0rrK/w89+tuuSj0MPE/SvmVlS60gHdOYCzwdEf+1J4EXgZN6DfJBs7cAzwO68h7XN4Elkibkg2unkbo3IO2FBqkv8QvApdp9HO6JkmZLeg7wD6Rui932oqtYx2+AQ3IdvUTE/aSuhzPyAcTXAH82wFtdCXw0H6waD3we+EZEbCd1RYyTdJKkMaT+57HAZZK25vc8S9LafBDsI7m9DiP9LN8i6bOSzpP0gNLBywclfQm4DvhUfv5RoPQg6vtz0ut53JNQ/07S8yQdAywFjsoH7G4nHR94KymxX9rP+70gr/eVue6Jkt6W7/+JpFn5vT4F/B54Nr/uN6T+4aFyAen/3nPQ9sC8/UE6YP+mku3nLPr+fE8AniHtye9D+n/W0/OBD0kak9utDfj3CuX62652Ktlmz8zb7GzKttmcxJ8F/pkW3ksHJ/U9dZ2kbtIBxiXAeyPirrysg/RhvxdYT+r7u1jSDFLyfU9OzEtJCX5RSb1XAKeTfu7PICWfSiquIy9bC9wF/DonwkreCbyG9KH+HOmn6zP9vN+LSR+UH5AOcv4+x0BEbCb1XV9E+rXwFKl7BdIH7/PA9fn9PELaS74T2JHb4a2kA4ALSXuNt5MS/2Okg573k/bsJ5XU25/vALeR+m5/ArwiIiYCbyN1udxOavcf9lVBRFxD+v+syt0TdwJ/mhc/l3SQ+/Ec22OkL2iAZcBhuXvk2wCSblDJCJYanUvqMrpR0hbSQdNZOea7gA+StoWHc3x9tdelOfaHgLtzPVWTdJekd/ZT5BbgZaRflUuAUyp0NUI/21UFf0V6r78jfUYqfSlfChzOrh2c1tToTn3f0o1BHmSr87q/AZxZ5zrvI/28P4P0IfsCsDovW8eug2XvJ+3hjh+grkWkD/SkktetKylTOjroMuD6Puq6uLydSb8cVpNGnjye7x+Sl70D2FBW/qPAtfn+/qRfFE8C/036kux39FFJPfOBm0l99JuBe9j9QOjOdhopN6oYfTWE635Po9bdTDfvqbeg3IXwkjxO+ARS/+q3h3CV+5BGUNxaYdkbgDWRDgz3ZwMpyX28ivW9gdQdsRtJ04A/J+1Rl9oLuITUv3so6eDk+XnZdcDLJb2spPxfkfaIIR0wfQp4AfDefBuMWcDPSQcpTweuVsnYc6tOPobwAdIxq5bmpN6aXkBKkN3Al4H/FxGVEm6tvk3awz4O+D6V+273J3UXVOOzQIf6OGGovzol/QOpG+ULEfGL0mUR8VhEXBURT0fEFlKXwbF52dOkLp15uZ6XAa8Ars3HQ/4COD2/9m7SAbvBeAT4UkRsi4hvAP8LnDTIOlqapONJv7J+w64v25blU8CbRETMH8Z1XUfaAx1qJ0fEdwco8xhpvPWAIuJOSatJXxRdg6kzIv4e+PtKhfNe3r8AJ5C6YgAmSBoVqd//CtIBuLNIe+nfjoin8zDW0ew+jLLS8ND+VDv8b0SIiOWkrsThXOd/kMblG95Tt8b7LnB82XC1/pxOOvOyv3H53yXtQVfrY6QTX2ZFxHNJZyzCriF+NwEHSjqStMfeszf4W9IQvENK6hrs5G7VDv8zq4qTujXa10l7t1dJekXu599faarYE8sLR8TPSAd2P1S+rMTpwNFKc7+8AEDSSyVdJmlShfITSP3oT+T+7NPL1rkN+BbpYO9+pCRP3ou/mjQ8dJ98UtN7BvXuqx/+Z1YVJ3VrqIh4hnRg8x5SsnySNH79ANLQuErOop+f2xHxc9KQzWnAXZI2A1eRDrZuqfCSL5HGsD9KGt63pkKZK3Kc34rdx1H/LWko5q9JX1ArKRkeWq/hf5Jel4fRmvVLu3fnmVktJC0FXhARA46CkTSfNGRx9pAHZi3De+pmNchdRn+Uzy4+ClgAXNPouKx1efSLWW0mkLpcDiINqftn0hBIs4Zw94uZWYG4+8XMrECc1M3MCmRI+tQPOOCAmDZt2lBUPShPPfUU++7rE81KuU16c5v05japrFnaZePGjY9GRMXpMoYkqU+bNo0NGzYMRdWDsm7dOubMmdPoMJqK26Q3t0lvbpPKmqVdJPV5FbCqul8kTVK6Ivw9SldYf039wjMzs3qpdk/9XNL0qKfkq6rsM9ALzMxs+A2Y1PMlyI4hTX5PRPyBdJVyMzNrMtV0v7yINBvdJZJulXTRIGbUMzOzYTTgyUeSZpImOXptRNwi6VzgyTw/dWm5haTrTDJ58uQZq1atGqKQq9fd3c348eMbHUZTcZv05jbpzW1SWbO0S3t7+8aImFlpWTVJ/QWkq9pPy49fByyKiD6vzjJz5szw6Jfm5DbpzW3Sm9uksmZpF0l9JvUBu18i4tfALyW9PD81l3QFcjMzazLVjn7pAC7PI1/uBf566EIyM7M9VVVSj4hNQMVdfTOzZnf4isPrV9lgLy1ewR3vvaP2SvrgqXfNrPC2dJ3NfWf3eRiwavXoU5+26Pqa4+iPJ/QyMysQJ3UzswJxUjczKxAndTOzAnFSNzMrECd1M7MCcVI3MysQJ3UzswLxyUdm1hLqdtLPmtrqmbj3mPrE0QcndTMrvHqcTQrpi6FedQ0Vd7+YmRWIk7qZWYE4qZuZFYiTuplZgTipm5kViJO6mVmBOKmbmRWIk7qZWYH45CMzM0BSdeWW9r88IuoQzZ7znrqZGSkZD3Tr7OwcsEyjOambmRWIk7qZWYE4qZuZFYiTuplZgTipm5kViJO6mVmBOKmbmRWIk7qZWYH4jFKzFlTt2ZMDaYaTbWx33lM3a0EDnRU59ZOrqzrD0pqPk7qZWYE4qZuZFYiTuplZgTipm5kViJO6mVmBOKmbmRVIVePUJd0HbAF2ANsjYuZQBmVmZntmMCcftUfEo0MWiZmZ1czdL2ZmBaJqzgqT9AvgcSCAr0bE1yqUWQgsBJg8efKMVatW1TnUwevu7mb8+PGNDqOpuE16c5v0Nn/NUyw/Yd9Gh9F0mmVbaW9v39hXN3i13S+zI+IhSc8HbpJ0T0T8oLRATvRfA5g5c2bMmTOnlpjrYt26dTRDHM3EbdKb26SCNde7TSoYCdtKVUk9Ih7Kfx+RdA1wFPCD/l9lZo1wxJk3snnrtprrmbbo+prrmLj3GG47/bia67HqDZjUJe0L7BURW/L944CzhjwyM9sjm7du476zT6qpjnrtkdbji8EGp5o99cnANXmqztHAFRGxZkijMjOzPTJgUo+Ie4EjhiEWMzOrkYc0mpkViJO6mVmBOKlby1q5ciXTp09n7ty5TJ8+nZUrVzY6JLOa+Rql1pJWrlzJ4sWLWbZsGTt27GDUqFEsWLAAgHnz5jU4OrM95z11a0lLlixh2bJltLe3M3r0aNrb21m2bBlLlixpdGhmNXFSt5bU1dXF7Nmzd3tu9uzZdHV1NSgis/pw94u1pLa2NtavX097e/vO59avX09bW1sDo6qPCW2LOHzFotorWlGPWABqOxHKBsdJ3VrS4sWLWbBgwc4+9c7OThYsWFCI7pctXWf7jNIW5qRuLWnevHksX76cuXPnEhFI4o1vfKMPktqI5z51a0kdHR2sXbuWc845hxtuuIFzzjmHtWvX0tHR0ejQzGripG4t6cILL2Tp0qWcdtppjBs3jtNOO42lS5dy4YUXNjo0s5q4+8Va0jPPPMOpp56623OnnnoqH/vYxxoUUX3VpS97TX2m3rXh5aRuLWns2LEsXLiQTZs20dXVRVtbG0ceeSRjx45tdGg1q/UgKaQvhXrUY8PP3S/Wko499lguv/xyjjnmGL7zne9wzDHHcPnll3Psscc2OjSzmnhP3VrSQw89xMknn8zFF1/MV77yFcaOHcvJJ5/MT3/600aHZlYTJ3VrSV1dXdx6662MGTNm55jsbdu2MW7cuEaHZlYTd79YS+o5o7RUUc4otdbmpG4tqeeM0s7OTrZv377zjNLFixc3OjSzmrj7xVpSz5mjHR0dO0e/LFmyxGeU2ojnpG4ta968ecybN69u85yYNQMndSs8SXWpJyLqUo/ZUHKfuhVeRPR7m/rJ1QOWKVpCl9Tv7f6lbxqwTL2+LK2+nNTNWtBAX2CdnZ0t90VXFE7qZmYF4qRuZlYgTupmZgXipG5mViBO6mZmBeKkbmZWIE7qZmYF4qRuZlYgTupmZgXipG5mViBO6mZmBeKkbmZWIE7qZmYFUnVSlzRK0q2SVg9lQGZmtucGs6f+YaBrqAIxM7PaVZXUJR0CnARcNLThmJlZLardU/8S8Ang2SGMxczMajTgNUolvQl4JCI2SprTT7mFwEKAyZMns27dunrFuMe6u7ubIo5m4japzG2yO28nlY2EdtFAl6SS9I/Au4HtwDjgucDVEfGuvl4zc+bM2LBhQz3j3CO+SnxvbpPepi26nvvOPqnRYTQVbyeVNUu7SNoYETMrLRuw+yUiPhURh0TENOAvgbX9JXQzM2ucAbtfzJrZEWfeyOat22quZ9qi62uuY+LeY7jt9ONqrsesFoNK6hGxDlg3JJGY7YHNW7fV3HVSr5/U9fhiMKuVzyg1MysQJ3UzswJxUjczKxAndTOzAnFSNzMrECd1M7MCcVI3MysQn3xkI9qEtkUcvmJR7RWtqEcskCYzNWscJ/WCkVSXegaaE6hZbOk62ycfmZVw90vBRES/t6mfXD1gmZGS0M2sNyd1M7MCcVI3MysQJ3UzswJxUjczKxCPfrERry6jTtbUZz51s0ZzUrcRrR6XofPl7KxI3P1iZlYgTupmZgXipG5mViBO6mZmBeKkbmZWIIVM6itXrmT69OnMnTuX6dOns3LlykaHZGY2LAo3pHHlypUsXryYZcuWsWPHDkaNGsWCBQsAmDdvXoOjMzMbWoXbU1+yZAnLli2jvb2d0aNH097ezrJly1iyZEmjQzMzG3KFS+pdXV3Mnj17t+dmz55NV1dXgyIyMxs+hUvqbW1trF+/frfn1q9fT1tbW4MiMjMbPoVL6osXL2bBggV0dnayfft2Ojs7WbBgAYsXL250aGZmQ65wB0p7DoZ2dHTQ1dVFW1sbS5Ys8UFSM2sJhUvqkBL7vHnz6nbtSTOzkaJw3S9mZq3MSd3MrECc1M3MCsRJ3cysQAp5oLSojjjzRjZv3VZzPfW4/NvEvcdw2+nH1VzPcJA0cJmlA9cTEXWIxmxoOamPIJu3bqv5smv1GhFUl+uCDpOBkrFHSVmRuPvFzKxAnNTNzApkwKQuaZykH0u6TdJdks4cjsDMzGzwqulTfwZ4fUR0SxoDrJd0Q0T8aIhjMzOzQRowqUc6ytSdH47JNw8DMDNrQlX1qUsaJWkT8AhwU0TcMrRhmZnZntBgxt5KmgRcA3RExJ1lyxYCCwEmT548Y9WqVfWMc490d3czfvz4RodRN/PXPMXyE/atqY56tUk9YmkWRdtO6sFtUlmztEt7e/vGiJhZcWFEDOoGfBb4eH9lZsyYEc2gs7Oz0SHU1dRPrq65jnq1ST1iaRZF207qwW1SWbO0C7Ah+si/A/apSzoQ2BYRT0jaG3gjUMX5d1ZvE9oWcfiKRbVXtKIesQDUdiKUmdVfNaNfXgiskDSK1Af/zYhYPbRhWSVbus72GaVm1q9qRr/cDvzxMMRiZmY18hmlZmYF4qRuZlYgTupmZgXipG5mViBO6mZmBeKLZIwwdRlKuKY+Vz4ys+bjpD6C1DpGHdKXQj3qMbPm5O4XM7MCKWRS7+joYNy4cbS3tzNu3Dg6OjoaHZKZ2bAoXPdLR0cH559//s7HzzzzzM7H5513XqPCMjMbFoXbU+9J4EcffTTf+ta3OProo3d73sysyAqX1AFmzZrFzTffzAEHHMDNN9/MrFmzGh2SmdmwKGRSP/744/t9bGZWVCO2T11Sn8vOOusszjrrrKpfE4O4+pOZWTMbsXvqfV31Y8qUKRXLT5kypb+rOZmZFcKITep9eeCBB3ol9ilTpvDAAw80KCIzs+FTuKQOKbFHBFM/uZqIcEI3s5ZRyKRuZtaqnNTNzArESd3MrECc1M3MCsRJ3cysQEbsyUdWWX8nZe0ss3Tgejx+32xk8p56wfR1glXPrbOzc8AyTuhmI5eTuplZgTipm5kViJO6mVmBOKmbmRWIk7qZWYE4qZuZFUjTjlM/4swb2bx1W831TFt0fU2vn7j3GG47/bia4zAzGw5Nm9Q3b93GfWefVFMd69atY86cOTXVUeuXgpnZcHL3i5lZgTipm5kViJO6mVmBOKmbmRWIk7qZWYEMmNQlTZHUKeluSXdJ+vBwBGZmZoNXzZDG7cDHIuJ/JE0ANkq6KSLuHuLYzMxskAZM6hHxMPBwvr9FUhdwMDCkSX1C2yIOX7Go9opW1BoHQG3j5c3MhosGc0EESdOAHwDTI+LJsmULgYUAkydPnrFq1aqaApu/5imWn7BvTXV0d3czfvz4hsfRTOrRJkXjNunNbVJZs7RLe3v7xoiYWXFhNVfByYl/PLAR+POBys6YMSNqNfWTq2uuo7OzsyniaCb1aJOicZv05japrFnaBdgQfeTfqqYJkDQGuAq4PCKurs93zcDqcor+mtrnfjEzGykGTOpKVzJeBnRFxBeHPqSk1nlfIH0p1KMeM7ORoppx6q8F3g28XtKmfDtxiOMyM7M9UM3ol/WAhiEWMzOrkc8oNTMrECd1M7MCcVI3MysQJ3UzswJxUjczKxAndTOzAnFSNzMrECd1M7MCcVI3MysQJ3UzswJxUjczKxAndTOzAqlqPvVmlGYErqLc0oHLxCCu/mRm1sxG7J56X1f9KL11dnZWe1UnM7NCGLFJ3czMenNSNzMrECd1M7MCcVI3MysQJ3UzswJxUjczKxAndTOzAnFSNzMrEA3FyTeSfgvcX/eKB+8A4NFGB9Fk3Ca9uU16c5tU1iztMjUiDqy0YEiSerOQtCEiZjY6jmbiNunNbdKb26SykdAu7n4xMysQJ3UzswIpelL/WqMDaEJuk97cJr25TSpr+nYpdJ+6mVmrKfqeuplZSxnxSV3SFEmdku6WdJekD1coI0lflvQzSbdLelUjYh0uksZJ+rGk23KbnFmhzFhJ38htcoukacMf6fCTNErSrZJWV1jWcm0i6T5Jd0jaJGlDheUt9dkBkDRJ0pWS7pHUJek1Zcubuk1GfFIHtgMfi4jDgFcDH5R0WFmZPwVelm8Lga8Mb4jD7hng9RFxBHAkcIKkV5eVWQA8HhEvBf4FqOIaUYXwYaCrj2Wt2ibtEXFkH0P1Wu2zA3AusCYiXgEcQe/tpanbZMQn9Yh4OCL+J9/fQvoHHFxW7C3ApZH8CJgk6YXDHOqwye+zOz8ck2/lB0/eAqzI968E5qraawSOUJIOAU4CLuqjSMu1SRVa6rMjaSJwDLAMICL+EBFPlBVr6jYZ8Um9VP65/MfALWWLDgZ+WfL4QXon/kLJ3QybgEeAmyKizzaJiO3AZmD/4Y1y2H0J+ATwbB/LW7FNArhR0kZJCyssb7XPzouA3wKX5G66iyTtW1amqdukMEld0njgKuAjEfFko+NptIjYERFHAocAR0ma3uiYGknSm4BHImJjo2NpMrMj4lWkLoUPSjqm0QE12GjgVcBXIuKPgaeARY0NaXAKkdQljSEl9Msj4uoKRR4CppQ8PiQ/V3j5p2MncELZop1tImk0MBF4bHijG1avBd4s6T5gFfB6SZeVlWm1NiEiHsp/HwGuAY4qK9Jqn50HgQdLftleSUrypZq6TUZ8Us99nsuAroj4Yh/FrgXek49avxrYHBEPD1uQw0zSgZIm5ft7A28E7ikrdi3w3nz/FGBtFPikhYj4VEQcEhHTgL8kvd93lRVrqTaRtK+kCT33geOAO8uKtdRnJyJ+DfxS0svzU3OBu8uKNXWbjG50AHXwWuDdwB25Dxng08ChABFxAfDvwInAz4Cngb9uQJzD6YXACkmjSF/c34yI1ZLOAjZExLWkL8KvS/oZ8DtSoms5Ld4mk4Fr8rHg0cAVEbFG0qnQsp8dgA7gcknPAe4F/noktYnPKDUzK5AR3/1iZma7OKmbmRWIk7qZWYE4qZuZFYiTuplZgTipW0NImi/poGFe5zRJd+b7MyV9eRjXPafSzJBlZd4sqeLZi5K6Kz1vVq4I49RtZJpPOtHlV41YeURsAHpNNdtIeaz8tY2Ow0Y276lb3Uh6V57HfZOkr+ZJxUZJWi7pzjxv90clnQLMJJ3gsSmf9dpXnWdIWiHph5Lul/Tnkv4p17UmTxGBpBmSvp8npvqPnlnz8vO3SboN+GBJvTv3nCUdJem/8gRO/9lzNmH+NXF1Xs9PJf1TFW2wXNIFkjZI+kmec6a8zH6Svq00F/ePJP1RyfrOz/dflGO6Q9LnSl57qaSTSx5fLuktA8VlrcNJ3epCUhvwDuC1eSKxHcA7SfO5HxwR0yPicOCSiLiStJf8zjyP99YBqn8J8HrgzcBlQGeuaytwUk7s5wGnRMQM4GJgSX7tJUBHnlu+L/cAr8sTOH0W+HzJsiPz+zoceIekKRVeX24aaQ6Vk4ALJI0rW34mcGtE/BHp7OdLK9RxLmlSqcOB0lPQl5F+5fRME3s0cH0VMVmLcPeL1ctcYAbw3/m0871J0/5eB7xY0nmk5HPjHtR9Q0Rsk3QHMApYk5+/g5RAXw5MB27K6x4FPJznv5kUET/I5b9Omo2w3ETStAovI01FO6Zk2ToeHdQAAAHaSURBVPciYjOApLuBqew+7Wol34yIZ4GfSroXeEXZ8tnAXwBExFpJ+0t6blmZ1/aUyXEvzeW/L+nfJB2Yl1+Vpwk2A5zUrX4ErIiIT/VaIB0BHA+cCrwdeN8g634GICKelbStZJKtZ0nbsIC7IqL8smOTqqz/H0h7/29VmpN/Xfm6sx1U95kpn3tjT+fi6Ot1lwLvIs1N01TzjljjufvF6uV7wCmSng87+42nSjoA2CsirgI+w65pTLcAE+q07v8FDlS+lqSkMZJemacdfkLS7FzunX28fiK7pk6dX4d43iZpL0kvAV6c4yv1w55YJM0BHq1wDYCb2TWhWHncy4GPAERE+QyC1uKc1K0ucnL5DOkqOrcDN5FmizwYWJdn0LwM6NmTX07qb94kaW9JZ0l68x6u+w+kqXKX5gOim0h9zZD2ZP81r7+vS9P9E/CPkm6lyl+vSlfEqXRNT4AHgB8DNwCnRsTvy5afAczI7XQ2u6b7LfVh0kUr7qDsqjoR8RvSZRsvqSZWay2epdGsjiQtB1bng8FDtY59SMcTXtXT32/Ww3vqZiOIpDeQ9tLPc0K3SrynbmZWIN5TNzMrECd1M7MCcVI3MysQJ3UzswJxUjczKxAndTOzAvn/7O3WrHDkkLwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# PICNIC average ploidy vs estimated copy-neutral \n",
    "tmp = df_ploidies.boxplot(column=\"PICNIC avg. pl.\", by = \"est. median. ploidy\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert gene-level integer CN into log2R-like format in order to make it compatible with TCGA and CCLE\n",
    "1 Copy-neutral state was defined from average ploiy, as median of integer CN values in non-disrupted genes.\n",
    "\n",
    "2 Compute log2(CN/neutral-CN) for min and max CN; keep the value with most extreme estimate\n",
    "\n",
    "3 Replace estimates below thresholds with zeroes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "estimated_CN = est_ploidies[\"median_pl\"].to_dict()\n",
    "estimated_CN[1287381]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1287381</th>\n",
       "      <th>924100</th>\n",
       "      <th>910924</th>\n",
       "      <th>687561</th>\n",
       "      <th>1287706</th>\n",
       "      <th>687452</th>\n",
       "      <th>906798</th>\n",
       "      <th>906797</th>\n",
       "      <th>906800</th>\n",
       "      <th>910922</th>\n",
       "      <th>...</th>\n",
       "      <th>909785</th>\n",
       "      <th>909904</th>\n",
       "      <th>909905</th>\n",
       "      <th>687592</th>\n",
       "      <th>1303911</th>\n",
       "      <th>946358</th>\n",
       "      <th>909907</th>\n",
       "      <th>1298146</th>\n",
       "      <th>908452</th>\n",
       "      <th>908450</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OR4G11P</th>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 996 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               1287381  924100   910924    687561   1287706  687452   906798   \\\n",
       "OR4G11P       0.736966      0.0      0.0  0.321928      0.0      0.0      0.0   \n",
       "OR4F5         0.736966      0.0      0.0  0.321928      0.0      0.0      0.0   \n",
       "RP11-34P13.7  0.736966      0.0      0.0  0.321928      0.0      0.0      0.0   \n",
       "\n",
       "              906797   906800   910922   ...  909785    909904   909905   \\\n",
       "OR4G11P           0.0      0.0     -1.0  ...      0.0  0.415037      0.0   \n",
       "OR4F5             0.0      0.0     -1.0  ...      0.0  0.415037      0.0   \n",
       "RP11-34P13.7      0.0      0.0     -1.0  ...      0.0  0.415037      0.0   \n",
       "\n",
       "              687592    1303911   946358   909907    1298146   908452   \\\n",
       "OR4G11P           0.0 -0.584963  0.321928     -1.0 -0.584963  0.584963   \n",
       "OR4F5             0.0 -0.584963  0.321928     -1.0 -0.584963  0.584963   \n",
       "RP11-34P13.7      0.0 -0.584963  0.321928     -1.0 -0.584963  0.584963   \n",
       "\n",
       "              908450   \n",
       "OR4G11P           0.0  \n",
       "OR4F5             0.0  \n",
       "RP11-34P13.7      0.0  \n",
       "\n",
       "[3 rows x 996 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc = gdsc.apply(lambda x : CN2log2R(x,estimated_CN[x.name] ))\n",
    "# drop genes without any determined value\n",
    "gdsc = gdsc.dropna(axis=0,how=\"all\")\n",
    "# fill with zeroes the remaining ones\n",
    "gdsc.fillna(0,inplace=True)\n",
    "gdsc.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdsc = gdsc.applymap(lambda x :  clean_logR(x, pos_seg_mean_thr, neg_seg_mean_thr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Symbol rows are not empty.\n",
      "Ok: All Symbol are mapped to GeneID\n",
      "9 Symbol mapped to multiple GeneID\n",
      "Ok: All GeneID are unique\n",
      "60207 Symbol can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "NCBI = pd.read_csv(\"Homo_sapiens.gene_info\",sep = \"\\t\")\n",
    "NCBI = NCBI[[\"#tax_id\",\"GeneID\",\"Symbol\",\"Synonyms\",\"type_of_gene\"]]\n",
    "NCBI = NCBI.loc[NCBI[\"#tax_id\"] == 9606]\n",
    "NCBI = NCBI.loc[NCBI[\"type_of_gene\"] != \"unknown\"]\n",
    "ncbi_symbols = parse_mapping_table(NCBI, \"Symbol\",\"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Synonyms rows are not empty.\n",
      "Ok: All Synonyms are mapped to GeneID\n",
      "3261 Synonyms mapped to multiple GeneID\n",
      "50720 different Synonyms mapped to the same GeneID\n",
      "10915 Synonyms can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "ncbi_synonyms = expand(NCBI[[\"Synonyms\",\"GeneID\"]],column=\"Synonyms\",sep=\"|\") \n",
    "ncbi_synonyms = parse_mapping_table(ncbi_synonyms, \"Synonyms\",\"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped: 24521 \n",
      "\tdirectly via main_mapper 22038 \n",
      "\tvia alternative mapper 826 \n",
      "\tvia one of multiple synonyms in alternative mapper 1657 \n",
      "\tLOC 0 \n",
      "Unmapped: 21611 \n",
      "\trecognized symbols without Entrez ID 0 \n",
      "\tmultiple query_ids map to the same target_id 0 \n",
      "\tquery_ids map to multiple target_ids in the main mapper 0 \n",
      "\tquery_ids map to multiple target_ids in the alternative mapper 94 \n",
      "\tLOC not found in Entrez 0 \n",
      "\tNot found at all: 21517\n",
      "Warning: query IDs mapping to duplicated target IDs in mapping table: 171\n",
      "Warning: query IDs not mapped to any target IDs excluded: 21611\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IDs mapped to multiple target IDs are kept:\n",
      " [143872, 414212, 414213, 284697, 341019, 51236, 414761, 442416, 728113, 26165, 6218, 221262, 100132948, 768096, 6241, 642658, 401508, 100289124, 728695, 125050, 387707, 84631, 4253, 643749, 54438, 3255, 100034743, 2749, 653505, 26824, 23254, 728798, 100462820, 119016, 26871, 286464, 51463, 642826, 653067, 100132115, 9502, 10529, 23334, 440107, 256815, 645425, 647507, 728917, 84316, 200030, 729438, 100302179, 106478953, 246126, 100033392, 92017, 374, 445815, 79741, 399761, 647060, 284565, 6551, 161176, 643479, 644509, 100418977, 29099, 10160, 653234, 644019, 114112, 79817, 463, 677844, 26583, 23523, 84458, 1517, 574445, 26095, 653308]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1287381</th>\n",
       "      <th>924100</th>\n",
       "      <th>910924</th>\n",
       "      <th>687561</th>\n",
       "      <th>1287706</th>\n",
       "      <th>687452</th>\n",
       "      <th>906798</th>\n",
       "      <th>906797</th>\n",
       "      <th>906800</th>\n",
       "      <th>910922</th>\n",
       "      <th>...</th>\n",
       "      <th>909785</th>\n",
       "      <th>909904</th>\n",
       "      <th>909905</th>\n",
       "      <th>687592</th>\n",
       "      <th>1303911</th>\n",
       "      <th>946358</th>\n",
       "      <th>909907</th>\n",
       "      <th>1298146</th>\n",
       "      <th>908452</th>\n",
       "      <th>908450</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.736966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 996 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    1287381   924100   910924    687561    1287706   687452    906798   \\\n",
       "1  0.000000  0.000000      0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "2  0.000000  0.584963      0.0 -0.415037  0.321928  0.000000  0.584963   \n",
       "9 -0.584963  0.584963      0.0 -0.415037 -0.415037 -0.584963  0.000000   \n",
       "\n",
       "    906797   906800   910922   ...  909785    909904    909905    687592   \\\n",
       "1  0.584963      0.0      0.0  ...      0.0 -0.584963  0.584963  0.415037   \n",
       "2  0.584963      0.0      0.0  ...      0.0 -0.584963  0.000000  0.000000   \n",
       "9  0.000000      0.0      0.0  ...      0.0 -0.584963  0.000000  0.415037   \n",
       "\n",
       "    1303911   946358   909907    1298146   908452    908450   \n",
       "1  0.000000  0.000000      0.0  0.000000  0.584963  0.415037  \n",
       "2  0.736966  0.321928      0.0 -0.584963  0.000000  0.736966  \n",
       "9  0.000000  0.000000     -1.0 -0.584963  0.000000  0.000000  \n",
       "\n",
       "[3 rows x 996 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc,query2target,not_mapped  = apply_mappers(gdsc, ncbi_symbols, ncbi_synonyms, verbose = True,handle_duplicates = \"keep\")\n",
    "gdsc.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "82 duplicated IDs in 171 rows found.\n",
      "duplicate rows removed due to low correlation of duplicated profiles 25\n",
      "Merged  146 duplicated rows into 70 rows\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>683665</th>\n",
       "      <th>683667</th>\n",
       "      <th>684052</th>\n",
       "      <th>684055</th>\n",
       "      <th>684057</th>\n",
       "      <th>684059</th>\n",
       "      <th>684062</th>\n",
       "      <th>684072</th>\n",
       "      <th>684681</th>\n",
       "      <th>687448</th>\n",
       "      <th>...</th>\n",
       "      <th>1659818</th>\n",
       "      <th>1659819</th>\n",
       "      <th>1659823</th>\n",
       "      <th>1659928</th>\n",
       "      <th>1659929</th>\n",
       "      <th>1660034</th>\n",
       "      <th>1660035</th>\n",
       "      <th>1660036</th>\n",
       "      <th>1674021</th>\n",
       "      <th>1789883</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.321928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>-0.415037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 996 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         683665    683667   684052    684055    684057   684059    684062   \\\n",
       "gene_id                                                                      \n",
       "1            0.0  0.000000      0.0  0.000000 -0.415037      0.0 -0.415037   \n",
       "2            0.0  0.000000      0.0  0.584963  0.000000      0.0  0.000000   \n",
       "9            0.0  0.321928      0.0  0.584963  0.584963      0.0  0.321928   \n",
       "10           0.0  0.321928      0.0  0.584963  0.584963      0.0  0.321928   \n",
       "12           0.0 -1.000000      0.0  0.000000 -1.000000      0.0 -0.415037   \n",
       "\n",
       "          684072    684681   687448   ...  1659818  1659819   1659823  \\\n",
       "gene_id                               ...                               \n",
       "1        0.000000  0.415037      0.0  ...      0.0      0.0  0.000000   \n",
       "2        0.584963  0.000000      0.0  ...      0.0      0.0  0.000000   \n",
       "9        0.000000 -0.584963      0.0  ...      0.0      0.0 -0.415037   \n",
       "10       0.000000 -0.584963      0.0  ...      0.0      0.0 -0.415037   \n",
       "12       0.000000  0.000000      0.0  ...     -1.0      0.0  0.321928   \n",
       "\n",
       "         1659928   1659929  1660034   1660035   1660036   1674021   1789883  \n",
       "gene_id                                                                      \n",
       "1            0.0 -0.415037      0.0 -0.584963  0.000000  0.000000  0.321928  \n",
       "2            0.0  0.000000      0.0  0.000000  0.000000  0.584963 -0.415037  \n",
       "9            0.0  0.000000     -1.0 -1.584963  0.000000 -1.000000 -1.000000  \n",
       "10           0.0  0.000000     -1.0 -1.584963  0.000000 -1.000000 -1.000000  \n",
       "12           0.0 -0.415037      0.0  0.415037  0.584963  0.000000  0.000000  \n",
       "\n",
       "[5 rows x 996 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc = handle_dups(gdsc,corr_thr = 0.75)\n",
    "gdsc.index.name = \"gene_id\"\n",
    "gdsc = gdsc.T.sort_index().T\n",
    "gdsc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdsc.to_csv(preprocessed_dir+\"/GDSC\"+\".Segment_Mean.CNA.tsv\",\n",
    "                 sep = \"\\t\",header=True,index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PDX\n",
    "For PDX dataset only gene-level estimated copy-number (non-integer) reported. From ploidy distributions, calculated as average over all genes we concluded that CN estimates were called under assumption that copy-neutral state of each xenograft corresponds CN = 2.\n",
    "\n",
    "For gene ID conversion we used the same approach as for RNA-seq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23852, 375)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>2.58</td>\n",
       "      <td>1.60</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.00</td>\n",
       "      <td>3.94</td>\n",
       "      <td>2.04</td>\n",
       "      <td>11.39</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.01</td>\n",
       "      <td>...</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.14</td>\n",
       "      <td>2.95</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.99</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.43</td>\n",
       "      <td>2.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>2.58</td>\n",
       "      <td>1.60</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.00</td>\n",
       "      <td>3.94</td>\n",
       "      <td>2.04</td>\n",
       "      <td>11.39</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.01</td>\n",
       "      <td>...</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.14</td>\n",
       "      <td>2.95</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.99</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.43</td>\n",
       "      <td>2.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>2.87</td>\n",
       "      <td>2.97</td>\n",
       "      <td>2.01</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.10</td>\n",
       "      <td>1.58</td>\n",
       "      <td>2.01</td>\n",
       "      <td>1.64</td>\n",
       "      <td>1.89</td>\n",
       "      <td>1.99</td>\n",
       "      <td>...</td>\n",
       "      <td>2.04</td>\n",
       "      <td>0.97</td>\n",
       "      <td>1.58</td>\n",
       "      <td>2.08</td>\n",
       "      <td>1.95</td>\n",
       "      <td>1.92</td>\n",
       "      <td>1.54</td>\n",
       "      <td>1.28</td>\n",
       "      <td>1.33</td>\n",
       "      <td>2.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2LD1</th>\n",
       "      <td>5.74</td>\n",
       "      <td>1.64</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.01</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.74</td>\n",
       "      <td>2.06</td>\n",
       "      <td>1.59</td>\n",
       "      <td>1.40</td>\n",
       "      <td>2.53</td>\n",
       "      <td>...</td>\n",
       "      <td>2.03</td>\n",
       "      <td>2.07</td>\n",
       "      <td>2.25</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.01</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.08</td>\n",
       "      <td>1.85</td>\n",
       "      <td>1.93</td>\n",
       "      <td>1.45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          X-1004  X-1008  X-1027  X-1095  X-1119  X-1156  X-1167  X-1169  \\\n",
       "Sample                                                                     \n",
       "A1BG        2.58    1.60    2.17    2.08    2.00    3.94    2.04   11.39   \n",
       "A1BG-AS1    2.58    1.60    2.17    2.08    2.00    3.94    2.04   11.39   \n",
       "A1CF        2.87    2.97    2.01    2.06    2.10    1.58    2.01    1.64   \n",
       "A2LD1       5.74    1.64    2.06    2.01    2.07    1.74    2.06    1.59   \n",
       "\n",
       "          X-1172  X-1173  ...  X-5694  X-5696  X-5713  X-5717  X-5727  X-5739  \\\n",
       "Sample                    ...                                                   \n",
       "A1BG        2.17    2.01  ...    2.08    2.10    2.14    2.95    2.06    2.07   \n",
       "A1BG-AS1    2.17    2.01  ...    2.08    2.10    2.14    2.95    2.06    2.07   \n",
       "A1CF        1.89    1.99  ...    2.04    0.97    1.58    2.08    1.95    1.92   \n",
       "A2LD1       1.40    2.53  ...    2.03    2.07    2.25    2.00    1.01    2.00   \n",
       "\n",
       "          X-5808  X-5959  X-5975  X-6047  \n",
       "Sample                                    \n",
       "A1BG        1.99    2.07    1.43    2.03  \n",
       "A1BG-AS1    1.99    2.07    1.43    2.03  \n",
       "A1CF        1.54    1.28    1.33    2.10  \n",
       "A2LD1       1.08    1.85    1.93    1.45  \n",
       "\n",
       "[4 rows x 375 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PDX_xls = \"PDX/nm.3954-S2.xlsx\"\n",
    "pdx = pd.read_excel(PDX_xls,\"copy number\")\n",
    "pdx.set_index(\"Sample\",drop=True,inplace=True)\n",
    "focal  = pdx.T[\"FocalCNScore\"]\n",
    "pdx.drop([\"ArmLevelCNScore\",\"FocalCNScore\"],inplace = True)\n",
    "print(pdx.shape)\n",
    "pdx.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Strings containing duplicated gene IDs: 544\n",
      "268 duplicated IDs in 544 rows found.\n",
      "duplicate rows removed due to low correlation of duplicated profiles 134\n",
      "Merged  410 duplicated rows into 205 rows\n"
     ]
    }
   ],
   "source": [
    "pdx.index.name = \"gene_id\"\n",
    "ids = pdx.index\n",
    "ids = list(set(ids[ids.duplicated()]))\n",
    "print(\"Strings containing duplicated gene IDs:\",pdx.loc[ids,:].shape[0])\n",
    "pdx = handle_dups(pdx,corr_thr = 0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, 'CN Averaged over all')"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAbMElEQVR4nO3de5wcZZ3v8c+Xm1wChJwM2QgZBxBFFAk4ArvCLqAgl8MBVmENd5Y90VUQ3JUlclaBRTHsS2FlcT0bBLmIXETuFwG5I4gkEHIBohGCXAIJQoCAoCG//aOeMe3QPVOTTFVn5vm+X69+pbrq6apf12vy7erqp55SRGBmZvlYpd0FmJlZvRz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfBbJSTdKekfSrTrlLRY0qotlp8s6YfLsf0+12vvtLz72oYeB78tN0nzJP0+BewLks6XNGIg64iI30bEiIh4ezBrW5H1StpZ0tL0vl6TNEfSkWlZl6RIy3re9/WSdmt4/Yi0bw5umLeupN9K+vTgvEOz5efgtxW1T0SMALYFuoF/bXM9g+W59L7WA04AzpG0ZcPykWn51sCtwFWSjgCIiMXAZ4H/kNSR2v87MDUirqjrDZi14uC3QRERzwI3AR/qvUzSKpL+VdJTkhZIulDS+mlZzxH0aun5JpLuSkfatwKjG9Zzg6Rjeq17hqT9m2yz93rvlHSqpJ+ndd8iaXTv1zV5XxERVwMvA1s2Wf58RHwHOBk4XdIqaf7NwA3AWZJ2Bg4EPt9qO5KOkPREqu3Jnm8LkjaTdLuk30l6UdLFkkY2vG6epOPTfnhd0rmSxki6Ka3rZ5I26LVPJkp6TtJ8SV/uo6YdJN0naZGkR9L76LNeGxoc/DYoJI0D9gIebrL4iPTYBdgUGAGc3WJVPwKmUQT+qcDhDcsuAA5p2ObWwEYUAVvGQcCRwIbAGkDL0GvYxirpg2UkMLOPplem9b6/Yd6XgJ2BK4AvR8TzLbaxDnAWsGdErAv8FTC9ZzHwTeDdwAeAcRQfMo0+BewGvA/Yh+ID+ESgg+L/+Bd7td8F2BzYHThB0iea1NSzX78OjKLYVz+R1NFPvTYEOPhtRV0taRFwL3AXcFqTNgcDZ0TEE+k0yFeAz/QcjfeQ1Al8FPhqRLwVEXcD1zU0uRZ4n6TN0/NDgcsi4g8la/1BRPwqIn4PXA6M76Ptu9P7ehE4CTg0Iub00f659O+onhkR8TIwG1ib4oOhL0uBD0laKyLmR8TstI65EXFr2h8LgTOAv+n12v+MiBfSt657gAci4uGIeBO4CtimV/tTIuL1iJgJ/ACY0KSeQ4AbI+LGiFgaEbcCUyk+3FvWa0ODg99W1H4RMTIi3hMRn0+h2tu7gacanj8FrAaMadLu5Yh4vVdbAFKQXQYckk6pTAAuGkCtjUfcb1B882jlufS+RkXE+Ii4tJ91b5T+falnhqRDgC7gZ8DprV6Y3u/fAZ8D5qdTWlukdYyRdKmkZyW9CvyQhtNfyQsN079v8rz3+3y6Yfopiv3e23uAA9JpnkXpQ3BHYGxf9drQ4OC3OjxHESQ9OoEl/HlAAcwHNkinEhrbNrqA4hvEx4E3IuL+Qa51ee0PLADmAEjaEDgT+L8UP/QeKGmnVi+OiJsjYjdgLPA4cE5adBoQwFYRsR7FkbhWsNZxDdOdLPu20uhp4KL04dfzWCciJvdTrw0BDn6rwyXAl9IPtyMowuyyiFjS2CginqI4nXCKpDUk7Uhxzrqxzf0Upxm+zcCO9iuRjsiPpjgd9JWIWJoWnQ1cHRF3RMR84F8oega9q8U69k0feG8BiyneI8C66fkr6bz78YNQ9lclrS3pgxS/eVzWpM0PgX0kfVLSqpLWVNHNdeN+6rUhwMFvdTiPIqTvBp4E3gSOadH2IGB7ilMmJwEXNmlzIbAVRTi1yyJJr1P84LsXcEBEnAcgaT+K0yJ/CumI+D7FkfXXmqxrFeCf0vKXKM7h/2NadgpFV9lXKH5s7e+3gjLuAuYCtwHfiohbejeIiKeBfSl+JF5I8Q3g+FRrX/XaECDfiMWGGkmHARMjYsd21zKUSOqi+OBdvfe3LcuLj/htSJG0NkV/+CntrsVsqHLw25Ah6ZMUpx1eoOjvb2bLwad6zMwy4yN+M7PMrNZ/k/YbPXp0dHV1tbsMM7MhZdq0aS9GREfv+UMi+Lu6upg6dWq7yzAzG1IkPdVsvk/1mJllxsFvZpYZB7+ZWWYc/GZmmXHwm5llxsFvZpaZyoI/DeP6y3SvztmSTknzz0/36JyeHn3dBcnMzAZZlf343wJ2jYjFklYH7pV0U1p2fERcUeG2zcyshcqCP4pBgBanp6unhwcGMjNrs0qv3JW0KjANeC/w3Yh4QNI/At+Q9DWKG0FMioi3mrx2IjARoLOz9933bKjrmnRDqXbzJu9dcSVm+an0x92IeDsixgMbA9tJ+hDwFWAL4KPAKOCEFq+dEhHdEdHd0fGOoSbMzGw51dKrJyIWAXcAe0TE/Ci8BfwA2K6OGszMrFBlr54OSSPT9FrAbsDjksameQL2A2ZVVYOZmb1Tlef4xwIXpPP8qwCXR8T1km6X1AEImA58rsIazMyslyp79cwAtmkyf9eqtmlmZv3zlbtmZplx8JuZZcbBb2aWGQe/mVlmHPxmZplx8JuZZcbBb2aWGQe/mVlmHPxmZplx8JuZZcbBb2aWGQe/mVlmHPxmZplx8JuZZcbBb2aWGQe/mVlmHPxmZpmp8taLlpmuSTe0uwQzK8FH/GZmmXHwm5llprLgl7SmpF9KekTSbEmnpPmbSHpA0lxJl0lao6oazMzsnao84n8L2DUitgbGA3tI2gE4HTgzIt4LvAwcVWENZmbWS2XBH4XF6enq6RHArsAVaf4FwH5V1WBmZu9Uaa8eSasC04D3At8FfgMsioglqckzwEYtXjsRmAjQ2dlZZZm2EivbU2je5L0rrsRs+Kj0x92IeDsixgMbA9sBWwzgtVMiojsiujs6Oiqr0cwsN7X06omIRcAdwF8CIyX1fNPYGHi2jhrMzKxQZa+eDkkj0/RawG7AYxQfAJ9OzQ4HrqmqBjMze6cqz/GPBS5I5/lXAS6PiOslPQpcKunrwMPAuRXWYGZmvVQW/BExA9imyfwnKM73m5lZG3isHuvXUBiDx71/zMrzkA1mZplx8JuZZcbBb2aWGQe/mVlmHPxmZplxr55haCj0wjGz9vERv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlhkHv5lZZhz8ZmaZcfCbmWXGY/VYVnynLjMf8ZuZZcfBb2aWmcqCX9I4SXdIelTSbEnHpvknS3pW0vT02KuqGszM7J2qPMe/BPjniHhI0rrANEm3pmVnRsS3Kty2mZm1UFnwR8R8YH6afk3SY8BGVW3PzMzKqeUcv6QuYBvggTTraEkzJJ0naYMWr5koaaqkqQsXLqyjTDOzLFQe/JJGAD8BjouIV4HvAZsB4ym+EXy72esiYkpEdEdEd0dHR9Vlmpllo9Lgl7Q6RehfHBFXAkTECxHxdkQsBc4BtquyBjMz+3NV9uoRcC7wWESc0TB/bEOz/YFZVdVgZmbvVGWvno8BhwIzJU1P804EJkgaDwQwD/hshTWYmVkvVfbquRdQk0U3VrVNMzPrn6/cNTPLjIPfzCwzDn4zs8w4+M3MMuPgNzPLjIPfzCwzDn4zs8w4+M3MMuPgNzPLTL/BL2kzSe9K0ztL+qKkkdWXZmZmVShzxP8T4G1J7wWmAOOAH1ValZmZVaZM8C+NiCUUI2n+Z0QcD4zt5zVmZraSKjNI2x8lTQAOB/ZJ81avriSz9uuadEOpdvMm711xJWaDr8wR/5HAXwLfiIgnJW0CXFRtWWZmVpV+j/gj4lFJJwCd6fmTwOlVF2ZmZtUo06tnH2A68NP0fLyka6suzMzMqlHmVM/JFPfFXQQQEdOBTSusyczMKlQm+P8YEa/0mre0imLMzKx6ZXr1zJZ0ELCqpM2BLwL3VVuWmZlVpcwR/zHAB4G3gEuAV4HjqizKzMyqU6ZXzxvA/0uP0iSNAy4ExgABTImI70gaBVwGdAHzgAMj4uWBlW1mZsurZfBLuo4isJuKiP/Tz7qXAP8cEQ9JWheYJulW4AjgtoiYLGkSMAk4YcCVm5nZcunriP9bK7LiiJgPzE/Tr0l6DNgI2BfYOTW7ALgTB7+ZWW1aBn9E3NUzLWkNYAuKbwBzIuIPA9mIpC5gG+ABYEz6UAB4nuJUULPXTAQmAnR2dg5kc2Zm1ocyF3DtDfwGOAs4G5grac+yG5A0gmKEz+Mi4tXGZRERtDidFBFTIqI7Iro7OjrKbs7MzPpRpjvnt4FdImIuFOPzAzcAN/X3QkmrU4T+xRFxZZr9gqSxETFf0lhgwfKVbmZmy6NMd87XekI/eQJ4rb8XSRJwLvBYRJzRsOhaipE+Sf9eU7JWMzMbBGWO+KdKuhG4nOK0zAHAg5L+FqDhSL63jwGHAjMlTU/zTgQmA5dLOgp4CjhwBeo3M7MBKhP8awIvAH+Tni8E1qIYmz+ApsEfEfcCarHOjw+sTDMzGyxlLuA6so5CzMysHv0Gf7rxyjEUV9r+qX2JC7jMzGwlVOZUz9UUP9Jeh0flNDMb8soE/5sRcVbllZiZWS3KBP93JJ0E3EIxQicAEfFQZVWZmVllygT/VhTdMndl2ameSM/NzGyIKRP8BwCbDnR8HjMzWzmVuXJ3FjCy6kLMzKweZY74RwKPS3qQPz/H7+6cZmZDUJngP6nyKszMrDZlrty9q782ZmY2dJQZj38HSQ9KWizpD5LelvRqf68zM7OVU5kfd88GJgC/phic7R+A71ZZlJmZVadM8JPG4181It6OiB8Ae1RblpmZVaXMj7tvpHvuTpf07xQ3UC/1gWFmZiufMgF+aGp3NPA6MA74VJVFmZlZdcr06nkqTb4p6SxgXK9bMZqZ2RBSplfPnZLWkzQKeAg4R9IZ/b3OzMxWTmVO9awfEa8CfwtcGBHbA5+otiwzM6tKmeBfTdJYipuiX19xPWZmVrEywf9vwM3A3Ih4UNKmFH36zcxsCOo3+CPixxHx4Yj4fHr+RET026tH0nmSFkia1TDvZEnPSpqeHnutWPlmZjZQVfbHP5/mF3qdGRHj0+PGCrdvZmZNVBb8EXE38FJV6zczs+XTjitwj5Y0I50K2qBVI0kTJU2VNHXhwoV11mdmNqz1ewGXpHdRXKnb1dg+Iv5tObb3PeBUinv2ngp8G/j7Zg0jYgowBaC7uzuWY1tmZtZEmbF6rgFeAabRcAeu5RERL/RMSzoHdw81M6tdmeDfOCIGZTROSWMjYn56uj/F/XzNzKxGZYL/PklbRcTMgaxY0iXAzsBoSc9Q3MJxZ0njKU71zAM+O7ByzcxsRZUJ/h2BIyQ9SXGqR0BExIf7elFETGgy+9yBl2hmZoOpTPDvWXkVZmZWm4EMy2xmZsOA76RlZpYZB7+ZWWYc/GZmmXHwm5llxsFvZpYZB7+ZWWYc/GZmmXHwm5llxsFvZpYZB7+ZWWbKjNVjy6lr0g2l2s2bvHfFlZiZLeMjfjOzzDj4zcwy4+A3M8uMg9/MLDMOfjOzzLhXj9kKaGfPLfcas+XlI34zs8xUFvySzpO0QNKshnmjJN0q6dfp3w2q2r6ZmTVX5RH/+cAeveZNAm6LiM2B29JzMzOrUWXBHxF3Ay/1mr0vcEGavgDYr6rtm5lZc3Wf4x8TEfPT9PPAmJq3b2aWvbb16omIkBStlkuaCEwE6OzsrK2u/pTtSdHudZqZtVL3Ef8LksYCpH8XtGoYEVMiojsiujs6Omor0MxsuKs7+K8FDk/ThwPX1Lx9M7PsVdmd8xLgfuD9kp6RdBQwGdhN0q+BT6TnZmZWo8rO8UfEhBaLPl7VNs3MrH++ctfMLDMeq8esBgPpueWxdaxqPuI3M8uMg9/MLDMOfjOzzDj4zcwy4+A3M8uMg9/MLDMOfjOzzDj4zcwy4+A3M8uMg9/MLDMessFsJTPYN+Ypu76yQ0UM9vqsfj7iNzPLjIPfzCwzDn4zs8w4+M3MMuPgNzPLjIPfzCwzDn4zs8w4+M3MMtOWC7gkzQNeA94GlkREdzvqMDPLUTuv3N0lIl5s4/bNzLLkUz1mZplpV/AHcIukaZImNmsgaaKkqZKmLly4sObyzMyGr3YF/44RsS2wJ/AFSX/du0FETImI7ojo7ujoqL9CM7Nhqi3BHxHPpn8XAFcB27WjDjOzHNUe/JLWkbRuzzSwOzCr7jrMzHLVjl49Y4CrJPVs/0cR8dM21GFmlqXagz8ingC2rnu7ZmZW8B24zKwSA7mTmO/WVS/34zczy4yD38wsMw5+M7PMOPjNzDLj4Dczy4x79SQD6YFgNhwNp/8DZd9Lrr2JfMRvZpYZB7+ZWWYc/GZmmXHwm5llxsFvZpYZRUS7a+hXd3d3TJ06dbleO5x6KphZewzV3j+SpkVEd+/5PuI3M8uMg9/MLDMOfjOzzDj4zcwy4+A3M8uMx+oxMxskVfQirKJHkY/4zcwy4+A3M8tMW4Jf0h6S5kiaK2lSO2owM8tV7cEvaVXgu8CewJbABElb1l2HmVmu2nHEvx0wNyKeiIg/AJcC+7ahDjOzLLWjV89GwNMNz58Btu/dSNJEYGJ6uljSnBpqWxGjgRfbXcRKwPthGe+LZYb0vtDpg7aqAe+HFdz2e5rNXGm7c0bEFGBKu+soS9LUZoMh5cb7YRnvi2W8Lwory35ox6meZ4FxDc83TvPMzKwG7Qj+B4HNJW0iaQ3gM8C1bajDzCxLtZ/qiYglko4GbgZWBc6LiNl111GBIXNaqmLeD8t4XyzjfVFYKfbDkLgRi5mZDR5fuWtmlhkHv5lZZhz8AyDpPEkLJM1qsXx9SddJekTSbElH1l1jHSSNk3SHpEfT+zy2SRtJOisNyzFD0rbtqLVqJffFwWkfzJR0n6St21Frlcrsh4a2H5W0RNKn66yxLmX3haSdJU1Pbe6qtciI8KPkA/hrYFtgVovlJwKnp+kO4CVgjXbXXcF+GAtsm6bXBX4FbNmrzV7ATYCAHYAH2l13G/fFXwEbpOk9h+O+KLMf0rJVgduBG4FPt7vuNv5NjAQeBTrT8w3rrNFH/AMQEXdThHnLJsC6kgSMSG2X1FFbnSJifkQ8lKZfAx6juCK70b7AhVH4BTBS0tiaS61cmX0REfdFxMvp6S8orl0ZVkr+TQAcA/wEWFBjebUquS8OAq6MiN+mdrXuDwf/4Dob+ADwHDATODYilra3pGpJ6gK2AR7otajZ0BzNgmDY6GNfNDqK4pvQsNVqP0jaCNgf+F79VbVHH38T7wM2kHSnpGmSDquzrpV2yIYh6pPAdGBXYDPgVkn3RMSr7S2rGpJGUBy9HTdc32NZZfaFpF0ogn/HOmurUz/74T+AEyJiafGleHjrZ1+sBnwE+DiwFnC/pF9ExK/qqM3BP7iOBCZHcdJurqQngS2AX7a3rMEnaXWKP+qLI+LKJk2yGZqjxL5A0oeB7wN7RsTv6qyvLiX2QzdwaQr90cBekpZExNU1llmLEvviGeB3EfE68Lqku4GtKX4PqJxP9Qyu31J8giNpDPB+4Im2VlSB9BvGucBjEXFGi2bXAoel3j07AK9ExPzaiqxJmX0hqRO4Eji0riO6upXZDxGxSUR0RUQXcAXw+WEa+mX+f1wD7ChpNUlrU4xQ/FhdNfqIfwAkXQLsDIyW9AxwErA6QET8f+BU4HxJMyl6s5wQEUN2KNo+fAw4FJgpaXqadyLQCX/aFzdS9OyZC7xB8W1oOCqzL74G/C/gv9LR7pJYCUZoHGRl9kMu+t0XEfGYpJ8CM4ClwPcjomk38Sp4yAYzs8z4VI+ZWWYc/GZmmXHwm5llxsFvZpYZB7+ZWWYc/FY7SX8h6VJJv0mXq98o6X2SuiSFpGMa2p4t6Yg+1jVd0qW1FL6CJC1udw29STq/Z5TMNHzAcOtmak04+K1W6eKWq4A7I2KziPgI8BVgTGqyADg23Y+5v3V9gGK0x50krTMItQ2L61rSRXP+v20t+Y/D6rYL8MfGC3oi4pGIuCc9XQjcBhxeYl0TgIuAWyhGA0XSLyR9sKdBz1GspHVU3E/hl5IeltTT/ghJ10q6HbhN0ghJt0l6KI2fv2/Dur4qaY6keyVdIunLaf5mkn6avr3cI2mLNH8TSfen9Xy91ZuQ9E+SZqXHcWneZElfaGhzcsP2jpf0oIox/k9J87pSbRcCs/jz4TKQ9LX0mlmSpiiHwXKstXaPXe1HXg/gi8CZLZZ1UYTWpsAciqP5s4EjWrSfQ3E15O7AdWnel4BT0vRYYE6aPg04JE2PpBgTZR3gCIpxU0alZasB66Xp0RRXHgv4KMUAfGtSjLH+a+DLqd1twOZpenvg9jR9LXBYmv4CsLjJe/gIxUiu61AM5T2bYjTHbYC7Gto9ShHmu1PcsFsUB27XU9wnooviCtAdWuyrUQ3TFwH7pOnzSePiA3cC3e3+G/Gj+oeP+G2lExFPUAxje1CrNulc9ItRjGd+G7CNpFHA5UDPnZ0OpBgTBorAnJQuob+TIsA707JbI6LnPgsCTpM0A/gZxVDSYyguw78mIt6MYoz161IdIyhutPLjtO7/pvjAIb3mkjR9UYu3siNwVUS8HhGLKcb02SkiHgY2lPRuFXfsejkink7vY3fgYeAhikEAN0/reiqKex80s4ukB9JwIrsCH2zRzjIwLM5p2pAym2XB3JfTKEK71S3pJgBbSJqXnq8HfCoizpH0uzQa5t8Bn0vLlZbPaVyJpO2B1xtmHUxx97SPRMQf0/rX7KPOVYBFETG+xfIVGRPlxxT76i+Ay3pKBr4ZEf/d2FDFuO+N76Nx2ZrAf1EczT8t6WT6fk82zPmI3+p2O/AuSRN7Zkj6sKSdGhtFxOMUpzf26b2C9MPlgcBWsWy0x30pPgygCMl/AdaPiBlp3s3AMT3ntiVt06K+9YEFKfR3Ad6T5v8c2EfSmuko/3+nOl8FnpR0QFqvtOyeuj8HPpOmD26xvXuA/SStnX6g3j/N63kfn6EI/x83vI+/TzUgaSNJG7ZYd4+ekH8xvW5Y3uvWynPwW60iIijC7ROpO+ds4JvA802af4PmtyncCXg2Ip5rmHc3sKWK2zteQRGYlzcsP5ViJNUZaZuntijxYqA7nRI5DHg81f0gxTn7GRR30JoJvJJeczBwlKRHKL7R9PwgfCzwhbSupncfi+IWfedT3LPhAYpRGh9Oy2ZT/J7wbKQhrSPiFuBHFDfumJne67ot3kvPNhYB51D8fnIz8GBf7W348+icZiVJGhERi1WMn343MDEFt9mQ4nP8ZuVNkbQlxamTCxz6NlT5iN/MLDM+x29mlhkHv5lZZhz8ZmaZcfCbmWXGwW9mlpn/AWddyz9LZF0wAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "average_ploidies = pdx.apply(np.mean)\n",
    "p = plt.hist(average_ploidies,bins=30)\n",
    "plt.title(\"Ploidy in PDX samples\")\n",
    "plt.ylabel(\"n samples\")\n",
    "plt.xlabel(\"CN Averaged over all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.007232</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.021480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.007232</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.021480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.521051</td>\n",
       "      <td>0.570463</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>-0.286304</td>\n",
       "      <td>-0.081614</td>\n",
       "      <td>-0.007232</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>-1.043943</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>-0.036526</td>\n",
       "      <td>-0.058894</td>\n",
       "      <td>-0.377070</td>\n",
       "      <td>-0.643856</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.070389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            X-1004    X-1008    X-1027    X-1095    X-1119    X-1156  \\\n",
       "gene_id                                                                \n",
       "A1BG      0.367371 -0.321928  0.117695  0.056584  0.000000  0.978196   \n",
       "A1BG-AS1  0.367371 -0.321928  0.117695  0.056584  0.000000  0.978196   \n",
       "A1CF      0.521051  0.570463  0.007196  0.042644  0.070389 -0.340075   \n",
       "\n",
       "            X-1167    X-1169    X-1172    X-1173  ...    X-5694    X-5696  \\\n",
       "gene_id                                           ...                       \n",
       "A1BG      0.028569  2.509696  0.117695  0.007196  ...  0.056584  0.070389   \n",
       "A1BG-AS1  0.028569  2.509696  0.117695  0.007196  ...  0.056584  0.070389   \n",
       "A1CF      0.007196 -0.286304 -0.081614 -0.007232  ...  0.028569 -1.043943   \n",
       "\n",
       "            X-5713    X-5717    X-5727    X-5739    X-5808    X-5959  \\\n",
       "gene_id                                                                \n",
       "A1BG      0.097611  0.560715  0.042644  0.049631 -0.007232  0.049631   \n",
       "A1BG-AS1  0.097611  0.560715  0.042644  0.049631 -0.007232  0.049631   \n",
       "A1CF     -0.340075  0.056584 -0.036526 -0.058894 -0.377070 -0.643856   \n",
       "\n",
       "            X-5975    X-6047  \n",
       "gene_id                       \n",
       "A1BG     -0.483985  0.021480  \n",
       "A1BG-AS1 -0.483985  0.021480  \n",
       "A1CF     -0.588574  0.070389  \n",
       "\n",
       "[3 rows x 375 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdx = pdx.applymap(lambda x: np.log2(x/2))\n",
    "pdx.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.521051</td>\n",
       "      <td>0.570463</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.286304</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.043943</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.37707</td>\n",
       "      <td>-0.643856</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            X-1004    X-1008  X-1027  X-1095  X-1119    X-1156  X-1167  \\\n",
       "gene_id                                                                  \n",
       "A1BG      0.367371 -0.321928     0.0     0.0     0.0  0.978196     0.0   \n",
       "A1BG-AS1  0.367371 -0.321928     0.0     0.0     0.0  0.978196     0.0   \n",
       "A1CF      0.521051  0.570463     0.0     0.0     0.0 -0.340075     0.0   \n",
       "\n",
       "            X-1169  X-1172  X-1173  ...  X-5694    X-5696    X-5713    X-5717  \\\n",
       "gene_id                             ...                                         \n",
       "A1BG      2.509696     0.0     0.0  ...     0.0  0.000000  0.000000  0.560715   \n",
       "A1BG-AS1  2.509696     0.0     0.0  ...     0.0  0.000000  0.000000  0.560715   \n",
       "A1CF     -0.286304     0.0     0.0  ...     0.0 -1.043943 -0.340075  0.000000   \n",
       "\n",
       "          X-5727  X-5739   X-5808    X-5959    X-5975  X-6047  \n",
       "gene_id                                                        \n",
       "A1BG         0.0     0.0  0.00000  0.000000 -0.483985     0.0  \n",
       "A1BG-AS1     0.0     0.0  0.00000  0.000000 -0.483985     0.0  \n",
       "A1CF         0.0     0.0 -0.37707 -0.643856 -0.588574     0.0  \n",
       "\n",
       "[3 rows x 375 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdx = pdx.applymap(lambda x :  clean_logR(x, pos_seg_mean_thr, neg_seg_mean_thr))\n",
    "pdx.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped: 23282 \n",
      "\tdirectly via main_mapper 20801 \n",
      "\tvia alternative mapper 525 \n",
      "\tvia one of multiple synonyms in alternative mapper 1178 \n",
      "\tLOC 778 \n",
      "Unmapped: 231 \n",
      "\trecognized symbols without Entrez ID 0 \n",
      "\tmultiple query_ids map to the same target_id 0 \n",
      "\tquery_ids map to multiple target_ids in the main mapper 0 \n",
      "\tquery_ids map to multiple target_ids in the alternative mapper 69 \n",
      "\tLOC not found in Entrez 35 \n",
      "\tNot found at all: 127\n",
      "Warning: query IDs mapping to duplicated target IDs in mapping table: 99\n",
      "Warning: query IDs not mapped to any target IDs excluded: 231\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IDs mapped to multiple target IDs are kept:\n",
      " [143872, 286464, 51463, 642826, 653067, 399761, 647060, 284565, 84631, 161176, 284697, 341019, 83869, 4253, 9502, 83871, 10529, 5414, 51754, 440107, 645425, 728113, 26165, 445815, 114112, 728695, 6218, 100132948, 100134869, 84316, 200030, 729438, 6241, 642658, 100302179, 401508, 286183, 84968, 119016, 84458, 26220, 574445, 26095, 80759, 3192, 387707, 79741]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.761285</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.700440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.201634</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.327687</td>\n",
       "      <td>-0.494109</td>\n",
       "      <td>-0.535332</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.761285</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.700440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.201634</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.327687</td>\n",
       "      <td>-0.494109</td>\n",
       "      <td>-0.535332</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           X-1004    X-1008  X-1027    X-1095  X-1119    X-1156  X-1167  \\\n",
       "gene_id                                                                   \n",
       "1        0.367371 -0.321928     0.0  0.000000     0.0  0.978196     0.0   \n",
       "2        0.761285  0.000000     0.0  0.500802     0.0  0.700440     0.0   \n",
       "3        0.761285  0.000000     0.0  0.500802     0.0  0.700440     0.0   \n",
       "\n",
       "           X-1169    X-1172  X-1173  ...  X-5694  X-5696  X-5713    X-5717  \\\n",
       "gene_id                              ...                                     \n",
       "1        2.509696  0.000000     0.0  ...     0.0     0.0     0.0  0.560715   \n",
       "2        0.000000  0.201634     0.0  ...     0.0     0.0     0.0  0.739848   \n",
       "3        0.000000  0.201634     0.0  ...     0.0     0.0     0.0  0.739848   \n",
       "\n",
       "         X-5727    X-5739    X-5808    X-5959    X-5975  X-6047  \n",
       "gene_id                                                          \n",
       "1           0.0  0.000000  0.000000  0.000000 -0.483985     0.0  \n",
       "2           0.0  0.739848  0.327687 -0.494109 -0.535332     0.0  \n",
       "3           0.0  0.739848  0.327687 -0.494109 -0.535332     0.0  \n",
       "\n",
       "[3 rows x 375 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdx,query2target,not_mapped = apply_mappers(pdx, ncbi_symbols, ncbi_synonyms, verbose = True,handle_duplicates = \"keep\")\n",
    "pdx.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143872</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.330973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.350497</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.349235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.514573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143872</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.330973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.350497</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.349235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.514573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286464</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.378512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.524915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.902389</td>\n",
       "      <td>-1.321928</td>\n",
       "      <td>0.367371</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>0.718088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.494109</td>\n",
       "      <td>-1.494109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286464</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.378512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.524915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.902389</td>\n",
       "      <td>-1.321928</td>\n",
       "      <td>0.367371</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>0.718088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.494109</td>\n",
       "      <td>-1.494109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286464</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.378512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.524915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.902389</td>\n",
       "      <td>-1.321928</td>\n",
       "      <td>0.367371</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>0.718088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.494109</td>\n",
       "      <td>-1.494109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3192</th>\n",
       "      <td>1.350497</td>\n",
       "      <td>0.632268</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.220330</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.448901</td>\n",
       "      <td>-0.268817</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.480265</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.207893</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.330973</td>\n",
       "      <td>0.250962</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.021480</td>\n",
       "      <td>0.599318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387707</th>\n",
       "      <td>0.389567</td>\n",
       "      <td>-0.666576</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>-0.473931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.386468</td>\n",
       "      <td>-0.610433</td>\n",
       "      <td>-0.577767</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387707</th>\n",
       "      <td>0.389567</td>\n",
       "      <td>-0.666576</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>-0.473931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.386468</td>\n",
       "      <td>-0.610433</td>\n",
       "      <td>-0.577767</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79741</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.589763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.242977</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.207893</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.220330</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.798366</td>\n",
       "      <td>-0.985645</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.70044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.459432</td>\n",
       "      <td>0.269033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79741</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.589763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.242977</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.207893</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.798366</td>\n",
       "      <td>-0.985645</td>\n",
       "      <td>-0.358454</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.70044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.459432</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           X-1004    X-1008  X-1027    X-1095  X-1119    X-1156  X-1167  \\\n",
       "gene_id                                                                   \n",
       "143872   0.000000  0.560715     0.0  0.000000     0.0 -0.330973     0.0   \n",
       "143872   0.000000  0.560715     0.0  0.000000     0.0 -0.330973     0.0   \n",
       "286464   0.000000  0.378512     0.0  0.550901     0.0 -0.524915     0.0   \n",
       "286464   0.000000  0.378512     0.0  0.550901     0.0 -0.524915     0.0   \n",
       "286464   0.000000  0.378512     0.0  0.550901     0.0 -0.524915     0.0   \n",
       "...           ...       ...     ...       ...     ...       ...     ...   \n",
       "3192     1.350497  0.632268     0.0  0.000000     0.0  0.220330     0.0   \n",
       "387707   0.389567 -0.666576     0.0  0.000000     0.0 -0.340075     0.0   \n",
       "387707   0.389567 -0.666576     0.0  0.000000     0.0 -0.340075     0.0   \n",
       "79741    0.000000  0.589763     0.0  0.000000     0.0 -0.242977     0.0   \n",
       "79741    0.000000  0.589763     0.0  0.000000     0.0 -0.242977     0.0   \n",
       "\n",
       "           X-1169    X-1172    X-1173  ...    X-5694    X-5696    X-5713  \\\n",
       "gene_id                                ...                                 \n",
       "143872  -0.367732  0.350497  0.000000  ...  0.000000 -0.349235  0.000000   \n",
       "143872  -0.367732  0.350497  0.000000  ...  0.000000 -0.349235  0.000000   \n",
       "286464  -0.902389 -1.321928  0.367371  ...  0.000000 -1.014500  0.718088   \n",
       "286464  -0.902389 -1.321928  0.367371  ...  0.000000 -1.014500  0.718088   \n",
       "286464  -0.902389 -1.321928  0.367371  ...  0.000000 -1.014500  0.718088   \n",
       "...           ...       ...       ...  ...       ...       ...       ...   \n",
       "3192     0.448901 -0.268817  0.000000  ...  0.480265  0.000000  0.207893   \n",
       "387707  -0.367732  0.000000  0.000000  ...  0.000000 -1.014500 -0.473931   \n",
       "387707  -0.367732  0.000000  0.000000  ...  0.000000 -1.014500 -0.473931   \n",
       "79741    0.207893  0.000000  0.220330  ... -0.798366 -0.985645  0.000000   \n",
       "79741    0.207893  0.000000  0.000000  ... -0.798366 -0.985645 -0.358454   \n",
       "\n",
       "           X-5717   X-5727    X-5739    X-5808    X-5959    X-5975    X-6047  \n",
       "gene_id                                                                       \n",
       "143872  -0.321928 -1.01450 -0.588574  0.000000  0.000000  0.000000 -0.514573  \n",
       "143872  -0.321928 -1.01450 -0.588574  0.000000  0.000000  0.000000 -0.514573  \n",
       "286464   0.000000 -1.01450  0.000000  0.000000  0.000000 -1.494109 -1.494109  \n",
       "286464   0.000000 -1.01450  0.000000  0.000000  0.000000 -1.494109 -1.494109  \n",
       "286464   0.000000 -1.01450  0.000000  0.000000  0.000000 -1.494109 -1.494109  \n",
       "...           ...      ...       ...       ...       ...       ...       ...  \n",
       "3192     0.000000  0.00000 -0.330973  0.250962  0.000000  1.021480  0.599318  \n",
       "387707   0.000000  0.00000  0.000000 -0.386468 -0.610433 -0.577767  0.000000  \n",
       "387707   0.000000  0.00000  0.000000 -0.386468 -0.610433 -0.577767  0.000000  \n",
       "79741    0.000000  0.70044  0.000000 -0.367732 -0.588574  0.459432  0.269033  \n",
       "79741    0.000000  0.70044  0.000000 -0.367732 -0.588574  0.459432  0.000000  \n",
       "\n",
       "[99 rows x 375 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dups = list(set(pdx[pdx.index.duplicated(keep=False)].index.values))\n",
    "pdx.loc[dups,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47 duplicated IDs in 99 rows found.\n",
      "duplicate rows removed due to low correlation of duplicated profiles 4\n",
      "Merged  95 duplicated rows into 45 rows\n"
     ]
    }
   ],
   "source": [
    "# most of these dupliates correspond to genes merged in the current assembly, e.g. gene - gene-AS\n",
    "pdx = handle_dups(pdx,corr_thr = 0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdx = pdx.T.sort_index().T\n",
    "pdx.to_csv(preprocessed_dir+\"/PDX\"+\".Segment_Mean.CNA.tsv\",\n",
    "                 sep = \"\\t\",header=True,index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluation of the results\n",
    "1. How many common genes between four datasets?\n",
    "\n",
    "2. Do CNA profiles of the same cell line from GDSC and CCLE correlate?\n",
    "\n",
    "3. Do CNA profiles of the same cancer type from TCGA and PDX look similar?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 1). How many common genes between four datasets?\n",
    "# we take BRCA from TCGA because  \n",
    "tcga = pd.read_csv(preprocessed_dir +\"TCGA-BRCA-FFPE.Segment_Mean.CNA.tsv\",sep = \"\\t\", index_col=0)\n",
    "#print(tcga.head(3))\n",
    "gdsc = pd.read_csv(preprocessed_dir+ \"/GDSC\"+\".Segment_Mean.CNA.tsv\",sep = \"\\t\", index_col=0)\n",
    "#print(tcga.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'CCLE')"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIEAAAE/CAYAAADRztNjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAcaElEQVR4nO3de5BmZ10n8O+PmQQkROKS9rKZZCe1RssUuKJdEYmryMUKhEqs8pa4eEHWqa01LizssmG1ENlyDbC4sEW8TCEigmTDTadIMGAZi/WSbCbhsiQBK8aRzICmw82AaIz+9o9+Y3Wazsybme4+5+3z+VS9lfec9/T7fjsn03nm2895TnV3AAAAANjZHjF0AAAAAAC2nhIIAAAAYAKUQAAAAAAToAQCAAAAmAAlEAAAAMAEKIEAAAAAJkAJBAAAADABSiDgYamqS6rqxqr6QlXdPXv+72vVG6vqvqq6d/b4SFX9QlU9ds3Xn1xVr66qw1X1+ao6VFWvWfcZP1RVB2evf7Kq3lNV37793y0AwLgcbZxUVV9XVW+rqnuq6nNV9eGqemFV7aqqvVXVVbV7g/d8WVX9/ew9H3h8dvu/O2CrKYGAuVXVi5K8Nsmrknx1kq9K8u+SnJ/k5Nlhr+zuU5MsJXlukicl+aOqOmX2+kuSLCc5L8mpSZ6S5JY1n/HCJK9J8t9n739Wkl9KcvEWfmsAAKN3tHFSVf3LJDcmuSvJE7r7sUm+P6vjrlPnePv/3d2PWfM4bUu+CWBQ1d1DZwAWwGw2zyeS/Eh3v+MhjnljksPd/TNr9p2a5E+T/Hx3v66q3p3k97r7NRt8/WOTHEny3O5+2xZ8GwAAC+lY46SqenOSr+juCx/i6/cm+fMkJ3X3/etee1mSr+3u52xybGBkzAQC5vVtSR6Z5Hcezhd1971J3pfkX8923ZDkhbNLyJ5QVbXuMx6V5F2bkBcAYCc51jjp6Unevn1xgEWkBALmdXqSe9b+5qiq/riqPltVX6yq7zjK134iyT+bPf+FJK9I8m+SHExypKp+dPba49Z/BgAASY49Tnpckk+ewPv/wGxc98Dj+hN4L2CklEDAvD6V5PS1iwl295Nn14t/Kkf/eXJGkk/PvuYfuvvK7j4/yWlJfj7JG6rqGzb6DAAAkhx7nPSpJF9zAu9/dXeftubxXSfwXsBIKYGAef1Jkr/Lw1yguaoek9Xpyf9n/Wvd/cXuvjLJZ5Kcu+YzvueE0wIA7CzHGif9XpLv3b44wCJSAgFz6e7PJvm5JL9UVd9XVadW1SOq6puSnLL++Kp6ZFV9S5LfzmrJ8+uz/S+oqqdU1ZdV1e7ZpWCnJvlAd38uyUuTXFlV31NVj66qk6rqmVX1yu36XgEAxmaOcdLPJnlyVb2qqr46Sarqa6vqzVW19k5fj6yqR615+DshTIg/8MDcuvuVSV6Y5MVJ/mr2+NUk/yXJH88Oe3FV3ZvVKclvSnJzkid39xdmr/9Nklcn+csk9yT5ySTf2913zj7j1bPP+JkkK1m9zellWS2TAAAm62jjpO7+s6wuHr03ya1V9bkk78jqGoz3rnmbzyf54prHU2f7f7CqPr/u8ZXb8G0B28gt4gEAAAAmwEwgAAAAgAlQAgEAAABMgBIIAAAAYAKUQAAAAAAToAQCAAAAmIDdQ33w6aef3nv37h3q4wGALXbzzTff091LQ+fgwYzBAGBnO9oYbLASaO/evTl48OBQHw8AbLGq+ouhM/CljMEAYGc72hjM5WAAAAAAE6AEAgAAAJgAJRAAAADABCiBAAAAACZACQQAAAAwAUogAAAAgAlQAgEAAABMgBIIAAAAYAKUQAAAAAAToAQCAAAAmAAlEAAAAMAE7B46AMBm2Hv5NQ/aPnTFhQMlAQCYhvXjr8QYDMbOTCAAAACACVACAQAAAEyAEggAAABgApRAAAAAABOgBAIAAACYACUQAAAAwAQogQAAAAAmQAkEAAAAMAFKIAAAAIAJUAIBAAAATMAxS6CqekNV3V1VH3mI16uq/ldV3VFVH66qb978mAAAAACciHlmAr0xyQVHef2ZSc6ZPfYl+eUTjwUAAADAZjpmCdTd70/y6aMccnGSN/WqG5KcVlVfs1kBAQB4aFX1H6vq1qr6SFW9taoeNXQmAGCcNmNNoDOS3LVm+/BsHwAAW6iqzkjyH5Isd/fjk+xKcsmwqQCAsdrWhaGral9VHayqgysrK9v50QAAO9XuJF9WVbuTPDrJJwbOAwCM1GaUQEeSnLlme89s35fo7v3dvdzdy0tLS5vw0QAA09XdR5L8jyQfT/LJJJ/r7veuP84v4gCAZHNKoANJfmR2l7AnZXXw8clNeF8AAI6iqr4iq+sznp3knyc5paqes/44v4gDAJLV6cNHVVVvTfKUJKdX1eEkP5vkpCTp7l9Jcm2SZyW5I8nfJHnuVoUFAOBBnp7kz7t7JUmq6p1JnpzkzYOmAgBG6ZglUHdfeozXO8lPbloiAADm9fEkT6qqRyf5YpKnJTk4bCQAYKy2dWFoAAA2T3ffmOTtSW5J8v+yOrbbP2goAGC0jjkTCACA8erun83q5foAAEdlJhAAAADABCiBAAAAACZACQQAAAAwAUogAAAAgAlQAgEAAABMgBIIAAAAYAKUQAAAAAAToAQCAAAAmAAlEAAAAMAEKIEAAAAAJkAJBAAAADABSiAAAACACVACAQAAAEyAEggAAABgApRAAAAAABOgBAIAAACYACUQAAAAwAQogQAAAAAmQAkEAAAAMAFKIAAAAIAJUAIBAAAATIASCABgQVXV11fVB9c8/rqqXjB0LgBgnHYPHQAAgOPT3R9L8k1JUlW7khxJ8q5BQwEAo2UmEADAzvC0JH/W3X8xdBAAYJyUQAAAO8MlSd46dAgAYLyUQAAAC66qTk5yUZK3PcTr+6rqYFUdXFlZ2d5wAMBoKIEAABbfM5Pc0t1/tdGL3b2/u5e7e3lpaWmbowEAY6EEAgBYfJfGpWAAwDEogQAAFlhVnZLkGUneOXQWAGDc3CIeAGCBdfcXkjxu6BwAwPiZCQQAAAAwAUogAAAAgAlQAgEAAABMgBIIAAAAYAKUQAAAAAAToAQCAAAAmAAlEAAAAMAEKIEAAAAAJmCuEqiqLqiqj1XVHVV1+Qavn1VV11fVB6rqw1X1rM2PCgAAAMDxOmYJVFW7klyZ5JlJzk1yaVWdu+6wn0lydXc/McklSX5ps4MCAAAAcPzmmQl0XpI7uvvO7r4vyVVJLl53TCf58tnzxyb5xOZFBAAAAOBE7Z7jmDOS3LVm+3CSb113zMuSvLeqfirJKUmevinpAAAAANgUm7Uw9KVJ3tjde5I8K8lvVtWXvHdV7auqg1V1cGVlZZM+GgAAAIBjmacEOpLkzDXbe2b71npekquTpLv/JMmjkpy+/o26e393L3f38tLS0vElBgAAAOBhm6cEuinJOVV1dlWdnNWFnw+sO+bjSZ6WJFX1DVktgUz1AQAAABiJY5ZA3X1/ksuSXJfk9qzeBezWqnp5VV00O+xFSX6iqj6U5K1Jfqy7e6tCAwAAAPDwzLMwdLr72iTXrtv30jXPb0ty/uZGAwAAAGCzbNbC0AAAAACMmBIIAAAAYAKUQAAAAAAToAQCAAAAmAAlEADAAquq06rq7VX10aq6vaq+behMAMA4zXV3MAAARuu1SX63u7+vqk5O8uihAwEA46QEAgBYUFX12CTfkeTHkqS770ty35CZAIDxcjkYAMDiOjvJSpJfr6oPVNXrq+qUoUMBAOOkBAIAWFy7k3xzkl/u7icm+UKSy9cfVFX7qupgVR1cWVnZ7owAwEgogQAAFtfhJIe7+8bZ9tuzWgo9SHfv7+7l7l5eWlra1oAAwHgogQAAFlR3/2WSu6rq62e7npbktgEjAQAjZmFoAIDF9lNJ3jK7M9idSZ47cB4AYKSUQAAAC6y7P5hkeegcAMD4uRwMAAAAYALMBAJ2pL2XX/Ml+w5dceEASQAAAMbBTCAAAACACVACAQAAAEyAEggAAABgApRAAAAAABOgBAIAAACYACUQAAAAwAQogQAAAAAmQAkEAAAAMAFKIAAAAIAJUAIBAAAATIASCAAAAGAClEAAAAAAE6AEAgAAAJgAJRAAAADABCiBAAAAACZACQQAAAAwAUogAAAAgAlQAgEAAABMgBIIAAAAYAJ2Dx0AAIDjV1WHktyb5B+S3N/dy8MmAgDGSgkEALD4vqu77xk6BAAwbi4HAwAAAJgAJRAAwGLrJO+tqpurat/QYQCA8XI5GADAYvv27j5SVV+Z5H1V9dHufv/aA2bl0L4kOeuss4bICACMgJlAAAALrLuPzP55d5J3JTlvg2P2d/dydy8vLS1td0QAYCTmKoGq6oKq+lhV3VFVlz/EMT9QVbdV1a1V9VubGxMAgPWq6pSqOvWB50m+O8lHhk0FAIzVMS8Hq6pdSa5M8owkh5PcVFUHuvu2Nceck+QlSc7v7s/MpiMDALC1virJu6oqWR3X/VZ3/+6wkQCAsZpnTaDzktzR3XcmSVVdleTiJLetOeYnklzZ3Z9J/mk6MgAAW2g2PvtXQ+cAABbDPJeDnZHkrjXbh2f71vq6JF9XVX9UVTdU1QWbFRAAAACAE7dZdwfbneScJE9JsifJ+6vqCd392bUHuTMFAAAAwDDmmQl0JMmZa7b3zPatdTjJge7+++7+8yR/mtVS6EHcmQIAAABgGPOUQDclOaeqzq6qk5NckuTAumN+O6uzgFJVp2f18rA7NzEnAAAAACfgmCVQd9+f5LIk1yW5PcnV3X1rVb28qi6aHXZdkk9V1W1Jrk/yn7v7U1sVGgAAAICHZ641gbr72iTXrtv30jXPO8kLZw+ALbX38muGjgAAALBw5rkcDAAAAIAFpwQCAAAAmAAlEAAAAMAEzLUmEMBOsNFaQoeuuHCAJAAAANvPTCAAAACACVACAQAAAEyAEggAAABgApRAAAAAABOgBAIAAACYACUQAAAAwAQogQAAAAAmQAkEAAAAMAFKIAAAAIAJUAIBAAAATIASCAAAAGAClEAAAAAAE6AEAgBYcFW1q6o+UFXvHjoLADBeSiAAgMX3/CS3Dx0CABg3JRAAwAKrqj1JLkzy+qGzAADjpgQCAFhsr0ny4iT/OHQQAGDclEAAAAuqqp6d5O7uvvkYx+2rqoNVdXBlZWWb0gEAY6MEAgBYXOcnuaiqDiW5KslTq+rN6w/q7v3dvdzdy0tLS9udEQAYCSUQAMCC6u6XdPee7t6b5JIkv9/dzxk4FgAwUkogAAAAgAnYPXQAAABOXHf/QZI/GDgGADBiZgIBAAAATIASCAAAAGAClEAAAAAAE6AEAgAAAJgAC0MDk7b38msetH3oigsHSgIAALC1zAQCAAAAmAAlEAAAAMAEKIEAAAAAJkAJBAAAADABSiAAAACACVACAQAAAEyAEggAAABgApRAAAAAABOgBAIAAACYgLlKoKq6oKo+VlV3VNXlRznue6uqq2p58yICAAAAcKJ2H+uAqtqV5Mokz0hyOMlNVXWgu29bd9ypSZ6f5MatCApM197Lrxk6AgAAwMKbZybQeUnu6O47u/u+JFcluXiD4/5bklck+dtNzAcAAADAJpinBDojyV1rtg/P9v2TqvrmJGd2t1/XAwAAAIzQCS8MXVWPSPKLSV40x7H7qupgVR1cWVk50Y8GAAAAYE7zlEBHkpy5ZnvPbN8DTk3y+CR/UFWHkjwpyYGNFofu7v3dvdzdy0tLS8efGgAAAICHZZ4S6KYk51TV2VV1cpJLkhx44MXu/lx3n97de7t7b5IbklzU3Qe3JDEAAAAAD9sxS6Duvj/JZUmuS3J7kqu7+9aqenlVXbTVAQEA2FhVPaqq/m9Vfaiqbq2qnxs6EwAwXse8RXySdPe1Sa5dt++lD3HsU048FsAwNrod/aErLhwgCcBc/i7JU7v781V1UpI/rKr3dPcNQwcDAMZnrhIIAIDx6e5O8vnZ5kmzRw+XCAAYsxO+OxgAAMOpql1V9cEkdyd5X3ffOHQmAGCclEAAAAusu/+hu78pq3dwPa+qHr/+mKraV1UHq+rgysrK9ocEAEbB5WDAqGy0Jg8Ax9bdn62q65NckOQj617bn2R/kiwvL7tcDAAmykwgAIAFVVVLVXXa7PmXJXlGko8OmwoAGCszgQAAFtfXJPmNqtqV1V/uXd3d7x44EwAwUkogAIAF1d0fTvLEoXMAAIvB5WAAAAAAE2AmEMAxrF+s+tAVFw6UBAAA4PiZCQQAAAAwAUogAAAAgAlQAgEAAABMgBIIAAAAYAKUQAAAAAAToAQCAAAAmAAlEAAAAMAEKIEAAAAAJmD30AGA6dp7+TVDRzguG+U+dMWFAyQBAACYn5lAAAAAABOgBAIAAACYACUQAAAAwAQogQAAAAAmQAkEAAAAMAFKIAAAAIAJUAIBAAAATIASCAAAAGAClEAAAAAAE6AEAgAAAJgAJRAAwIKqqjOr6vqquq2qbq2q5w+dCQAYr91DBwCmY+/l1wwdYcts9L0duuLCAZIAE3N/khd19y1VdWqSm6vqfd1929DBAIDxMRMIAGBBdfcnu/uW2fN7k9ye5IxhUwEAY6UEAgDYAapqb5InJrlx2CQAwFgpgQAAFlxVPSbJO5K8oLv/eoPX91XVwao6uLKysv0BAYBRUAIBACywqjopqwXQW7r7nRsd0937u3u5u5eXlpa2NyAAMBpKIACABVVVleTXktze3b84dB4AYNyUQAAAi+v8JD+c5KlV9cHZ41lDhwIAxskt4oEtsZNvBw8wFt39h0lq6BwAwGIwEwgAAABgApRAAAAAABOgBAIAAACYgLnWBKqqC5K8NsmuJK/v7ivWvf7CJP82yf1JVpL8eHf/xSZnBUbK+j8bW//v5dAVFw6UBAAAYI6ZQFW1K8mVSZ6Z5Nwkl1bVuesO+0CS5e7+xiRvT/LKzQ4KAAAAwPGb53Kw85Lc0d13dvd9Sa5KcvHaA7r7+u7+m9nmDUn2bG5MAAAAAE7EPJeDnZHkrjXbh5N861GOf16S95xIKGDcXP4FAACweOZaE2heVfWcJMtJvvMhXt+XZF+SnHXWWZv50QAAAAAcxTyXgx1Jcuaa7T2zfQ9SVU9P8tNJLuruv9vojbp7f3cvd/fy0tLS8eQFAAAA4DjMUwLdlOScqjq7qk5OckmSA2sPqKonJvnVrBZAd29+TAAAAABOxDEvB+vu+6vqsiTXZfUW8W/o7lur6uVJDnb3gSSvSvKYJG+rqiT5eHdftIW5gW1i/R8AAICdYa41gbr72iTXrtv30jXPn77JuYCBKH0AAAB2pk1dGBqAh7ZRwXboigsHSAIAAEyREggmzKwfAACA6ZhnYWgAAAAAFpwSCAAAAGACXA4GE+HSLwAAgGkzEwgAAABgAswEAhjQ+hla7hYGAABsFTOBAAAAACbATCDYoawBBAAAwFpmAgEAAABMgJlAsAOY9QMwXVX1hiTPTnJ3dz9+6DwAwHgpgQBGZKNCz2LRwDG8Mcnrkrxp4BwAwMgpgWABmfkDwAO6+/1VtXfoHADA+FkTCABgh6uqfVV1sKoOrqysDB0HABiImUAwcmb9AHCiunt/kv1Jsry83APHATguLpuHE2cmEAAAAMAEmAkEI2LWDxvxWy8AAGAzmAkEALDAquqtSf4kyddX1eGqet7QmQCAcTITCAZk5g8AJ6q7Lx06AwCwGJRAsE0UPgAAAAxJCQSwgNaXitYIAgAAjsWaQAAAAAATYCYQbBGXfwEAADAmZgIBAAAATICZQLAJzPphaBv9N2idIABgI8YNMF1mAgEAAABMgJlAcAxm+bCo/JYPAABYy0wgAAAAgAkwE4jJMsOHKVr/372ZQQAAMB1KICZD6QMAwBi5hBvYLkogdiSFD8zHoBMAAKZDCcTCUfAAAADAw6cEYlQUPDA86wYBAMDOpARi2yh4YDG5ZAwAAHYGJRBbQuEDO5tiCAAAFo8SiIdNwQMAAACLRwnEUSl8gHlZSwgAWBTGLUyVEmjCFDzAVnLJGAAAjIsSaIdS8ABj5LduAAAwHCXQiChugKmZ9+eesgiAncS4HxjKXCVQVV2Q5LVJdiV5fXdfse71RyZ5U5JvSfKpJD/Y3Yc2N+pi84Me4PiZQQQP7VjjNACABxyzBKqqXUmuTPKMJIeT3FRVB7r7tjWHPS/JZ7r7a6vqkiSvSPKDWxF4HgoXgJ3teH/OK4/YaeYcpwEAJJlvJtB5Se7o7juTpKquSnJxkrWDi4uTvGz2/O1JXldV1d29iVkB4IQoj9iB5hmnAQAkma8EOiPJXWu2Dyf51oc6prvvr6rPJXlckns2IyQADGknzTBVaO0484zTAACSbPPC0FW1L8m+2ebnq+pj2/n5C+r0KNPGyrkZJ+dlvJybEahXfMmurTwv/2KL3peHyRjsuPiZNV6TODcb/Lwe++ednuSeBcy9qe8zUpP4M7OABhmDzVMCHUly5prtPbN9Gx1zuKp2J3lsVheIfpDu3p9k/xyfyUxVHezu5aFz8KWcm3FyXsbLuRkn52XhzTNOMwY7Dv5sjJdzM07Oy3g5N+M01Hl5xBzH3JTknKo6u6pOTnJJkgPrjjmQ5Ednz78vye9bDwgAYMvNM04DAEgyx0yg2Ro/lyW5Lqu3Hn1Dd99aVS9PcrC7DyT5tSS/WVV3JPl0VgcgAABsoYcapw0cCwAYqbnWBOrua5Ncu27fS9c8/9sk37+50ZgxdXu8nJtxcl7Gy7kZJ+dlwW00TmNT+LMxXs7NODkv4+XcjNMg56VctQUAAACw882zJhAAAAAAC04JtECq6kVV1VV1+tBZWFVVr6qqj1bVh6vqXVV12tCZpqyqLqiqj1XVHVV1+dB5SKrqzKq6vqpuq6pbq+r5Q2fiwapqV1V9oKrePXQWGCtjsHEx/hofY7DxMQYbtyHHX0qgBVFVZyb57iQfHzoLD/K+JI/v7m9M8qdJXjJwnsmqql1JrkzyzCTnJrm0qs4dNhVJ7k/you4+N8mTkvyk8zI6z09y+9AhYKyMwUbJ+GtEjMFGyxhs3AYbfymBFsf/TPLiJBZxGpHufm933z/bvCHJniHzTNx5Se7o7ju7+74kVyW5eOBMk9fdn+zuW2bP783q/+zOGDYVD6iqPUkuTPL6obPAiBmDjYzx1+gYg42QMdh4DT3+UgItgKq6OMmR7v7Q0Fk4qh9P8p6hQ0zYGUnuWrN9OP5HNypVtTfJE5PcOGwS1nhNVv9y+49DB4ExMgZbCMZfwzMGGzljsNEZdPw11y3i2XpV9XtJvnqDl346yX/N6jRkBnC0c9PdvzM75qezOuXyLduZDRZFVT0myTuSvKC7/3roPCRV9ewkd3f3zVX1lKHzwFCMwcbJ+As2hzHYuIxh/KUEGonufvpG+6vqCUnOTvKhqkpWp7veUlXndfdfbmPEyXqoc/OAqvqxJM9O8rTuNlV8OEeSnLlme89sHwOrqpOyOvh4S3e/c+g8/JPzk1xUVc9K8qgkX15Vb+7u5wycC7aVMdg4GX8tFGOwkTIGG6XBx1/lZ+ZiqapDSZa7+56hs7B6J4Qkv5jkO7t7Zeg8U1ZVu7O6OOTTsjrwuCnJD3X3rYMGm7ha/ZvTbyT5dHe/YOg8bGz2m6j/1N3PHjoLjJUx2HgYf42LMdg4GYON31DjL2sCwYl5XZJTk7yvqj5YVb8ydKCpmi0QeVmS67K68N3VBh+jcH6SH07y1NmfkQ/OfvMBAMfL+GtEjMFGyxiMDZkJBAAAADABZgIBAAAATIASCAAAAGAClEAAAAAAE6AEAgAAAJgAJRAAAADABCiBAAAAACZACQQAAAAwAUogAAAAgAn4/3S4QG3pMeJvAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1440x360 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "### distribution of logR values in GDSC and CCLE \n",
    "cn_values_gdsc  = []\n",
    "for row in df.iterrows():\n",
    "    cn_values_gdsc += [row[1][-1]]\n",
    "cn_values_ccle = []\n",
    "for row in cna_table.iterrows():\n",
    "    cn_values_ccle+=  [row[1][-1]]\n",
    "\n",
    "cn_values_gdsc = sorted (cn_values_gdsc)\n",
    "cn_values_ccle = sorted (cn_values_ccle)\n",
    "plt.figure(figsize=(20,5))\n",
    "plt.subplot(121)\n",
    "tmp = plt.hist(cn_values_gdsc,bins=100,density = True,range=(-5,4))\n",
    "plt.title(\"GDSC\")\n",
    "plt.subplot(122)\n",
    "tmp = plt.hist(cn_values_ccle,bins=100,density = True, range=(-5,4))\n",
    "plt.title(\"CCLE\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
